Index: src/core/org/apache/jmeter/gui/util/FileDialoger.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/core/org/apache/jmeter/gui/util/FileDialoger.java	(revision de4fe0c36b4ea94ef80f08b7339bca8a80cb6b17)
+++ src/core/org/apache/jmeter/gui/util/FileDialoger.java	(revision )
@@ -34,7 +34,7 @@
     /**
      * The last directory visited by the user while choosing Files.
      */
-    private static String lastJFCDirectory = null;
+    private static String lastJFCDirectory = System.getProperty("user.dir", "");
 
     private static JFileChooser jfc = new JFileChooser();
 
