Index: src/components/org/apache/jmeter/visualizers/Spline3.java
===================================================================
--- src/components/org/apache/jmeter/visualizers/Spline3.java	(revision 1489981)
+++ src/components/org/apache/jmeter/visualizers/Spline3.java	(working copy)
@@ -271,6 +271,7 @@
                 lineSum = lineSum - Math.abs(_A[i][i]);
                 if (lineSum > Math.abs(_A[i][i])) {
                     converge = false;
+                    break;
                 }
             }
         }
