--- jakarta-jmeter-2.1.1/src/protocol/http/org/apache/jmeter/protocol/http/sampler/SoapSampler.java	2005-07-12 13:50:56.000000000 -0700
+++ ../src/jakarta-jmeter-2.1.1/src/protocol/http/org/apache/jmeter/protocol/http/sampler/SoapSampler.java	2006-03-08 13:55:32.000000000 -0800
@@ -42,8 +42,9 @@
 	private static Logger log = LoggingManager.getLoggerForClass();
 
 	public static final String XML_DATA = "HTTPSamper.xml_data";
-
 	public static final String URL_DATA = "SoapSampler.URL_DATA";
+	public static final String SOAP_ACTION = "SoapSampler.SOAP_ACTION";
+	public static final String SEND_SOAP_ACTION = "SoapSampler.SEND_SOAP_ACTION";
 
 	public void setXmlData(String data) {
 		setProperty(XML_DATA, data);
@@ -61,6 +62,22 @@
 		setProperty(URL_DATA, url);
 	}
 
+	public String getSOAPAction() {
+		return getPropertyAsString(SOAP_ACTION);
+	}
+
+	public void setSOAPAction(String action) {
+		setProperty(SOAP_ACTION, action);
+	}
+
+	public boolean getSendSOAPAction() {
+		return getPropertyAsBoolean(SEND_SOAP_ACTION);
+	}
+
+	public void setSendSOAPAction(boolean action) {
+		setProperty(SEND_SOAP_ACTION, String.valueOf(action));
+	}
+
 	/**
 	 * Set the HTTP request headers in preparation to open the connection and
 	 * sending the POST data.
@@ -90,6 +107,7 @@
 		} else {
 			// otherwise we use "text/xml" as the default
 			connection.setRequestProperty("Content-Type", "text/xml");
+			if(getSendSOAPAction()) connection.setRequestProperty("SOAPAction", "\"" + getSOAPAction() + "\"");
 		}
 		connection.setDoOutput(true);
 	}
