Index: src/core/org/apache/jmeter/threads/JMeterThread.java
===================================================================
--- src/core/org/apache/jmeter/threads/JMeterThread.java	(Revision 1860784)
+++ src/core/org/apache/jmeter/threads/JMeterThread.java	(Arbeitskopie)
@@ -539,6 +539,9 @@
         threadVars.putObject(PACKAGE_OBJECT, pack);
 
         delay(pack.getTimers());
+        if (scheduler) {
+            stopSchedulerIfNeeded();
+        }
         SampleResult result = null;
         if (running) {
             Sampler sampler = pack.getSampler();
