Index: src/core/org/apache/jmeter/threads/JMeterThread.java
===================================================================
--- src/core/org/apache/jmeter/threads/JMeterThread.java	(revision 1647388)
+++ src/core/org/apache/jmeter/threads/JMeterThread.java	(working copy)
@@ -437,6 +437,14 @@
                     result.setGroupThreads(threadGroup.getNumberOfThreads());
                     result.setAllThreads(JMeterContextService.getNumberOfThreads());
                     result.setThreadName(threadName);
+                    SampleResult[]subResults = result.getSubResults();
+                    if(subResults != null) {
+                        for (SampleResult subResult : subResults) {
+                            subResult.setGroupThreads(threadGroup.getNumberOfThreads());
+                            subResult.setAllThreads(JMeterContextService.getNumberOfThreads());
+                            subResult.setThreadName(threadName);
+                        }
+                    }
                     threadContext.setPreviousResult(result);
                     runPostProcessors(pack.getPostProcessors());
                     checkAssertions(pack.getAssertions(), result, threadContext);
Index: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	(revision 1647388)
+++ src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	(working copy)
@@ -54,8 +54,8 @@
 import org.apache.jmeter.protocol.http.control.CacheManager;
 import org.apache.jmeter.protocol.http.control.Cookie;
 import org.apache.jmeter.protocol.http.control.CookieManager;
-import org.apache.jmeter.protocol.http.control.HeaderManager;
 import org.apache.jmeter.protocol.http.control.DNSCacheManager;
+import org.apache.jmeter.protocol.http.control.HeaderManager;
 import org.apache.jmeter.protocol.http.parser.HTMLParseException;
 import org.apache.jmeter.protocol.http.parser.HTMLParser;
 import org.apache.jmeter.protocol.http.util.ConversionUtils;
@@ -1206,9 +1206,6 @@
         // Iterate through the URLs and download each image:
         if (urls != null && urls.hasNext()) {
             if (container == null) {
-                // TODO needed here because currently done on sample completion in JMeterThread,
-                // but that only catches top-level samples.
-                res.setThreadName(Thread.currentThread().getName());
                 container = new HTTPSampleResult(res);
                 container.addRawSubResult(res);
             }
