Index: src/protocol/jms/org/apache/jmeter/protocol/jms/client/ReceiveSubscriber.java
===================================================================
--- src/protocol/jms/org/apache/jmeter/protocol/jms/client/ReceiveSubscriber.java	(revision 791718)
+++ src/protocol/jms/org/apache/jmeter/protocol/jms/client/ReceiveSubscriber.java	(working copy)
@@ -84,8 +84,8 @@
                 ConnectionFactory.getTopicConnectionFactory(ctx,connfactory);
                 _conn = ConnectionFactory.getTopicConnection();
                 _topic = InitialContextFactory.lookupTopic(ctx, topic);
-                _session = this.CONN.createTopicSession(false, TopicSession.AUTO_ACKNOWLEDGE);
-                _subscriber = this.SESSION.createSubscriber(this.TOPIC);
+                _session = _conn.createTopicSession(false, TopicSession.AUTO_ACKNOWLEDGE);
+                _subscriber = _session.createSubscriber(this.TOPIC);
                 log.info("created the topic connection successfully");
             } catch (JMSException e) {
                 log.error("Connection error: " + e.getMessage());
