From e1824eebb323129278949bf0a079aea396115317 Mon Sep 17 00:00:00 2001
From: Felix Schumacher <felix.schumacher@internetallee.de>
Date: Sat, 20 Jan 2018 14:49:22 +0100
Subject: [PATCH] Use a setenv.sh script if available to preset variables.

---
 bin/jmeter     | 80 ++++++++++++++++++++++++++++++++++++++++++----------------
 bin/jmeter.bat | 57 +++++++++++++++++++++++++++++++++++++----
 2 files changed, 110 insertions(+), 27 deletions(-)

diff --git a/bin/jmeter b/bin/jmeter
index 1060110..95f8b80 100755
--- a/bin/jmeter
+++ b/bin/jmeter
@@ -22,7 +22,59 @@
 ##   e.g.
 ##   JVM_ARGS="-Xms512m -Xmx512m" jmeter etc.
 ##
+##   Do not set the variables in this script. Instead put them into a script
+##   setenv.sh in JMETER_HOME/bin to keep your customizations separate.
+##
+##   JAVA_HOME        Must point at your Java Development Kit installation.
+##                    Required to run the with the "debug" argument.
+##
+##   JRE_HOME         Must point at your Java Runtime installation.
+##                    Defaults to JAVA_HOME if empty. If JRE_HOME and JAVA_HOME
+##                    are both empty, JMeter will try to guess JAVA_HOME.
+##                    If JRE_HOME and JAVA_HOME are both set, JAVA_HOME is used.
+##
+##   JMETER_GC_ALGO   (Optional) Java runtime options to specify JVM garbage collection
+##                    algorithm
+##                    Defaults to "-XX:+UseG1GC -XX:MaxGCPauseMillis=250 -XX:G1ReservePercent=20"
+##
+##   JMETER_HEAP      (Optional) Java runtime options for memory management
+##                    used when JMeter is started.
+##                    Defaults to "-Xms512m -Xmx512m -X:MaxMetaspaceSize=256m"
+##
+##   JMETER_HOME      (Optional) May point to your JMeter install dir. If empty
+##                    it will be set relativ to this script.
+##
+##   JMETER_LANGUAGE  (Optional) Java runtime options to specify used language
+##                    Defaults to "-Duser.language=en -Duser.region=EN"
+##
+##   JMETER_OPTS      (Optional) Java runtime options used when JMeter is started.
+##                    Special options for operating systems might be added by JMeter.
+##
 ##   ==============================================
+
+# resolve links - $0 may be a softlink (code as used by Tomcat)
+# N.B. readlink would be a lot simpler but is not supported on Solaris
+PRG="$0"
+
+while [ -h "$PRG" ]; do
+  ls=`ls -ld "$PRG"`
+  link=`expr "$ls" : '.*-> \(.*\)$'`
+  if expr "$link" : '/.*' > /dev/null; then
+    PRG="$link"
+  else
+    PRG=`dirname "$PRG"`/"$link"
+  fi
+done
+
+PRGDIR=`dirname "$PRG"`
+
+# Only set JMETER_HOME if not already set
+[ -z "$JMETER_HOME" ] && JMETER_HOME=`cd "$PRGDIR/.." >/dev/null; pwd`
+
+if [ -r "${JMETER_HOME}/setenv.sh" ]; then
+  . "${JMETER_HOME}/setenv.sh"
+fi
+
 # Make sure prerequisite environment variables are set
 if [ -z "$JAVA_HOME" -a -z "$JRE_HOME" ]; then
   if [ "`uname`" = "Darwin" ]; then
@@ -77,27 +129,11 @@ if [ "$CURRENT_VERSION" -gt "$MINIMAL_VERSION" ]; then
     ADD_MODS="--add-modules java.activation"
 fi
 
-# resolve links - $0 may be a softlink (code as used by Tomcat)
-# N.B. readlink would be a lot simpler but is not supported on Solaris
-PRG="$0"
-
-while [ -h "$PRG" ]; do
-  ls=`ls -ld "$PRG"`
-  link=`expr "$ls" : '.*-> \(.*\)$'`
-  if expr "$link" : '/.*' > /dev/null; then
-    PRG="$link"
-  else
-    PRG=`dirname "$PRG"`/"$link"
-  fi
-done
-
-PRGDIR=`dirname "$PRG"`
-
-JMETER_OPTS=""
+: "${JMETER_OPTS:=""}"
 case `uname` in
    Darwin*)
    # Add Mac-specific property - should be ignored elsewhere (Bug 47064)
-   JMETER_OPTS="-Xdock:name=JMeter -Xdock:icon=\"${PRGDIR}/../docs/images/jmeter_square.png\" -Dapple.laf.useScreenMenuBar=true -Dapple.eawt.quitStrategy=CLOSE_ALL_WINDOWS"
+   JMETER_OPTS="${JMETER_OPTS} -Xdock:name=JMeter -Xdock:icon=\"${PRGDIR}/../docs/images/jmeter_square.png\" -Dapple.laf.useScreenMenuBar=true -Dapple.eawt.quitStrategy=CLOSE_ALL_WINDOWS"
    ;;
 esac
 
@@ -120,11 +156,11 @@ esac
 
 # This is the base heap size -- you may increase or decrease it to fit your
 # system's memory availability:
-HEAP="-Xms512m -Xmx512m -XX:MaxMetaspaceSize=256m"
+: "${JMETER_HEAP:="-Xms512m -Xmx512m -XX:MaxMetaspaceSize=256m"}"
 
 # Set language
 # Default to en_EN
-JMETER_LANGUAGE="-Duser.language=en -Duser.region=EN"
+: "${JMETER_LANGUAGE="-Duser.language=en -Duser.region=EN"}"
 
 # Uncomment this to generate GC verbose file with Java prior to 9
 # VERBOSE_GC="-verbose:gc -Xloggc:gc_jmeter_%p.log -XX:+PrintGCDetails -XX:+PrintGCCause -XX:+PrintTenuringDistribution -XX:+PrintHeapAtGC -XX:+PrintGCApplicationConcurrentTime -XX:+PrintAdaptiveSizePolicy -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCDateStamps"
@@ -137,7 +173,7 @@ JMETER_LANGUAGE="-Duser.language=en -Duser.region=EN"
 # RUN_IN_DOCKER="-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap"
 
 # Finally, some tracing to help in case things go astray:
-GC_ALGO="-XX:+UseG1GC -XX:MaxGCPauseMillis=250 -XX:G1ReservePercent=20"
+: "${JMETER_GC_ALGO:="-XX:+UseG1GC -XX:MaxGCPauseMillis=250 -XX:G1ReservePercent=20"}"
 
 
 # Always dump on OOM (does not cost anything unless triggered)
@@ -145,6 +181,6 @@ DUMP="-XX:+HeapDumpOnOutOfMemoryError"
 SYSTEM_PROPS="-Djava.security.egd=file:/dev/urandom"
 SERVER="-server"
 
-ARGS="$SERVER $DUMP $HEAP $VERBOSE_GC $GC_ALGO $SYSTEM_PROPS $JMETER_LANGUAGE $RUN_IN_DOCKER"
+ARGS="$SERVER $DUMP $JMETER_HEAP $VERBOSE_GC $JMETER_GC_ALGO $SYSTEM_PROPS $JMETER_LANGUAGE $RUN_IN_DOCKER"
 
 "$JAVA_HOME/bin/java" $ADD_MODS $ARGS $JVM_ARGS $JMETER_OPTS -jar "$PRGDIR/ApacheJMeter.jar" "$@"
diff --git a/bin/jmeter.bat b/bin/jmeter.bat
index 193d83f..4bd723c 100644
--- a/bin/jmeter.bat
+++ b/bin/jmeter.bat
@@ -18,18 +18,59 @@ rem   limitations under the License.
 rem   =====================================================
 rem   Environment variables that can be defined externally:
 rem
-rem   JMETER_BIN - JMeter bin directory (must end in \)
-rem   JM_LAUNCH - java.exe (default) or javaw.exe
-rem   JVM_ARGS - additional java options, e.g. -Dprop=val
-rem   JM_START - set this to "start" to launch JMeter in a separate window
-rem              this is used by the jmeterw.cmd script.
+rem   Do not set the variables in this script. Instead put them into a script
+rem   setenv.bat in JMETER_HOME/bin to keep your customizations separate.
+rem
+rem   DDRAW       - (Optional) JVM options to influence usage of direct draw,
+rem                 e.g. '-Dsun.java2d.ddscale=true'
+rem
+rem   JMETER_BIN  - JMeter bin directory (must end in \)
+rem
+rem   JMETER_HOME - installation directory. Will be guessed from location of jmeter.bat
+rem
+rem   JM_LAUNCH   - java.exe (default) or javaw.exe
+rem
+rem   JM_START    - set this to "start" to launch JMeter in a separate window
+rem                 this is used by the jmeterw.cmd script.
+rem
+rem   JVM_ARGS    - (Optional) Java options used when starting JMeter, e.g. -Dprop=val
+rem                 Defaults to '-Duser.language="en" -Duser.region="EN"'
+rem
+rem   GC_ALGO     - (Optional) JVM garbage collector options 
+rem                 Defaults to '-XX:+UseG1GC -XX:MaxGCPauseMillis=250 -XX:G1ReservePercent=20'
+rem
+rem   HEAP        - (Optional) JVM memory settings used when starting JMeter
+rem                 Defaults to '-Xms512m -Xmx512m -XX:MaxMetaspaceSize=256m'
 rem
 rem   =====================================================
 
 setlocal
+
+rem Guess JMETER_HOME if not defined
+set "CURRENT_DIR=%cd%"
+if not "%JMETER_HOME%" == "" goto gotHome
+set "JMETER_HOME=%CURRENT_DIR%"
+if exist "%JMETER_HOME%\bin\jmeter.bat" goto okHome
+cd ..
+set "JMETER_HOME=%cd%"
+cd "%CURRENT_DIR%"
+:gotHome
+
+if exist "%JMETER_HOME%\bin\jmeter.bat" goto okHome
+echo The JMETER_HOME environment variable is not defined correctly
+echo This environment variable is needed to run this program
+goto end
+:okHome
+
+rem Get standard environment variables
+if exist "%JMETER_HOME%\bin\setenv.bat" call "%JMETER_HOME%\bin\setenv.bat"
+
+if not "%JVM_ARGS%" == "" goto gotJvmArgs
 # Set language
 # Default to en_EN
 set JVM_ARGS=-Duser.language="en" -Duser.region="EN"
+:gotJvmArgs
+
 rem Minimal version to run JMeter
 set MINIMAL_VERSION=1.8.0
 
@@ -91,9 +132,11 @@ set JMETER_CMD_LINE_ARGS=%*
 rem The following link describes the -XX options:
 rem http://docs.oracle.com/javase/8/docs/technotes/tools/unix/java.html
 
+if not "%HEAP%" == "" goto gotHeap
 rem See the unix startup file for the rationale of the following parameters,
 rem including some tuning recommendations
 set HEAP=-Xms512m -Xmx512m -XX:MaxMetaspaceSize=256m
+:gotHeap
 
 rem Uncomment this to generate GC verbose file with Java prior to 9 
 rem set VERBOSE_GC=-verbose:gc -Xloggc:gc_jmeter_%%p.log -XX:+PrintGCDetails -XX:+PrintGCCause -XX:+PrintTenuringDistribution -XX:+PrintHeapAtGC -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCDateStamps -XX:+PrintAdaptiveSizePolicy
@@ -101,7 +144,9 @@ rem set VERBOSE_GC=-verbose:gc -Xloggc:gc_jmeter_%%p.log -XX:+PrintGCDetails -XX
 rem Uncomment this to generate GC verbose file with Java 9 and above
 rem set VERBOSE_GC=-Xlog:gc*,gc+age=trace,gc+heap=debug:file=gc_jmeter_%%p.log
 
+if not "%GC_ALGO%" == "" goto :gotGcAlgo
 set GC_ALGO=-XX:+UseG1GC -XX:MaxGCPauseMillis=250 -XX:G1ReservePercent=20
+:gotGcAlgo
 
 set SYSTEM_PROPS=-Djava.security.egd=file:/dev/urandom
 
@@ -115,6 +160,7 @@ rem set RUN_IN_DOCKER=-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimit
 rem Additional settings that might help improve GUI performance on some platforms
 rem See: http://www.oracle.com/technetwork/java/perf-graphics-135933.html
 
+if not "%DDRAW%" == "" goto gotDDraw
 set DDRAW=
 rem  Setting this flag to true turns off DirectDraw usage, which sometimes helps to get rid of a lot of rendering problems on Win32.
 rem set DDRAW=%DDRAW% -Dsun.java2d.noddraw=true
@@ -124,6 +170,7 @@ rem set DDRAW=%DDRAW% -Dsun.java2d.ddoffscreen=false
 
 rem Setting this flag to true enables hardware-accelerated scaling.
 rem set DDRAW=%DDRAW% -Dsun.java2d.ddscale=true
+:gotDDraw
 
 rem Server mode
 set SERVER=-server
-- 
2.7.4

