diff --git a/src/core/org/apache/jmeter/samplers/SampleResult.java b/src/core/org/apache/jmeter/samplers/SampleResult.java
index a9df618f4..fe3792a81 100644
--- a/src/core/org/apache/jmeter/samplers/SampleResult.java
+++ b/src/core/org/apache/jmeter/samplers/SampleResult.java
@@ -614,7 +614,10 @@ public class SampleResult implements Serializable, Cloneable, Searchable {
      *            the {@link SampleResult} to be added
      */
     public void addSubResult(SampleResult subResult) {
-        addSubResult(subResult, true);
+        if (getSampleLabel().equals(subResult.getSampleLabel())) {
+            addSubResult(subResult, true);
+        }
+        addSubResult(subResult, false);
     }
     /**
      * Add a subresult and adjust the parent byte count and end-time.
@@ -652,7 +655,10 @@ public class SampleResult implements Serializable, Cloneable, Searchable {
      *            the {@link SampleResult} to be added
      */
     public void addRawSubResult(SampleResult subResult){
-        storeSubResult(subResult, true);
+        if (getSampleLabel().equals(subResult.getSampleLabel())) {
+            storeSubResult(subResult, true);
+        }
+        storeSubResult(subResult, false);
     }
     
     /**
@@ -676,7 +682,10 @@ public class SampleResult implements Serializable, Cloneable, Searchable {
      *            the {@link SampleResult} to be added
      */
     public void storeSubResult(SampleResult subResult) {
-        storeSubResult(subResult, true);
+        if (getSampleLabel().equals(subResult.getSampleLabel())) {
+            storeSubResult(subResult, true);
+        }
+        storeSubResult(subResult, false);
     }
     
     /**
