Index: C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/components/org/apache/jmeter/modifiers/CounterConfig.java
===================================================================
--- C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/components/org/apache/jmeter/modifiers/CounterConfig.java	(revision 454855)
+++ C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/components/org/apache/jmeter/modifiers/CounterConfig.java	(working copy)
@@ -116,7 +116,11 @@
 	}
 
 	public long getEnd() {
-		return getPropertyAsLong(END);
+		long propertyAsLong = getPropertyAsLong(END);
+		if (propertyAsLong == 0 && "".equals(getProperty(END).getStringValue())) {
+			propertyAsLong = Long.MAX_VALUE;
+		}
+		return propertyAsLong;
 	}
 
 	public void setIncrement(long inc) {
