Index: test/src/org/apache/jmeter/junit/JMeterTest.java
===================================================================
--- test/src/org/apache/jmeter/junit/JMeterTest.java	(revision 1049987)
+++ test/src/org/apache/jmeter/junit/JMeterTest.java	(working copy)
@@ -36,6 +36,7 @@
 import java.util.Set;
 
 import javax.swing.JComponent;
+import javax.swing.JDialog;
 
 import junit.framework.Test;
 import junit.framework.TestSuite;
@@ -486,7 +487,11 @@
      * Test serializable elements - test the object
      */
     public void runSerialTest() throws Exception {
-        if (!(serObj instanceof JComponent)) {
+        if (
+            ! (serObj instanceof JComponent ||
+              serObj instanceof JDialog )
+            ) {
+            
             try {
                 ByteArrayOutputStream bytes = new ByteArrayOutputStream();
                 ObjectOutputStream out = new ObjectOutputStream(bytes);
