Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java	(Revision 1836751)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java	(Arbeitskopie)
@@ -1560,7 +1560,10 @@
             keyPassword = storePassword; // we use same password for both
             setPassword(storePassword);
             log.info(
-                    "Creating HTTP(S) Test Script Recorder Root CA in {}, ensure you install certificate in your Browser for recording",
+                    "Creating HTTP(S) Test Script Recorder Root CA that is valid for {} days in {},"
+                    + " ensure you install certificate in your Browser for recording. You can change"
+                    + " the validity period for newly generated Test CAs by defining the property proxy.cert.validity.",
+                    CERT_VALIDITY,
                     CERT_PATH_ABS);
             KeyToolUtils.generateProxyCA(CERT_PATH, storePassword, CERT_VALIDITY);
             log.info("Created keystore in {}", CERT_PATH_ABS);
