diff --git a/extras/proxycert.sh b/extras/proxycert.sh
old mode 100644
new mode 100755
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
index 6e657ba..9b75d75 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
@@ -384,15 +384,15 @@ public abstract class HTTPSamplerBase extends AbstractSampler
         if (getPostBodyRaw()) {
             return true;
         } else {
-            boolean noArgumentsHasName = true;
+            boolean hasArguments = false;
             for (JMeterProperty jMeterProperty : getArguments()) {
+                hasArguments = true;
                 HTTPArgument arg = (HTTPArgument) jMeterProperty.getObjectValue();
                 if (arg.getName() != null && arg.getName().length() > 0) {
-                    noArgumentsHasName = false;
-                    break;
+                    return false;
                 }
             }
-            return noArgumentsHasName;
+            return hasArguments;
         }
     }
 
