Index: ViewResultsFullVisualizer.java
===================================================================
--- ViewResultsFullVisualizer.java	(revision 1181023)
+++ ViewResultsFullVisualizer.java	(working copy)
@@ -150,7 +148,8 @@
             jTree.expandPath(new TreePath(root));
         }
         if (autoScrollCB.isSelected() && root.getChildCount() > 1) {
-            jTree.scrollRowToVisible(root.getChildCount() - 1);
+            jTree.scrollPathToVisible(new TreePath(new Object[] { root,
+                    treeModel.getChild(root, root.getChildCount() - 1) }));
         }
     }
