Index: JDBCSamplerBeanInfo.java
===================================================================
--- JDBCSamplerBeanInfo.java	(revision 388876)
+++ JDBCSamplerBeanInfo.java	(working copy)
@@ -50,7 +50,7 @@
 		p.setValue(NOT_UNDEFINED, Boolean.TRUE);
 		p.setValue(DEFAULT, JDBCSampler.SELECT);
 		p.setValue(NOT_OTHER,Boolean.TRUE);
-		p.setValue(TAGS,new String[]{JDBCSampler.SELECT,"Update Statement"});
+		p.setValue(TAGS,new String[]{JDBCSampler.SELECT,JDBCSampler.UPDATE,JDBCSampler.STATEMENT});
 
 		p = property("query");
 		p.setValue(NOT_UNDEFINED, Boolean.TRUE);
