Index: src/monitor/components/org/apache/jmeter/monitor/util/Stats.java
===================================================================
--- src/monitor/components/org/apache/jmeter/monitor/util/Stats.java	(revision 1166194)
+++ src/monitor/components/org/apache/jmeter/monitor/util/Stats.java	(working copy)
@@ -118,7 +118,7 @@
                 return WARNING;
             } else if (per >= ACTIVE_PER && per <= WARNING_PER) {
                 return ACTIVE;
-            } else if (per < ACTIVE_PER && per > HEALTHY_PER) {
+            } else if (per < ACTIVE_PER && per >= HEALTHY_PER) {
                 return HEALTHY;
             } else {
                 return DEAD;
@@ -127,6 +127,7 @@
             return DEAD;
         }
     }
+  
 
     /**
      * Method will calculate the memory load: used / max = load. The load value
