Index: MainFrame.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/gui/MainFrame.java,v
retrieving revision 1.20
diff -u -r1.20 MainFrame.java
--- MainFrame.java	12 May 2003 21:31:06 -0000	1.20
+++ MainFrame.java	14 Jun 2003 15:45:13 -0000
@@ -426,7 +426,7 @@
 
         // The setResizeWeight() method was added to JDK1.3. For now, JMeter
         // should remain compatible with JDK1.2.
-        //treeAndMain.setResizeWeight(.2);
+        treeAndMain.setResizeWeight(.2); //TODO: requires JDK1.3 is this now OK?
         treeAndMain.setContinuousLayout(true);
         all.add(treeAndMain, BorderLayout.CENTER);
 
