Index: C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/core/org/apache/jmeter/samplers/StatisticalSampleResult.java
===================================================================
--- C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/core/org/apache/jmeter/samplers/StatisticalSampleResult.java	(revision 466937)
+++ C:/Documents and Settings/ecerulm/workspace32jmeter/JMeter2_2/src/core/org/apache/jmeter/samplers/StatisticalSampleResult.java	(working copy)
@@ -52,6 +52,9 @@
 		}
 
 		// Set start/end times
+        if (getStartTime() == 0){
+            this.setStartTime(res.getStartTime());
+        }
 		this.setStartTime(Math.min(getStartTime(), res.getStartTime()));
 		this.setEndTime(Math.max(getEndTime(), res.getEndTime()));
 	}
