### Eclipse Workspace Patch 1.0
#P jmeter-svn
Index: src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/Receiver.java
===================================================================
--- src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/Receiver.java	(revision 780322)
+++ src/protocol/jms/org/apache/jmeter/protocol/jms/sampler/Receiver.java	(working copy)
@@ -81,7 +81,7 @@
                     if (reply.getJMSCorrelationID() == null) {
                         log.warn("Received message with correlation id null. Discarding message ...");
                     } else {
-                        MessageAdmin.getAdmin().putReply(reply.getJMSMessageID(), reply);
+                        MessageAdmin.getAdmin().putReply(reply.getJMSCorrelationID(), reply);
                     }
                 }
 
