*** /home/jmheight/jakarta-jmeter/src/org/apache/jmeter/protocol/http/config/UrlConfig.java	Wed Oct 24 19:34:31 2001
--- UrlConfig.java	Tue Oct 30 10:51:38 2001
***************
*** 106,109 ****
--- 106,116 ----
  	public final static String GET = "GET";
  
+         /** 
+          * Description of the field
+          */
+         public final static String BODY = "BODY";
+ 
+         public final static String CONTENT_TYPE = "CONTENT_TYPE";
+ 
  	/**
  	 *  Constructor for the UrlConfig object
