diff -u -r jakarta-jmeter-2.0.3-orig/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java jakarta-jmeter-2.0.3/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java
--- jakarta-jmeter-2.0.3-orig/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java	2005-03-06 12:45:28.000000000 +0100
+++ jakarta-jmeter-2.0.3/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java	2005-10-19 16:29:44.000000000 +0200
@@ -106,7 +106,6 @@
 	    sizePanel.add(xpath);
 	    
 	    xml = new XMLConfPanel();
-	    xpath = new XPathPanel();
 	    
 	    xml.setBorder(
 		        BorderFactory.createTitledBorder(
