Index: src/protocol/java/org/apache/jmeter/protocol/java/sampler/JSR223Sampler.java
===================================================================
--- src/protocol/java/org/apache/jmeter/protocol/java/sampler/JSR223Sampler.java	(revision 1828970)
+++ src/protocol/java/org/apache/jmeter/protocol/java/sampler/JSR223Sampler.java	(working copy)
@@ -29,6 +29,7 @@
 
 import org.apache.jmeter.config.ConfigTestElement;
 import org.apache.jmeter.engine.util.ConfigMergabilityIndicator;
+import org.apache.jmeter.gui.GUIMenuSortOrder;
 import org.apache.jmeter.samplers.Entry;
 import org.apache.jmeter.samplers.SampleResult;
 import org.apache.jmeter.samplers.Sampler;
@@ -37,7 +38,7 @@
 import org.apache.jmeter.util.JSR223TestElement;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
+@GUIMenuSortOrder(4)
 public class JSR223Sampler extends JSR223TestElement implements Cloneable, Sampler, TestBean, ConfigMergabilityIndicator {
     private static final Set<String> APPLIABLE_CONFIG_CLASSES = new HashSet<>(
             Arrays.asList("org.apache.jmeter.config.gui.SimpleConfigGui"));
