Index: src/components/org/apache/jmeter/visualizers/RequestViewRaw.java
===================================================================
--- src/components/org/apache/jmeter/visualizers/RequestViewRaw.java	(revision 1754657)
+++ src/components/org/apache/jmeter/visualizers/RequestViewRaw.java	(working copy)
@@ -87,6 +87,7 @@
             }
             if (sb.length() > 0) {
                 sampleDataField.setText(sb.toString());
+                sampleDataField.setCaretPosition(1);
             } else {
                 // add a message when no request data (ex. Java request)
                 sampleDataField.setText(JMeterUtils
Index: src/components/org/apache/jmeter/visualizers/SamplerResultTab.java
===================================================================
--- src/components/org/apache/jmeter/visualizers/SamplerResultTab.java	(revision 1754657)
+++ src/components/org/apache/jmeter/visualizers/SamplerResultTab.java	(working copy)
@@ -334,6 +334,7 @@
                 statsBuff.append(JMeterUtils.getResString("view_results_assertion_failure_message")).append(assertionResult.getFailureMessage()).append(NL); //$NON-NLS-1$
                 statsDoc.insertString(statsDoc.getLength(), statsBuff.toString(), null);
             }
+            stats.setCaretPosition(1);
         } catch (BadLocationException exc) {
             stats.setText(exc.getLocalizedMessage());
         }
