Index: TableDataModel.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/components/org/apache/jmeter/visualizers/TableDataModel.java,v
retrieving revision 1.8
diff -u -r1.8 TableDataModel.java
--- TableDataModel.java	3 Feb 2003 14:29:00 -0000	1.8
+++ TableDataModel.java	8 Jun 2003 19:13:04 -0000
@@ -64,7 +64,7 @@
 
 import org.apache.jmeter.samplers.SampleResult;
 import org.apache.jmeter.util.JMeterUtils;
-import org.apache.log.Hierarchy;
+import org.apache.jorphan.logging.LoggingManager;
 import org.apache.log.Logger;
 
 
@@ -78,8 +78,8 @@
  */
 public class TableDataModel extends GraphModel implements TableModel
 {
-    transient private static Logger log = Hierarchy.getDefaultHierarchy().getLoggerFor(
-                    "jmeter.gui");
+    transient private static Logger log = LoggingManager.getLoggerForClass();
+
     List urlList = new ArrayList();
 
     /**
@@ -207,10 +207,10 @@
         }
         else if (columnIndex == 1)
         {
-            log.info("rowIndex = " + rowIndex);
+            log.debug("rowIndex = " + rowIndex);
             if ((rowIndex >= 0) && (rowIndex < urlList.size()))
             {
-                log.info(" url = " + urlList.get(rowIndex));
+                log.debug(" url = " + urlList.get(rowIndex));
                 return urlList.get(rowIndex);
             }
         }
