diff --git a/bin/jmeter.properties b/bin/jmeter.properties
index 4fcc6a1..33fcc3a 100644
--- a/bin/jmeter.properties
+++ b/bin/jmeter.properties
@@ -91,7 +91,7 @@ xml.parser=org.apache.xerces.parsers.SAXParser
 # This is needed when server does not support Socket version negotiation, this can lead to:
 # javax.net.ssl.SSLPeerUnverifiedException: peer not authenticated
 # java.net.SocketException: Connection reset
-# see https://issues.apache.org/bugzilla/show_bug.cgi?id=54759
+# see https://bz.apache.org/bugzilla/show_bug.cgi?id=54759
 #https.socket.protocols=SSLv2Hello SSLv3 TLSv1
 
 # Control if we allow reuse of cached SSL context between iterations
@@ -132,7 +132,7 @@ jmeter.laf.mac=System
 
 # If none of the above jmeter.laf properties are defined, JMeter uses the CrossPlatform LAF.
 # This is because the CrossPlatform LAF generally looks better than the System LAF.
-# See https://issues.apache.org/bugzilla/show_bug.cgi?id=52026 for details
+# See https://bz.apache.org/bugzilla/show_bug.cgi?id=52026 for details
 # N.B. the laf can be defined in user.properties.
 
 # LoggerPanel display
@@ -667,7 +667,7 @@ HTTPResponse.parsers=htmlParser wmlParser
 # Define the HTML parser to be used.
 # Default parser:
 # This new parser (since 2.10) should perform better than all others
-# see https://issues.apache.org/bugzilla/show_bug.cgi?id=55632
+# see https://bz.apache.org/bugzilla/show_bug.cgi?id=55632
 #htmlParser.className=org.apache.jmeter.protocol.http.parser.LagartoBasedHtmlParser
 
 # Other parsers:
diff --git a/bin/templates/build-adv-web-test-plan.jmx b/bin/templates/build-adv-web-test-plan.jmx
index 354aca2..096e4c0 100644
--- a/bin/templates/build-adv-web-test-plan.jmx
+++ b/bin/templates/build-adv-web-test-plan.jmx
@@ -26,7 +26,7 @@
           </elementProp>
           <elementProp name="bugzilla_host" elementType="Argument">
             <stringProp name="Argument.name">bugzilla_host</stringProp>
-            <stringProp name="Argument.value">issues.apache.org</stringProp>
+            <stringProp name="Argument.value">bz.apache.org</stringProp>
             <stringProp name="Argument.metadata">=</stringProp>
           </elementProp>
         </collectionProp>
@@ -185,7 +185,7 @@
             <stringProp name="TestPlan.comments">Shows how to extract bug id from a link in response</stringProp>
             <stringProp name="RegexExtractor.useHeaders">false</stringProp>
             <stringProp name="RegexExtractor.refname">bugId</stringProp>
-            <stringProp name="RegexExtractor.regex">https://issues.apache.org/bugzilla/show_bug.cgi\?id=([^&quot;]+?)&quot;</stringProp>
+            <stringProp name="RegexExtractor.regex">https://bz.apache.org/bugzilla/show_bug.cgi\?id=([^&quot;]+?)&quot;</stringProp>
             <stringProp name="RegexExtractor.template">$1$</stringProp>
             <stringProp name="RegexExtractor.default">nv_bugId</stringProp>
             <stringProp name="RegexExtractor.match_number">0</stringProp>
diff --git a/bin/testfiles/jmeter_home_page.html b/bin/testfiles/jmeter_home_page.html
index a4e59af..8d0897b 100644
--- a/bin/testfiles/jmeter_home_page.html
+++ b/bin/testfiles/jmeter_home_page.html
@@ -55,7 +55,7 @@ Content-Type: text/html
 </li>
 		  				<li>	 <a href="./changes.html">Changes</a>
 </li>
-		  				<li>	 <a href="http://issues.apache.org/bugzilla/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&email1=&emailtype1=substring&emailassigned_to1=1&email2=&emailtype2=substring&emailreporter2=1&bugidtype=include&bug_id=&changedin=&votes=&chfieldfrom=&chfieldto=Now&chfieldvalue=&product=JMeter&short_desc=&short_desc_type=substring&long_desc=&long_desc_type=substring&bug_file_loc=&bug_file_loc_type=substring&keywords=&keywords_type=anywords&field0-0-0=noop&type0-0-0=noop&value0-0-0=&cmdtype=doit&order=Reuse+same+sort+as+last+time">Known Bugs</a>
+		  				<li>	 <a href="http://bz.apache.org/bugzilla/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&email1=&emailtype1=substring&emailassigned_to1=1&email2=&emailtype2=substring&emailreporter2=1&bugidtype=include&bug_id=&changedin=&votes=&chfieldfrom=&chfieldto=Now&chfieldvalue=&product=JMeter&short_desc=&short_desc_type=substring&long_desc=&long_desc_type=substring&bug_file_loc=&bug_file_loc_type=substring&keywords=&keywords_type=anywords&field0-0-0=noop&type0-0-0=noop&value0-0-0=&cmdtype=doit&order=Reuse+same+sort+as+last+time">Known Bugs</a>
 </li>
 		  				<li>	 <a href="./license.html">License</a>
 </li>
diff --git a/bin/testfiles/load_bug_list.jmx b/bin/testfiles/load_bug_list.jmx
index 8c9379a..e0dc7dd 100644
--- a/bin/testfiles/load_bug_list.jmx
+++ b/bin/testfiles/load_bug_list.jmx
@@ -210,7 +210,7 @@
           </elementProp>
         </collectionProp>
       </elementProp>
-      <stringProp name="HTTPSampler.domain">issues.apache.org</stringProp>
+      <stringProp name="HTTPSampler.domain">bz.apache.org</stringProp>
       <stringProp name="HTTPSampler.port">80</stringProp>
       <stringProp name="HTTPSampler.protocol">http</stringProp>
       <stringProp name="HTTPSampler.contentEncoding"></stringProp>
diff --git a/doap_JMeter.rdf b/doap_JMeter.rdf
index d28fe41..004e832 100644
--- a/doap_JMeter.rdf
+++ b/doap_JMeter.rdf
@@ -28,7 +28,7 @@
     <asfext:pmc rdf:resource="http://jmeter.apache.org/" />
     <shortdesc>Pure Java application for load and functional testing</shortdesc>
     <description>Apache JMeter may be used to test performance both on static and dynamic resources (files, Servlets, Perl scripts, Java Objects, Data Bases and Queries, FTP Servers and more). It can be used to simulate a heavy load on a server, network or object to test its strength or to analyze overall performance under different load types. You can use it to make a graphical analysis of performance or to test your server/script/object behavior under heavy concurrent load. </description>
-    <bug-database rdf:resource="http://issues.apache.org/bugzilla/" />
+    <bug-database rdf:resource="http://bz.apache.org/bugzilla/" />
     <mailing-list rdf:resource="http://jmeter.apache.org/mail2.html" />
     <download-page rdf:resource="http://jmeter.apache.org/download_jmeter.cgi" />
     <programming-language>Java</programming-language>
diff --git a/res/maven/ApacheJMeter_parent.pom b/res/maven/ApacheJMeter_parent.pom
index 89f0c47..67adbc3 100644
--- a/res/maven/ApacheJMeter_parent.pom
+++ b/res/maven/ApacheJMeter_parent.pom
@@ -41,7 +41,7 @@ under the License.
     </licenses>
     <issueManagement>
         <system>bugzilla</system>
-        <url>https://issues.apache.org/bugzilla/describecomponents.cgi?product=JMeter</url>
+        <url>https://bz.apache.org/bugzilla/describecomponents.cgi?product=JMeter</url>
     </issueManagement>
     <scm>
         <connection>http://svn.apache.org/repos/asf/jmeter/trunk/</connection>
diff --git a/src/components/org/apache/jmeter/visualizers/RenderAsHTML.java b/src/components/org/apache/jmeter/visualizers/RenderAsHTML.java
index 3c80233..7185bd6 100644
--- a/src/components/org/apache/jmeter/visualizers/RenderAsHTML.java
+++ b/src/components/org/apache/jmeter/visualizers/RenderAsHTML.java
@@ -95,7 +95,7 @@ public class RenderAsHTML extends SamplerResultTab implements ResultRenderer {
         /*
          * Get round problems parsing <META http-equiv='content-type'
          * content='text/html; charset=utf-8'> See
-         * http://issues.apache.org/bugzilla/show_bug.cgi?id=23315
+         * http://bz.apache.org/bugzilla/show_bug.cgi?id=23315
          *
          * Is this due to a bug in Java?
          */
@@ -111,7 +111,7 @@ public class RenderAsHTML extends SamplerResultTab implements ResultRenderer {
             resultsScrollPane.setViewportView(results);
         } catch (NumberFormatException e) {
             // Java Bug : http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=9001188.
-            // See https://issues.apache.org/bugzilla/show_bug.cgi?id=54586
+            // See https://bz.apache.org/bugzilla/show_bug.cgi?id=54586
             log.warn("An error occured rendering html code", e);
             results.setText("Failed to render HTML: " + e.getMessage() +", use Text renderer");            
         }
diff --git a/src/components/org/apache/jmeter/visualizers/backend/graphite/GraphiteBackendListenerClient.java b/src/components/org/apache/jmeter/visualizers/backend/graphite/GraphiteBackendListenerClient.java
index 4592821..b1123d3 100644
--- a/src/components/org/apache/jmeter/visualizers/backend/graphite/GraphiteBackendListenerClient.java
+++ b/src/components/org/apache/jmeter/visualizers/backend/graphite/GraphiteBackendListenerClient.java
@@ -155,7 +155,7 @@ public class GraphiteBackendListenerClient extends AbstractBackendListenerClient
         graphiteMetricsManager.addMetric(timestampInSeconds, contextName, METRIC_OK_COUNT, Integer.toString(metric.getSuccesses()));
         graphiteMetricsManager.addMetric(timestampInSeconds, contextName, METRIC_KO_COUNT, Integer.toString(metric.getFailures()));
         graphiteMetricsManager.addMetric(timestampInSeconds, contextName, METRIC_ALL_COUNT, Integer.toString(metric.getTotal()));
-        // See https://issues.apache.org/bugzilla/show_bug.cgi?id=57350
+        // See https://bz.apache.org/bugzilla/show_bug.cgi?id=57350
         if(metric.getTotal() > 0) { 
             if(metric.getSuccesses()>0) {
                 graphiteMetricsManager.addMetric(timestampInSeconds, contextName, METRIC_OK_MIN_RESPONSE_TIME, Double.toString(metric.getOkMinTime()));
diff --git a/src/core/org/apache/jmeter/engine/ClientJMeterEngine.java b/src/core/org/apache/jmeter/engine/ClientJMeterEngine.java
index 9cd9361..5eaab8f 100644
--- a/src/core/org/apache/jmeter/engine/ClientJMeterEngine.java
+++ b/src/core/org/apache/jmeter/engine/ClientJMeterEngine.java
@@ -107,7 +107,7 @@ public class ClientJMeterEngine implements JMeterEngine {
     public void runTest() throws JMeterEngineException {
         log.info("running clientengine run method");
         
-        // See https://issues.apache.org/bugzilla/show_bug.cgi?id=55510
+        // See https://bz.apache.org/bugzilla/show_bug.cgi?id=55510
         JMeterContextService.clearTotalThreads();
         HashTree testTree = test;
 
@@ -123,7 +123,7 @@ public class ClientJMeterEngine implements JMeterEngine {
             /*
              * Add fix for Deadlocks, see:
              * 
-             * See https://issues.apache.org/bugzilla/show_bug.cgi?id=48350
+             * See https://bz.apache.org/bugzilla/show_bug.cgi?id=48350
             */
             File baseDirRelative = FileServer.getFileServer().getBaseDirRelative();
             String scriptName = FileServer.getFileServer().getScriptName();
diff --git a/src/core/org/apache/jmeter/gui/tree/JMeterTreeListener.java b/src/core/org/apache/jmeter/gui/tree/JMeterTreeListener.java
index 43bc436..a34c9fd 100644
--- a/src/core/org/apache/jmeter/gui/tree/JMeterTreeListener.java
+++ b/src/core/org/apache/jmeter/gui/tree/JMeterTreeListener.java
@@ -154,8 +154,8 @@ public class JMeterTreeListener implements TreeSelectionListener, MouseListener,
         currentPath = e.getNewLeadSelectionPath();
         // Call requestFocusInWindow to ensure current component loses focus and
         // all values are correctly saved
-        // see https://issues.apache.org/bugzilla/show_bug.cgi?id=55103
-        // see https://issues.apache.org/bugzilla/show_bug.cgi?id=55459
+        // see https://bz.apache.org/bugzilla/show_bug.cgi?id=55103
+        // see https://bz.apache.org/bugzilla/show_bug.cgi?id=55459
         tree.requestFocusInWindow();
         actionHandler.actionPerformed(new ActionEvent(this, 3333, ActionNames.EDIT)); // $NON-NLS-1$
     }
diff --git a/src/core/org/apache/jmeter/samplers/SampleResult.java b/src/core/org/apache/jmeter/samplers/SampleResult.java
index 7ba91dc..4b2cb64 100644
--- a/src/core/org/apache/jmeter/samplers/SampleResult.java
+++ b/src/core/org/apache/jmeter/samplers/SampleResult.java
@@ -588,7 +588,7 @@ public class SampleResult implements Serializable, Cloneable {
      */
     public void addSubResult(SampleResult subResult) {
         if(subResult == null) {
-            // see https://issues.apache.org/bugzilla/show_bug.cgi?id=54778
+            // see https://bz.apache.org/bugzilla/show_bug.cgi?id=54778
             return;
         }
         String tn = getThreadName();
diff --git a/src/monitor/components/org/apache/jmeter/visualizers/MonitorAccumModel.java b/src/monitor/components/org/apache/jmeter/visualizers/MonitorAccumModel.java
index 7394540..d422726 100644
--- a/src/monitor/components/org/apache/jmeter/visualizers/MonitorAccumModel.java
+++ b/src/monitor/components/org/apache/jmeter/visualizers/MonitorAccumModel.java
@@ -174,7 +174,7 @@ public class MonitorAccumModel implements Clearable, Serializable {
                 } 
                 // This part of code throws NullPointerException
                 // Don't think Monitor results can be loaded from files
-                // see https://issues.apache.org/bugzilla/show_bug.cgi?id=51810
+                // see https://bz.apache.org/bugzilla/show_bug.cgi?id=51810
 //                else {
 //                    noResponse(surl);
 //                }
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/parser/LagartoBasedHtmlParser.java b/src/protocol/http/org/apache/jmeter/protocol/http/parser/LagartoBasedHtmlParser.java
index ff72460..97cbd09 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/parser/LagartoBasedHtmlParser.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/parser/LagartoBasedHtmlParser.java
@@ -218,7 +218,7 @@ public class LagartoBasedHtmlParser extends HTMLParser {
             lagartoParser.parse(tagVisitor);
             return coll.iterator();
         } catch (LagartoException e) {
-            // TODO is it the best way ? https://issues.apache.org/bugzilla/show_bug.cgi?id=55634
+            // TODO is it the best way ? https://bz.apache.org/bugzilla/show_bug.cgi?id=55634
             if(log.isDebugEnabled()) {
                 log.debug("Error extracting embedded resource URLs from:'"+baseUrl+"', probably not text content, message:"+e.getMessage());
             }
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/parser/URLCollection.java b/src/protocol/http/org/apache/jmeter/protocol/http/parser/URLCollection.java
index 0c69a18..fb04184 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/parser/URLCollection.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/parser/URLCollection.java
@@ -92,7 +92,7 @@ public class URLCollection {
             }
             // No point in adding the URL as String as it will result in null 
             // returned during iteration, see URLString
-            // See https://issues.apache.org/bugzilla/show_bug.cgi?id=55092
+            // See https://bz.apache.org/bugzilla/show_bug.cgi?id=55092
             return false;
         }
         return b;
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java b/src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java
index df89c85..ce43ec8 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java
@@ -271,7 +271,7 @@ public class Proxy extends Thread {
              */
             if (headers != null) {
                 headers.removeHeaderNamed(HTTPConstants.HEADER_COOKIE);// Always remove cookies
-                // See https://issues.apache.org/bugzilla/show_bug.cgi?id=25430
+                // See https://bz.apache.org/bugzilla/show_bug.cgi?id=25430
                 // HEADER_AUTHORIZATION won't be removed, it will be used
                 // for creating Authorization Manager
                 // Remove additional headers
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java
index 5097d40..19199ed 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java
@@ -158,7 +158,7 @@ public class HTTPHC4Impl extends HTTPHCAbstractImpl {
      * Special interceptor made to keep metrics when connection is released for some method like HEAD
      * Otherwise calling directly ((HttpConnection) localContext.getAttribute(ExecutionContext.HTTP_CONNECTION)).getMetrics();
      * would throw org.apache.http.impl.conn.ConnectionShutdownException
-     * See https://issues.apache.org/jira/browse/HTTPCLIENT-1081
+     * See https://bz.apache.org/jira/browse/HTTPCLIENT-1081
      */
     private static final HttpResponseInterceptor METRICS_SAVER = new HttpResponseInterceptor(){
         @Override
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
index 81b78e9..ccb1a28 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
@@ -1586,7 +1586,7 @@ public abstract class HTTPSamplerBase extends AbstractSampler
                 // downloaded for the sample made for the redirected url
                 // otherwise, use null so the container is created if necessary unless
                 // the flag is false, in which case revert to broken 2.1 behaviour 
-                // Bug 51939 -  https://issues.apache.org/bugzilla/show_bug.cgi?id=51939
+                // Bug 51939 -  https://bz.apache.org/bugzilla/show_bug.cgi?id=51939
                 if(!wasRedirected) {
                     HTTPSampleResult container = (HTTPSampleResult) (
                             areFollowingRedirect ? res.getParent() : SEPARATE_CONTAINER ? null : res);
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerProxy.java b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerProxy.java
index 1af54b3..6893ee8 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerProxy.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerProxy.java
@@ -66,7 +66,7 @@ public final class HTTPSamplerProxy extends HTTPSamplerBase implements Interrupt
                 return errorResult(ex, new HTTPSampleResult());
             }
         }
-        // see https://issues.apache.org/bugzilla/show_bug.cgi?id=51380
+        // see https://bz.apache.org/bugzilla/show_bug.cgi?id=51380
         if(notifyFirstSampleAfterLoopRestart) {
             impl.notifyFirstSampleAfterLoopRestart();
             notifyFirstSampleAfterLoopRestart = false;
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/util/ConversionUtils.java b/src/protocol/http/org/apache/jmeter/protocol/http/util/ConversionUtils.java
index 573f4f5..7bd2c7a 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/util/ConversionUtils.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/util/ConversionUtils.java
@@ -99,7 +99,7 @@ public class ConversionUtils {
      * @param location the location, possibly with extraneous leading "../"
      * @return URL with extraneous ../ removed
      * @throws MalformedURLException when the given <code>URL</code> is malformed
-     * @see <a href="https://issues.apache.org/bugzilla/show_bug.cgi?id=46690">Bug 46690 - handling of 302 redirects with invalid relative paths</a>
+     * @see <a href="https://bz.apache.org/bugzilla/show_bug.cgi?id=46690">Bug 46690 - handling of 302 redirects with invalid relative paths</a>
      */
     public static URL makeRelativeURL(URL baseURL, String location) throws MalformedURLException{
         URL initial = new URL(baseURL,location);
@@ -164,7 +164,7 @@ public class ConversionUtils {
      * 
      * @param url in which the '/..'s should be removed
      * @return collapsed URL
-     * @see <a href="https://issues.apache.org/bugzilla/show_bug.cgi?id=49083">Bug 49083 - collapse /.. in redirect URLs</a>
+     * @see <a href="https://bz.apache.org/bugzilla/show_bug.cgi?id=49083">Bug 49083 - collapse /.. in redirect URLs</a>
      */
     public static String removeSlashDotDot(String url)
     {
diff --git a/test/src/org/apache/jmeter/engine/util/TestValueReplacer.java b/test/src/org/apache/jmeter/engine/util/TestValueReplacer.java
index 44061f5..fe39ec4 100644
--- a/test/src/org/apache/jmeter/engine/util/TestValueReplacer.java
+++ b/test/src/org/apache/jmeter/engine/util/TestValueReplacer.java
@@ -132,7 +132,7 @@ public class TestValueReplacer extends JMeterTestCase {
          * Here, the string contains a valid variable reference, so all
          * backslashes are also processed.
          * 
-         * See https://issues.apache.org/bugzilla/show_bug.cgi?id=53534
+         * See https://bz.apache.org/bugzilla/show_bug.cgi?id=53534
          */
         public void testReplaceFunctionWithBackslash() throws Exception {
             ValueReplacer replacer = new ValueReplacer();
diff --git a/test/src/org/apache/jmeter/protocol/http/modifier/TestAnchorModifier.java b/test/src/org/apache/jmeter/protocol/http/modifier/TestAnchorModifier.java
index 155eb9e..cc5d03a 100644
--- a/test/src/org/apache/jmeter/protocol/http/modifier/TestAnchorModifier.java
+++ b/test/src/org/apache/jmeter/protocol/http/modifier/TestAnchorModifier.java
@@ -59,12 +59,12 @@ public class TestAnchorModifier extends JMeterTestCase {
             result.setResponseData(new TextFile(System.getProperty("user.dir") + HTMLFileName).getText(), null);
             result.setSampleLabel(context.toString());
             result.setSamplerData(context.toString());
-            result.setURL(new URL("http://issues.apache.org/fakepage.html"));
+            result.setURL(new URL("http://bz.apache.org/fakepage.html"));
             jmctx.setPreviousResult(result);
             AnchorModifier modifier = new AnchorModifier();
             modifier.setThreadContext(jmctx);
             modifier.process();
-            assertEquals("http://issues.apache.org/bugzilla/buglist.cgi?"
+            assertEquals("http://bz.apache.org/bugzilla/buglist.cgi?"
                     + "bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED"
                     + "&email1=&emailtype1=substring&emailassigned_to1=1"
                     + "&email2=&emailtype2=substring&emailreporter2=1" + "&bugidtype=include&bug_id=&changedin=&votes="
@@ -74,7 +74,7 @@ public class TestAnchorModifier extends JMeterTestCase {
                     + "&keywords_type=anywords" + "&field0-0-0=noop&type0-0-0=noop&value0-0-0="
                     + "&cmdtype=doit&order=Reuse+same+sort+as+last+time", config.toString());
             config.recoverRunningVersion();
-            assertEquals("http://issues.apache.org/bugzilla/buglist.cgi?"
+            assertEquals("http://bz.apache.org/bugzilla/buglist.cgi?"
                     + "bug_status=.*&bug_status=.*&bug_status=.*&email1="
                     + "&emailtype1=substring&emailassigned_to1=1&email2=" + "&emailtype2=substring&emailreporter2=1"
                     + "&bugidtype=include&bug_id=&changedin=&votes=" + "&chfieldfrom=&chfieldto=Now&chfieldvalue="
diff --git a/xdocs/issues.xml b/xdocs/issues.xml
index 9fedb76..1e72e8a 100644
--- a/xdocs/issues.xml
+++ b/xdocs/issues.xml
@@ -32,14 +32,14 @@ It's also worth checking first on the JMeter user mailing list; others may alrea
 <section name="Known Bugs"> 
 <ul>
 <li>
-<a href="http://issues.apache.org/bugzilla/buglist.cgi?query_format=specific&amp;order=relevance+desc&amp;bug_status=__open__&amp;product=JMeter&amp;content=">
+<a href="http://bz.apache.org/bugzilla/buglist.cgi?query_format=specific&amp;order=relevance+desc&amp;bug_status=__open__&amp;product=JMeter&amp;content=">
 Bugzilla - All open bugs</a>
 </li>
 <li>
-<a href="http://issues.apache.org/bugzilla/buglist.cgi?bug_status=NEW&amp;bug_status=ASSIGNED&amp;bug_status=REOPENED&amp;bug_status=NEEDINFO&amp;bug_severity=Blocker&amp;bug_severity=Critical&amp;bug_severity=Major&amp;bug_severity=Normal&amp;bug_severity=Regression&amp;bug_severity=Minor&amp;bug_severity=Trivial&amp;product=JMeter&amp;cmdtype=doit&amp;order=Bug+Number">
+<a href="http://bz.apache.org/bugzilla/buglist.cgi?bug_status=NEW&amp;bug_status=ASSIGNED&amp;bug_status=REOPENED&amp;bug_status=NEEDINFO&amp;bug_severity=Blocker&amp;bug_severity=Critical&amp;bug_severity=Major&amp;bug_severity=Normal&amp;bug_severity=Regression&amp;bug_severity=Minor&amp;bug_severity=Trivial&amp;product=JMeter&amp;cmdtype=doit&amp;order=Bug+Number">
 Bugzilla - Open bugs (excluding enhancements)</a></li>
 <li>
-<a href="http://issues.apache.org/bugzilla/buglist.cgi?bug_status=NEW&amp;bug_status=ASSIGNED&amp;bug_status=REOPENED&amp;bug_status=NEEDINFO&amp;bug_severity=Enhancement&amp;product=JMeter&amp;cmdtype=doit&amp;order=Bug+Number">
+<a href="http://bz.apache.org/bugzilla/buglist.cgi?bug_status=NEW&amp;bug_status=ASSIGNED&amp;bug_status=REOPENED&amp;bug_status=NEEDINFO&amp;bug_severity=Enhancement&amp;product=JMeter&amp;cmdtype=doit&amp;order=Bug+Number">
 Bugzilla - Enhancements only</a>
 </li>
 </ul>
@@ -51,7 +51,7 @@ Bugzilla is good for tracking progress and supplying patches, but is unwieldy fo
 </p> 
 <p>
 If you have not already done so, you need to register an account first, using the "New Account" link at the top of the
-main Bugzilla page: <a href="https://issues.apache.org/bugzilla/">https://issues.apache.org/bugzilla/</a>.
+main Bugzilla page: <a href="https://bz.apache.org/bugzilla/">https://bz.apache.org/bugzilla/</a>.
 </p>
 <p>
 Make sure you read and understand the information on the account creation page before signing up.
@@ -74,7 +74,7 @@ If reporting a bug, are you sure it really is a bug in JMeter, not just a misund
 </p>
 <p>
 If you have not already done so, you need to register an account first, using the "New Account" link at the top of the
-main Bugzilla page: <a href="https://issues.apache.org/bugzilla/">https://issues.apache.org/bugzilla/</a>.
+main Bugzilla page: <a href="https://bz.apache.org/bugzilla/">https://bz.apache.org/bugzilla/</a>.
 </p>
 <p>
 Make sure you read and understand the information on the account creation page before signing up.
@@ -90,7 +90,7 @@ Use the checklist below to guide you.
 </p>
 <ul>
 <li>JMeter version</li>
-<li>Java version (output from java -version)</li>
+li>Java version (output from java -version)</li>
 <li>OS version</li>
 <li>jmeter.log file (unlikely to contain sensitive information, but check before uploading)</li>
 <li>JMX file if relevant (redact any sensitive information first), providing a simplified Test Plan (using DEBUG sampler) will ensure BUG is fixed much more rapidely than without it</li>
@@ -108,8 +108,8 @@ Please do not use @author tags (credit will be given in the changes file).
 <li>In the case of patches for new features, please also provide documentation patches if at all possible. 
 Components are documented in xdocs/usermanual/component_reference.xml.</li>
 </ul>
-<p><b>See also the following <a href="https://issues.apache.org/bugwritinghelp.html">Bug writing guidelines</a>,
-also the terms and conditions noted on the <a href="https://issues.apache.org/bugzilla/createaccount.cgi">Bugzilla account creation page.</a></b></p>
+<p><b>See also the following <a href="https://bz.apache.org/bugwritinghelp.html">Bug writing guidelines</a>,
+also the terms and conditions noted on the <a href="https://bz.apache.org/bugzilla/createaccount.cgi">Bugzilla account creation page.</a></b></p>
 </section>
 </body>
 </document>
diff --git a/xdocs/stylesheets/site.vsl b/xdocs/stylesheets/site.vsl
index e91d687..59ee54f 100644
--- a/xdocs/stylesheets/site.vsl
+++ b/xdocs/stylesheets/site.vsl
@@ -233,7 +233,7 @@ No
 #end
 ##
 #macro ( bugzilla $id)
-<a href="https://issues.apache.org/bugzilla/show_bug.cgi?id=$id.getText()">Bug $id.getText()</a>
+<a href="https://bz.apache.org/bugzilla/show_bug.cgi?id=$id.getText()">Bug $id.getText()</a>
 #end
 
 ## Shorthand - automatically adds " - " before remaining text
diff --git a/xdocs/stylesheets/site_printable.vsl b/xdocs/stylesheets/site_printable.vsl
index b817284..378b83b 100644
--- a/xdocs/stylesheets/site_printable.vsl
+++ b/xdocs/stylesheets/site_printable.vsl
@@ -226,7 +226,7 @@ No
 #end
 ##
 #macro ( bugzilla $id)
-<a href="https://issues.apache.org/bugzilla/show_bug.cgi?id=$id.getText()">Bug $id.getText()</a>
+<a href="https://bz.apache.org/bugzilla/show_bug.cgi?id=$id.getText()">Bug $id.getText()</a>
 #end
 
 ## Shorthand - automatically adds " - " before remaining text
@@ -593,4 +593,4 @@ trademarks of the Apache Software Foundation.
 </table>
 </body>
 </html>
-#end
\ No newline at end of file
+#end
