Index: src/core/org/apache/jmeter/control/TransactionController.java
===================================================================
--- src/core/org/apache/jmeter/control/TransactionController.java	(revision 819480)
+++ src/core/org/apache/jmeter/control/TransactionController.java	(working copy)
@@ -202,6 +202,7 @@
                 }
                 res.setAllThreads(sampleResult.getAllThreads());
                 res.setGroupThreads(sampleResult.getGroupThreads());
+                res.setLatency(res.getLatency() + sampleResult.getLatency());
             }
         }
     }
