diff -crB jakarta-jmeter-2.3.4/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java jakarta-jmeter-2.3.4_WIP/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
*** jakarta-jmeter-2.3.4/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	2009-06-17 09:29:54.000000000 -0600
--- jakarta-jmeter-2.3.4_WIP/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	2010-05-14 16:11:23.428387418 -0600
***************
*** 1394,1401 ****
              if (frameDepth > MAX_FRAME_DEPTH) {
                  res.addSubResult(errorResult(new Exception("Maximum frame/iframe nesting depth exceeded."), res));
              } else {
-                 // If we followed redirects, we already have a container:
-                 if(!areFollowingRedirect) {
                  HTTPSampleResult container = (HTTPSampleResult) (areFollowingRedirect ? res.getParent() : res);
  
                  // Only download page resources if we were not redirected.
--- 1394,1399 ----
***************
*** 1404,1410 ****
                  if(!wasRedirected) {
                      res = downloadPageResources(res, container, frameDepth);
                  }
-                 }
              }
          }
          return res;
--- 1402,1407 ----
Only in jakarta-jmeter-2.3.4_WIP/src/protocol/http/org/apache/jmeter/protocol/http/sampler: HTTPSamplerBase.java~
