Index: src/functions/org/apache/jmeter/functions/FileToString.java
===================================================================
--- src/functions/org/apache/jmeter/functions/FileToString.java	(revision 1057358)
+++ src/functions/org/apache/jmeter/functions/FileToString.java	(working copy)
@@ -86,6 +86,9 @@
         String encoding = null;//means platform default
         if (values.length >= ENCODING) {
             encoding = ((CompoundVariable) values[ENCODING - 1]).execute().trim();
+            if (encoding.length() <= 0) { // empty encoding, return to platorm default
+                encoding = null;
+            }
         }
 
         String myName = "";//$NON-NLS-1$

