Index: src/reports/org/apache/jmeter/report/gui/tree/ReportTreeListener.java
===================================================================
--- src/reports/org/apache/jmeter/report/gui/tree/ReportTreeListener.java	(revision 1489981)
+++ src/reports/org/apache/jmeter/report/gui/tree/ReportTreeListener.java	(working copy)
@@ -208,6 +208,7 @@
         for (int i = 0; i < path.length; i++) {
             if (contains(source, path[i])) {
                 isValid = false;
+                break;
             }
         }
         return isValid;
