Index: src/core/org/apache/jmeter/testbeans/gui/WrapperEditor.java
===================================================================
--- src/core/org/apache/jmeter/testbeans/gui/WrapperEditor.java	(revision 1496951)
+++ src/core/org/apache/jmeter/testbeans/gui/WrapperEditor.java	(working copy)
@@ -445,4 +445,8 @@
             guiEditor.setAsText(lastValidValue);
         }
     }
+    
+    public void addChangeListener(PropertyChangeListener listener){
+    	guiEditor.addPropertyChangeListener(listener);
+    }
 }
\ No newline at end of file
