Index: trunk/src/protocol/http/org/apache/jmeter/protocol/http/sampler/AjpSampler.java
===================================================================
--- trunk/src/protocol/http/org/apache/jmeter/protocol/http/sampler/AjpSampler.java	(revision 900667)
+++ trunk/src/protocol/http/org/apache/jmeter/protocol/http/sampler/AjpSampler.java	(working copy)
@@ -91,6 +91,7 @@
     private transient ByteArrayOutputStream responseData = new ByteArrayOutputStream();
     private int inpos = 0;
     private int outpos = 0;
+    private transient String stringBody = null;
     private transient InputStream body = null;
 
     public AjpSampler() {
@@ -131,6 +132,7 @@
         }
         channel = null;
         body = null;
+        stringBody = null;
     }
 
     private void setupConnection(URL url, 
@@ -171,7 +173,7 @@
         } else {
             setString(HTTP_1_1);
         }
-        setString(url.getFile());
+        setString(url.getPath());
         setString(localAddress);
         setString(localName);
         setString(host);
@@ -181,7 +183,12 @@
         String hdr = setConnectionHeaders(url, host, method);
         res.setRequestHeaders(hdr);
         res.setCookies(setConnectionCookies(url, getCookieManager()));
-        setByte((byte)0xff); // Attributes not supported
+        String query = url.getQuery();
+        if (query != null) {
+            setByte((byte)0x05); // Marker for query string attribute
+            setString(query);
+        }
+        setByte((byte)0xff); // More general attributes not supported
     }
 
     private int getHeaderSize(String method, URL url) {
@@ -267,9 +274,10 @@
                     } else {
                         sb.append('&');
                     }
-                    sb.append(arg.getName()).append('=').append(arg.getStringValue());
+                    sb.append(arg.getStringValue());
                 }
-                byte [] sbody = sb.toString().getBytes(); //FIXME - encoding
+                stringBody = sb.toString();
+                byte [] sbody = stringBody.getBytes(); //FIXME - encoding
                 cl = sbody.length;
                 body = new ByteArrayInputStream(sbody);
             }
@@ -364,6 +372,7 @@
     throws IOException {
         send();
         if(method.equals(POST)) {
+            res.setQueryString(stringBody);
             sendPostBody();
         }
         handshake(res);
