Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/LoopController.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/LoopController.java	(revision 536755)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/LoopController.java	(working copy)
@@ -78,17 +78,19 @@
 		return getPropertyAsBoolean(CONTINUE_FOREVER);
 	}
 
-	/*
-	 * (non-Javadoc)
-	 * 
-	 * @see org.apache.jmeter.control.Controller#isDone()
-	 */
-	public boolean isDone() {
-		if (getLoops() != 0) {
-			return super.isDone();
-		}
-		return true;
-	}
+    /*
+     * (non-Javadoc)
+     * 
+     * @see org.apache.jmeter.control.Controller#next()
+     */
+    public Sampler next() {
+        if(endOfLoop()) {
+            return null;
+        }
+        else {
+            return super.next();
+        }
+    }
 
 	private boolean endOfLoop() {
 		return (getLoops() > -1) && loopCount >= getLoops();
