Index: C:/Documents and Settings/alf/workspace/Jmeter/src/core/org/apache/jmeter/gui/util/TextAreaCellRenderer.java
===================================================================
--- C:/Documents and Settings/alf/workspace/Jmeter/src/core/org/apache/jmeter/gui/util/TextAreaCellRenderer.java	(revision 519099)
+++ C:/Documents and Settings/alf/workspace/Jmeter/src/core/org/apache/jmeter/gui/util/TextAreaCellRenderer.java	(working copy)
@@ -35,6 +35,9 @@
 	public Component getTableCellRendererComponent(JTable arg0, Object arg1, boolean arg2, boolean arg3, int arg4,
 			int arg5) {
 		rend = new JTextArea(arg1.toString());
+		// Use two rows, so that we have room for horisontal scrollbar, if the text is one long line. Fix for 40371
+		// This is not an optimal solution, but makes it possible to see the line if it is long
+		rend.setRows(2);
 		rend.revalidate();
 		if (!arg3 && !arg2) {
 			rend.setBackground(JMeterColor.LAVENDER);
