From 1565afb533baa4345fcfa650251cd53637b3c6b9 Mon Sep 17 00:00:00 2001
From: Felix Schumacher <felix.schumacher@internetallee.de>
Date: Tue, 10 Mar 2020 21:27:10 +0100
Subject: [PATCH] Keep divider location when changing visualizer mode

---
 .../apache/jmeter/visualizers/ViewResultsFullVisualizer.java    | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/components/src/main/java/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java b/src/components/src/main/java/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java
index 5e5e1549e0..917ab3355a 100644
--- a/src/components/src/main/java/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java
+++ b/src/components/src/main/java/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java
@@ -515,11 +515,13 @@ implements ActionListener, TreeSelectionListener, Clearable, ItemListener {
                     // to restore last selected tab (better user-friendly)
                     selectedTab = rightSide.getSelectedIndex();
                     // Remove old right side
+                    int dividerLocation = mainSplit.getDividerLocation();
                     mainSplit.remove(rightSide);
 
                     // create and add a new right side
                     rightSide = new JTabbedPane();
                     mainSplit.add(rightSide);
+                    mainSplit.setDividerLocation(dividerLocation);
                     resultsRender.setRightSide(rightSide);
                     resultsRender.setLastSelectedTab(selectedTab);
                     log.debug("selectedTab={}", selectedTab);
-- 
2.17.1

