Index: src/core/org/apache/jmeter/samplers/SampleResult.java
===================================================================
--- src/core/org/apache/jmeter/samplers/SampleResult.java	(revision 497868)
+++ src/core/org/apache/jmeter/samplers/SampleResult.java	(working copy)
@@ -724,6 +724,10 @@
 		setEndTime(end);
 	}
 
+	public void incIdleTime(long pause) {
+		idleTime += pause;
+	}
+	
 	/**
 	 * Record the start time of a sample
 	 * 
