Index: src/core/org/apache/jmeter/visualizers/SamplingStatCalculator.java
===================================================================
--- src/core/org/apache/jmeter/visualizers/SamplingStatCalculator.java	(revision 1043833)
+++ src/core/org/apache/jmeter/visualizers/SamplingStatCalculator.java	(working copy)
@@ -131,7 +131,7 @@
         if (count == 0) {
             return 0;
         }
-        return calculator.getTotalBytes() / count;
+        return calculator.getTotalBytes() / (double)count;
     }
 
     public String getLabel() {
 
