Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/IfController.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/IfController.java	(revision 536755)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/control/IfController.java	(working copy)
@@ -137,16 +138,18 @@
 	}
 
 	/**
-	 * @see org.apache.jmeter.control.Controller#next() 'JMeterThread' iterates
-	 *      thru the Controller by calling this method. IF a valid 'Sampler' is
-	 *      returned, then it executes the sampler (calls sampler.sampler(xxx)
-	 *      method) . So here we make sure that the samplers belonging to this
-	 *      Controller do not get called - if isDone is true - if its the first
-	 *      time this is run. The first time is special cause it is called prior
-	 *      the iteration even starts !
+	 * @see org.apache.jmeter.control.Controller#next()
 	 */
 	public Sampler next() {
-		boolean result = evaluateCondition(getCondition());
+        // We should only evalute the condition if it is the first
+        // time ( first "iteration" ) we are called.
+        // For subsequent calls, we are inside the IfControllerGroup,
+        // so then we just pass the control to the next item inside the if control
+        boolean result = true;
+        if(isFirst()) {
+            result = evaluateCondition(getCondition());
+        }
+        
 		if (result) {
 			return super.next();
 		}
@@ -156,4 +159,4 @@
 			return null;
 		}
 	}
-}
\ No newline at end of file
+}
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java	(revision 536755)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java	(working copy)
@@ -104,7 +104,10 @@
 	public void modifyTestElement(TestElement tg) {
 		super.configureTestElement(tg);
 		if (tg instanceof ThreadGroup) {
-			((ThreadGroup) tg).setSamplerController((LoopController) loopPanel.createTestElement());
+            // Create and set the name of the LoopController which controls this ThreadGroup
+            LoopController loopController = (LoopController)loopPanel.createTestElement();
+            loopController.setName("Loop for ThreadGroup " + getName());
+			((ThreadGroup) tg).setSamplerController(loopController);
 		}
 
 		tg.setProperty(ThreadGroup.NUM_THREADS, threadInput.getText());
