Index: RegexExtractor.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/components/org/apache/jmeter/extractor/RegexExtractor.java,v
retrieving revision 1.11
diff -u -r1.11 RegexExtractor.java
--- RegexExtractor.java	15 Oct 2003 17:53:31 -0000	1.11
+++ RegexExtractor.java	21 Jan 2004 18:45:14 -0000
@@ -29,7 +29,7 @@
 import org.apache.oro.text.regex.Util;
 
 /**
- * @version $Revision: 1.11 $
+ * @version $Revision: 1.12 $
  */
 public class RegexExtractor
     extends AbstractTestElement
@@ -69,6 +69,7 @@
         }
         log.debug("RegexExtractor processing result");
         context.getVariables().put(getRefName(), getDefaultValue());
+        context.getVariables().put(getRefName()+"_matchNr", "0");
         Perl5Matcher matcher = (Perl5Matcher) localMatcher.get();
         PatternMatcherInput input =
             new PatternMatcherInput(
@@ -94,14 +95,29 @@
 			    x++;
 			}
 			while (x != getMatchNumber() && !done);
+			
 			try
 			{
-			    MatchResult match = getCorrectMatch(matches);
-			    if (match != null)
-			    {
-			        context.getVariables().put(getRefName(), generateResult(match));
-			        saveGroups(context.getVariables(), getRefName(), match);
-			    }
+				MatchResult match;
+				if (getMatchNumber() <0) {
+					context.getVariables().put(getRefName()+"_matchNr", ""+matches.size());
+					for (int i=1;i<=matches.size();i++) {
+					    match = getCorrectMatch(matches, i);
+					    if (match != null)
+					    {
+					        context.getVariables().put(getRefName()+"_"+i, generateResult(match));
+					        saveGroups(context.getVariables(), getRefName()+"_"+i, match);
+				    	}
+					}
+			    	
+				} else {
+				    match = getCorrectMatch(matches, getMatchNumber());
+				    if (match != null)
+				    {
+				        context.getVariables().put(getRefName(), generateResult(match));
+			    	    saveGroups(context.getVariables(), getRefName(), match);
+			    	}
+				}
 			}
 			catch (RuntimeException e)
 			{
@@ -233,16 +249,16 @@
      * @param matches
      * @return MatchResult
      */
-    protected MatchResult getCorrectMatch(List matches)
+    
+    protected MatchResult getCorrectMatch(List matches, int MatchNumber)
     {
-        if (getMatchNumber() == matches.size() && matches.size() > 0)
-        {
-            return (MatchResult) matches.get(matches.size() - 1);
-        }
-        else if (getMatchNumber() == 0 && matches.size() > 0)
+        if (MatchNumber == 0 && matches.size() > 0)
         {
             return (MatchResult) matches.get(
                 JMeterUtils.getRandomInt(matches.size()));
+        } else if (matches.size() > 0)
+        {
+            return (MatchResult) matches.get(MatchNumber - 1);
         }
         else
         {
