Index: C:/Documents and Settings/alf/workspace/Jmeter 2.2 writable/src/core/org/apache/jmeter/gui/action/ActionRouter.java
===================================================================
--- C:/Documents and Settings/alf/workspace/Jmeter 2.2 writable/src/core/org/apache/jmeter/gui/action/ActionRouter.java	(revision 570861)
+++ C:/Documents and Settings/alf/workspace/Jmeter 2.2 writable/src/core/org/apache/jmeter/gui/action/ActionRouter.java	(working copy)
@@ -67,7 +67,7 @@
 
 	private void performAction(final ActionEvent e) {
 		try {
-			GuiPackage.getInstance().updateCurrentNode();
+			GuiPackage.getInstance().updateCurrentGui();
 			Set commandObjects = (Set) commands.get(e.getActionCommand());
 			Iterator iter = commandObjects.iterator();
 			while (iter.hasNext()) {
