diff --git a/src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java b/src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java
index 2f6360ab2c..0e350a29eb 100644
--- a/src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java
+++ b/src/core/src/main/java/org/apache/jmeter/util/JMeterUtils.java
@@ -875,13 +875,13 @@ public class JMeterUtils implements UnitTestManager {
             errorMsg = "Unknown error - see log file";
             log.warn("Unknown error", new Throwable("errorMsg == null"));
         }
+        if (exception != null) {
+            log.error(errorMsg, exception);
+        } else {
+            log.error(errorMsg);
+        }
         GuiPackage instance = GuiPackage.getInstance();
         if (instance == null) {
-            if(exception != null) {
-                log.error(errorMsg, exception);
-            } else {
-                log.error(errorMsg);
-            }
             System.out.println(errorMsg); // NOSONAR intentional
             return; // Done
         }
