Index: src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java	(revision 1178423)
+++ src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java	(working copy)
@@ -24,9 +24,9 @@
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.net.URLConnection;
-import java.util.HashMap;
-import java.util.Map;
 import java.util.Date;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.commons.httpclient.HttpMethod;
 import org.apache.commons.httpclient.URIException;
@@ -331,7 +331,7 @@
         threadCache = new InheritableThreadLocal<Map<String, CacheEntry>>(){
             @Override
             protected Map<String, CacheEntry> initialValue(){
-                return new HashMap<String, CacheEntry>();
+                return new ConcurrentHashMap<String, CacheEntry>();
             }
         };
     }
