Index: src/jorphan/org/apache/jorphan/collections/Data.java
===================================================================
--- src/jorphan/org/apache/jorphan/collections/Data.java	(revision 1489451)
+++ src/jorphan/org/apache/jorphan/collections/Data.java	(working copy)
@@ -95,6 +95,7 @@
             for (int count = 0; count < dHeaders.length; count++) {
                 if (!header.contains(dHeaders[count])) {
                     valid = false;
+                    break;
                 }
             }
         }
