diff --git a/src/core/src/main/java/org/apache/jmeter/samplers/SampleResult.java b/src/core/src/main/java/org/apache/jmeter/samplers/SampleResult.java
index 11902de0c9..a8eb891583 100644
--- a/src/core/src/main/java/org/apache/jmeter/samplers/SampleResult.java
+++ b/src/core/src/main/java/org/apache/jmeter/samplers/SampleResult.java
@@ -706,11 +706,11 @@ public class SampleResult implements Serializable, Cloneable, Searchable {
      *            the {@link SampleResult} to be added
      * @param renameSubResults boolean do we rename subResults based on position
      */
-    private void storeSubResult(SampleResult subResult, boolean renameSubResults) {
+    public void storeSubResult(SampleResult subResult, boolean renameSubResults) {
         if (subResults == null) {
             subResults = new ArrayList<>();
         }
-        if(renameSubResults) {
+        if (renameSubResults) {
             subResult.setSampleLabel(getSampleLabel()+"-"+subResultIndex++);
         }
         subResults.add(subResult);
diff --git a/src/core/src/main/java/org/apache/jmeter/save/converters/SampleResultConverter.java b/src/core/src/main/java/org/apache/jmeter/save/converters/SampleResultConverter.java
index 49cbb093b5..857bb3f959 100644
--- a/src/core/src/main/java/org/apache/jmeter/save/converters/SampleResultConverter.java
+++ b/src/core/src/main/java/org/apache/jmeter/save/converters/SampleResultConverter.java
@@ -396,7 +396,7 @@ public class SampleResultConverter extends AbstractCollectionConverter {
         if (subItem instanceof AssertionResult) {
             res.addAssertionResult((AssertionResult) subItem);
         } else if (subItem instanceof SampleResult) {
-            res.storeSubResult((SampleResult) subItem);
+            res.storeSubResult((SampleResult) subItem, false);
         } else if (nodeName.equals(TAG_RESPONSE_HEADER)) {
             res.setResponseHeaders((String) subItem);
         } else if (nodeName.equals(TAG_REQUEST_HEADER)) {
