Index: src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BeanShellSamplerGui.java
===================================================================
--- src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BeanShellSamplerGui.java	(revision 1496951)
+++ src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BeanShellSamplerGui.java	(working copy)
@@ -33,6 +33,9 @@
 import org.apache.jmeter.testelement.TestElement;
 import org.apache.jmeter.testelement.property.BooleanProperty;
 import org.apache.jmeter.util.JMeterUtils;
+import org.fife.ui.rsyntaxtextarea.RSyntaxTextArea;
+import org.fife.ui.rsyntaxtextarea.SyntaxConstants;
+import org.fife.ui.rtextarea.RTextScrollPane;
 
 public class BeanShellSamplerGui extends AbstractSamplerGui {
 
@@ -44,7 +47,7 @@
 
     private JTextField parameters;// parameters to pass to script file (or script)
 
-    private JTextArea scriptField;// script area
+    private RSyntaxTextArea scriptField;// script area
 
     public BeanShellSamplerGui() {
         init();
@@ -153,8 +156,9 @@
     }
 
     private JPanel createScriptPanel() {
-        scriptField = new JTextArea();
-        scriptField.setRows(4);
+        scriptField = new RSyntaxTextArea(20,20);
+        scriptField.setSyntaxEditingStyle(SyntaxConstants.SYNTAX_STYLE_JAVA);
+        scriptField.setCodeFoldingEnabled(true);
         scriptField.setLineWrap(true);
         scriptField.setWrapStyleWord(true);
 
@@ -163,7 +167,7 @@
 
         JPanel panel = new JPanel(new BorderLayout());
         panel.add(label, BorderLayout.NORTH);
-        panel.add(new JScrollPane(scriptField), BorderLayout.CENTER);
+        panel.add(new RTextScrollPane(scriptField), BorderLayout.CENTER);
 
         JTextArea explain = new JTextArea(JMeterUtils.getResString("bsh_script_variables")); //$NON-NLS-1$
         explain.setLineWrap(true);
