Index: src/components/org/apache/jmeter/assertions/gui/XPathPanel.java
===================================================================
--- src/components/org/apache/jmeter/assertions/gui/XPathPanel.java	(revision 1463517)
+++ src/components/org/apache/jmeter/assertions/gui/XPathPanel.java	(working copy)
@@ -20,13 +20,12 @@
 
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
+import java.awt.BorderLayout;
 
-import javax.swing.Box;
 import javax.swing.JButton;
 import javax.swing.JCheckBox;
 import javax.swing.JOptionPane;
 import javax.swing.JPanel;
-import javax.swing.JTextField;
 import javax.xml.parsers.ParserConfigurationException;
 import javax.xml.transform.TransformerException;
 
@@ -37,6 +36,8 @@
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
 
+import org.apache.jorphan.gui.JLabeledTextArea;
+
 public class XPathPanel extends JPanel {
     private static final long serialVersionUID = 240L;
 
@@ -44,7 +45,7 @@
 
     private JCheckBox negated;
 
-    private JTextField xpath;
+    private JLabeledTextArea xpath;
 
     private JButton checkXPath;
 
@@ -52,24 +53,15 @@
      * 
      */
     public XPathPanel() {
-        super();
+        super(new BorderLayout(5, 0));
         init();
     }
 
     private void init() {
-        Box hbox = Box.createHorizontalBox();
-        hbox.add(Box.createHorizontalGlue());
-        hbox.add(getXPathTextField());
-        hbox.add(Box.createHorizontalGlue());
-        hbox.add(getCheckXPathButton());
-
-        Box vbox = Box.createVerticalBox();
-        vbox.add(hbox);
-        vbox.add(Box.createVerticalGlue());
-        vbox.add(getNegatedCheckBox());
-
-        add(vbox);
-
+        add(getXPathTextField(),BorderLayout.CENTER);
+        add(getCheckXPathButton(),BorderLayout.EAST);
+        add(getNegatedCheckBox(), BorderLayout.SOUTH);
+		
         setDefaultValues();
     }
 
@@ -145,9 +137,9 @@
         return checkXPath;
     }
 
-    public JTextField getXPathTextField() {
+    public JLabeledTextArea getXPathTextField() {
         if (xpath == null) {
-            xpath = new JTextField(50);
+            xpath = new JLabeledTextArea();
         }
         return xpath;
     }
