Index: src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LDAPExtSampler.java
===================================================================
--- src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LDAPExtSampler.java	(Revision 1797503)
+++ src/protocol/ldap/org/apache/jmeter/protocol/ldap/sampler/LDAPExtSampler.java	(Arbeitskopie)
@@ -800,6 +800,8 @@
                 xmlBuffer.tag("scope" , scopeStr); // $NON-NLS-1$
                 xmlBuffer.tag("countlimit",countLimit); // $NON-NLS-1$
                 xmlBuffer.tag("timelimit",timeLimit); // $NON-NLS-1$
+                xmlBuffer.tag("returnobject", Boolean.valueOf(isRetobj()).toString()); // $NON-NLS-1$
+                xmlBuffer.tag("dereference", Boolean.valueOf(isDeref()).toString()); // $NON-NLS-1$
 
                 NamingEnumeration<SearchResult> srch=null;
                 try {
