Index: src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java
===================================================================
--- src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java	(revision 1736376)
+++ src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java	(working copy)
@@ -740,6 +740,12 @@
     public JComponent getPrintableComponent() {
         if (saveGraphToFile) {
             saveGraphToFile = false;
+            
+            // (re)draw the graph first to take settings into account (Bug 58329)
+            if (model.getRowCount() > 1) {
+                makeGraph();
+            }
+            
             graphPanel.setBounds(graphPanel.getLocation().x,graphPanel.getLocation().y,
                     graphPanel.width,graphPanel.height);
             return graphPanel;
