Index: JavaTest.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/protocol/java/org/apache/jmeter/protocol/java/test/JavaTest.java,v
retrieving revision 1.1
diff -u -r1.1 JavaTest.java
--- JavaTest.java	6 Jun 2003 17:19:36 -0000	1.1
+++ JavaTest.java	6 Jun 2003 18:19:31 -0000
@@ -98,7 +98,7 @@
      * The default value of the SleepTime parameter, in milliseconds.
      */
     public static final long DEFAULT_SLEEP_TIME = 100;
-	private static final String SLEEP_NAME="SleepTime";
+	private static final String SLEEP_NAME="Sleep_Time";
 
     /**
      * The default value of the SleepMask parameter.
@@ -106,7 +106,7 @@
     public static final long DEFAULT_SLEEP_MASK = 0xff;
 	private static final String DEFAULT_MASK_STRING 
 		= "0x" + (Long.toHexString(DEFAULT_SLEEP_MASK)).toUpperCase();
-	private static final String MASK_NAME="SleepMask";
+	private static final String MASK_NAME="Sleep_Mask";
 
     /**
      * The base number of milliseconds to sleep during each sample.
@@ -192,8 +192,8 @@
         getLogger().debug(whoAmI() + "\tsetupTest()");
         listParameters(context);
 
-        sleepTime = context.getLongParameter("SleepTime", DEFAULT_SLEEP_TIME);
-        sleepMask = context.getLongParameter("SleepMask", DEFAULT_SLEEP_MASK);
+        sleepTime = context.getLongParameter(SLEEP_NAME, DEFAULT_SLEEP_TIME);
+        sleepMask = context.getLongParameter(MASK_NAME, DEFAULT_SLEEP_MASK);
 		responseMessage =context.getParameter(RESPONSEMESSAGE_NAME,RESPONSEMESSAGE_DEFAULT);
 		responseCode    =context.getParameter(RESPONSECODE_NAME,RESPONSEMESSAGE_DEFAULT);
 		success         =context.getParameter(SUCCESS_NAME,SUCCESS_DEFAULT).equalsIgnoreCase("OK");
