Index: HtmlPane.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/swing/HtmlPane.java,v
retrieving revision 1.1
diff -u -r1.1 HtmlPane.java
--- HtmlPane.java	7 Sep 2002 16:54:34 -0000	1.1
+++ HtmlPane.java	8 Jun 2003 19:57:03 -0000
@@ -1,5 +1,7 @@
 package org.apache.jmeter.swing;
 import javax.swing.JTextPane;
+import javax.swing.event.HyperlinkEvent;
+import javax.swing.event.HyperlinkListener;
 /**
  * @author Administrator
  *
@@ -8,6 +10,21 @@
  */
 public class HtmlPane extends JTextPane
 {
+	public HtmlPane(){	
+		this.addHyperlinkListener(
+			new HyperlinkListener(){
+				public void hyperlinkUpdate(HyperlinkEvent e) {
+					if (e.getEventType()==HyperlinkEvent.EventType.ACTIVATED){
+						String ref=e.getURL().getRef();
+						if (ref != null && ref.length() >0) {
+							scrollToReference(ref);
+						}
+					}
+				}
+			}
+		);
+	}
+
 	
 	public void scrollToReference(String reference)
 	{
