Index: src/core/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java
===================================================================
--- src/core/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java	(revision 1741506)
+++ src/core/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java	(working copy)
@@ -196,7 +196,15 @@
     @Override
     public void rexit() throws RemoteException {
         log.info("Exiting");
-        backingEngine.exit();
+        // Bug 59400 - allow rexit() to return
+        Thread et = new Thread() {
+            @Override
+            public void run() {
+                log.info("Stopping the backing engine");
+                backingEngine.exit();
+            }  
+        };
+        et.setDaemon(false);
         // Tidy up any objects we created
         Registry reg = LocateRegistry.getRegistry(this.rmiPort);        
         try {
@@ -207,6 +215,7 @@
         log.info("Unbound from registry");
         // Help with garbage control
         JMeterUtils.helpGC();
+        et.start();
     }
 
     @Override
