Index: src/protocol/http/org/apache/jmeter/protocol/http/sampler/SoapSampler.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/sampler/SoapSampler.java	(revision 895607)
+++ src/protocol/http/org/apache/jmeter/protocol/http/sampler/SoapSampler.java	(working copy)
@@ -58,6 +58,8 @@
 
 	private static final String SOAPACTION = "SOAPAction"; //$NON-NLS-1$
 
+	private static final String RESPONSE_CONTENT_ENCODING = "Content-Encoding"; //$NON-NLS-1$
+
 	public void setXmlData(String data) {
 		setProperty(XML_DATA, data);
 	}
@@ -241,7 +243,7 @@
 
             if (instream != null) {// will be null for HEAD
 
-                org.apache.commons.httpclient.Header responseHeader = httpMethod.getResponseHeader(TRANSFER_ENCODING);
+                org.apache.commons.httpclient.Header responseHeader = httpMethod.getResponseHeader(RESPONSE_CONTENT_ENCODING);
                 if (responseHeader != null && ENCODING_GZIP.equals(responseHeader.getValue())) {
                     instream = new GZIPInputStream(instream);
                 }
