Index: src/core/org/apache/jmeter/util/ScriptingBeanInfoSupport.java
===================================================================
--- src/core/org/apache/jmeter/util/ScriptingBeanInfoSupport.java	(revision 1754657)
+++ src/core/org/apache/jmeter/util/ScriptingBeanInfoSupport.java	(working copy)
@@ -29,6 +29,7 @@
 import javax.swing.JCheckBox;
 
 import org.apache.commons.lang3.StringUtils;
+import org.apache.jmeter.gui.ClearGui;
 import org.apache.jmeter.testbeans.BeanInfoSupport;
 import org.apache.jmeter.testbeans.TestBean;
 import org.apache.jmeter.testbeans.gui.FileEditor;
@@ -116,7 +117,7 @@
         
     }
     
-    public static class JSR223ScriptCacheCheckboxEditor extends PropertyEditorSupport implements ActionListener {
+    public static class JSR223ScriptCacheCheckboxEditor extends PropertyEditorSupport implements ActionListener, ClearGui {
 
         private final JCheckBox checkbox;
 
@@ -198,5 +199,11 @@
         public boolean supportsCustomEditor() {
             return true;
         }
+
+        @Override
+        public void clearGui() {
+            initialValue = null;
+            checkbox.setSelected(false);
+        }
     }
 }
