--- RemoteJMeterEngineImpl.java.orig	2009-06-17 11:29:16.000000000 -0400
+++ RemoteJMeterEngineImpl.java	2009-10-11 20:33:46.000000000 -0400
@@ -70,7 +70,11 @@
         log.info("Starting backing engine on " + port);
         InetAddress localHost=null;
         try {
+          String host = System.getProperties().getProperty("java.rmi.server.hostname");
+          if( host==null )
             localHost = InetAddress.getLocalHost();
+          else
+            localHost = InetAddress.getByName(host);
         } catch (UnknownHostException e1) {
             throw new RemoteException("Cannot start. Unable to get local host IP address.");
         }
