Index: src/protocol/java/org/apache/jmeter/protocol/java/test/SleepTest.java
===================================================================
--- src/protocol/java/org/apache/jmeter/protocol/java/test/SleepTest.java	(revision 1001614)
+++ src/protocol/java/org/apache/jmeter/protocol/java/test/SleepTest.java	(working copy)
@@ -134,7 +134,8 @@
                 sleep = sleepTime + (start % sleepMask);
             }
 
-            results.setSampleLabel("Sleep Test: time = " + sleep);
+            results.setSampleLabel(context.getParameter("TestElement.name"));
+            //results.setSampleLabel("Sleep Test: time = " + sleep);
 
             // Execute the sample. In this case sleep for the
             // specified time.
