Index: src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java	(revision 1748496)
+++ src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java	(working copy)
@@ -347,7 +347,7 @@
                 query = URLDecoder.decode(query, CHARSET_DECODE); // better ISO-8859-1 than UTF-8
             } catch(IllegalArgumentException e) {
                 log.warn("Error decoding query, maybe your request parameters should be encoded:" + query, e);
-                return null;
+                return query;
             } catch (UnsupportedEncodingException uee) {
                 log.warn("Error decoding query, maybe your request parameters should be encoded:" + query, uee);
                 return null;
