Index: src/core/org/apache/jmeter/control/TransactionController.java
===================================================================
--- src/core/org/apache/jmeter/control/TransactionController.java	(revision 780105)
+++ src/core/org/apache/jmeter/control/TransactionController.java	(working copy)
@@ -198,6 +198,8 @@
                     res.setSuccessful(false);
                     noFailingSamples++;
                 }
+                res.setAllThreads(sampleResult.getAllThreads());
+                res.setGroupThreads(sampleResult.getGroupThreads());
             }
         }
     }
