Index: src/core/org/apache/jmeter/NewDriver.java
===================================================================
--- src/core/org/apache/jmeter/NewDriver.java	(revision 497856)
+++ src/core/org/apache/jmeter/NewDriver.java	(working copy)
@@ -47,7 +47,7 @@
 
 		// Find JMeter home dir
 		StringTokenizer tok = new StringTokenizer(cp, File.pathSeparator);
-		if (tok.countTokens() == 1) {
+		if (tok.countTokens() == 1 || System.getProperty("os.name").toLowerCase().startsWith("mac os x") && tok.countTokens()  == 2) {
 			File jar = new File(tok.nextToken());
 			try {
 				jmDir = jar.getCanonicalFile().getParentFile().getParent();
