Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java	(revision 1769681)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java	(working copy)
@@ -531,7 +531,9 @@
                 }
                 if (HTTPConstants.HEADER_CONTENT_ENCODING.equalsIgnoreCase(parts[0])
                     &&
-                    HTTPConstants.ENCODING_GZIP.equalsIgnoreCase(parts[1])
+                    (HTTPConstants.ENCODING_GZIP.equalsIgnoreCase(parts[1]) ||
+                            HTTPConstants.ENCODING_DEFLATE.equalsIgnoreCase(parts[1]) ||
+                            HTTPConstants.ENCODING_BROTLI.equalsIgnoreCase(parts[1]))
                 ){
                     headerLines[i] = null; // We don't want this passed on to browser
                     fixContentLength = true;
Index: src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPConstantsInterface.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPConstantsInterface.java	(revision 1769681)
+++ src/protocol/http/org/apache/jmeter/protocol/http/util/HTTPConstantsInterface.java	(working copy)
@@ -64,7 +64,10 @@
     String HEADER_CONTENT_ENCODING = "content-encoding"; // $NON-NLS-1$
     String HTTP_1_1 = "HTTP/1.1"; // $NON-NLS-1$
     String HEADER_SET_COOKIE = "set-cookie"; // $NON-NLS-1$
+    String ENCODING_BROTLI = "br"; // $NON-NLS-1$
+    String ENCODING_DEFLATE = "deflate"; // $NON-NLS-1$
     String ENCODING_GZIP = "gzip"; // $NON-NLS-1$
+
     String HEADER_CONTENT_DISPOSITION = "Content-Disposition"; // $NON-NLS-1$
     String HEADER_CONTENT_TYPE = "Content-Type"; // $NON-NLS-1$
     String HEADER_CONTENT_LENGTH = "Content-Length"; // $NON-NLS-1$
