Index: src/components/org/apache/jmeter/control/ModuleController.java
===================================================================
--- src/components/org/apache/jmeter/control/ModuleController.java	(revision 821949)
+++ src/components/org/apache/jmeter/control/ModuleController.java	(working copy)
@@ -163,8 +163,8 @@
                 selectedNode = cloneTreeNode(selectedNode);
                 selectedNode.setEnabled(true);
             }
-            tree.add(selectedNode);
-            createSubTree(tree, selectedNode);
+            HashTree subtree = tree.add(selectedNode);
+            createSubTree(subtree, selectedNode);
         }
         return tree;
     }
