Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/IncludeControllerGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/IncludeControllerGui.java	(revision 535150)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/IncludeControllerGui.java	(working copy)
@@ -113,7 +113,16 @@
         IncludeController controller = (IncludeController)element;
         controller.setIncludePath(this.includePanel.getFilename());
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
 
+        includePanel.setFilename(""); // $NON-NLS-1$
+    }
+
 	public JPopupMenu createPopupMenu() {
 		JPopupMenu menu = new JPopupMenu();
 		JMenu addMenu = MenuFactory.makeMenus(new String[] { MenuFactory.CONFIG_ELEMENTS, MenuFactory.ASSERTIONS,
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/ModuleControllerGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/ModuleControllerGui.java	(revision 535150)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/control/gui/ModuleControllerGui.java	(working copy)
@@ -136,7 +136,18 @@
 			}
 		}
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
 
+        nodes.setSelectedIndex(-1);
+        selected = null;
+    }
+    
+
 	public JPopupMenu createPopupMenu() {
 		JPopupMenu menu = new JPopupMenu();
 		JMenu addMenu = MenuFactory.makeMenus(
