Index: src/core/org/apache/jmeter/control/WhileController.java
===================================================================
--- src/core/org/apache/jmeter/control/WhileController.java	(revision 1383743)
+++ src/core/org/apache/jmeter/control/WhileController.java	(working copy)
@@ -54,7 +54,7 @@
      */
     private boolean endOfLoop(boolean loopEnd) {
         String cnd = getCondition().trim();
-        log.debug("Condition string:" + cnd+".");
+        log.debug("Condition string at loop-" + (loopEnd ? "beginning " : "end ") + cnd+". (Called from " + (new Exception("Debug")).getStackTrace()[1].getMethodName() + ")");
         boolean res;
         // If blank, only check previous sample when at end of loop
         if ((loopEnd && cnd.length() == 0) || "LAST".equalsIgnoreCase(cnd)) {// $NON-NLS-1$
