Index: src/reports/org/apache/jmeter/gui/util/ReportMenuBar.java
===================================================================
--- src/reports/org/apache/jmeter/gui/util/ReportMenuBar.java	(revision 1043833)
+++ src/reports/org/apache/jmeter/gui/util/ReportMenuBar.java	(working copy)
@@ -491,11 +491,7 @@
         Component component = menu.getComponent();
 
         if (component.getName() != null) {
-            if (component instanceof JMenu) {
-                ((JMenu) component).setText(JMeterUtils.getResString(component.getName()));
-            } else {
-                ((JMenuItem) component).setText(JMeterUtils.getResString(component.getName()));
-            }
+            ((JMenuItem) component).setText(JMeterUtils.getResString(component.getName()));
         }
 
         MenuElement[] subelements = menu.getSubElements();
