Index: src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java	(revision 1176336)
+++ src/protocol/http/org/apache/jmeter/protocol/http/control/CookieManager.java	(working copy)
@@ -381,7 +381,7 @@
     }
 
 
-    public void addCookieFromHeader(String cookieHeader, URL url){
+    public synchronized void addCookieFromHeader(String cookieHeader, URL url){
         boolean debugEnabled = log.isDebugEnabled();
         if (debugEnabled) {
             log.debug("Received Cookie: " + cookieHeader + " From: " + url.toExternalForm());
@@ -459,7 +459,7 @@
         a.getDomain().equals(b.getDomain());
     }
 
-    private void removeMatchingCookies(Cookie newCookie){
+    private synchronized void removeMatchingCookies(Cookie newCookie){
         // Scan for any matching cookies
         PropertyIterator iter = getCookies().iterator();
         while (iter.hasNext()) {
