Index: ViewResultsFullVisualizer.java
===================================================================
--- ViewResultsFullVisualizer.java	(revision 1181638)
+++ ViewResultsFullVisualizer.java	(working copy)
@@ -357,6 +357,7 @@
                 response = JMeterUtils.getResString("view_results_response_too_large_message") //$NON-NLS-1$
                     + len + " > Max: "+MAX_DISPLAY_SIZE;
                 log.warn(response);
+                response = response + "\n" + res.getResponseDataAsString().substring(0, MAX_DISPLAY_SIZE) + "\n...";
             } else {
                 response = res.getResponseDataAsString();
             }
