Index: src/core/org/apache/jmeter/samplers/SampleResult.java
===================================================================
--- src/core/org/apache/jmeter/samplers/SampleResult.java	(revision 1671850)
+++ src/core/org/apache/jmeter/samplers/SampleResult.java	(working copy)
@@ -809,7 +809,8 @@
 
     // List of types that are known to be ascii, although they may appear to be binary
     private static final String[] NON_BINARY_TYPES = {
-        "video/f4m",       //$NON-NLS-1$ (Flash Media Manifest)
+        "audio/x-mpegurl",  //$NON-NLS-1$ (HLS Media Manifest)
+        "video/f4m"         //$NON-NLS-1$ (Flash Media Manifest)
         };
 
     /*
