diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java b/src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java
index c570db6..b1bf4ea 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/visualizers/RequestViewHTTP.java
@@ -339,7 +339,8 @@ public class RequestViewHTTP implements RequestView {
      * 
      * @param query
      *            to decode
-     * @return a decode query string
+     * @return the decoded query string, if it can be url-decoded. Otherwise the original
+     *            query will be returned.
      */
     public static String decodeQuery(String query) {
         if (query != null && query.length() > 0) {
@@ -347,14 +348,14 @@ public class RequestViewHTTP implements RequestView {
                 query = URLDecoder.decode(query, CHARSET_DECODE); // better ISO-8859-1 than UTF-8
             } catch(IllegalArgumentException e) {
                 log.warn("Error decoding query, maybe your request parameters should be encoded:" + query, e);
-                return null;
+                return query;
             } catch (UnsupportedEncodingException uee) {
                 log.warn("Error decoding query, maybe your request parameters should be encoded:" + query, uee);
-                return null;
-            } 
+                return query;
+            }
             return query;
         }
-        return null;
+        return "";
     }
 
     @Override
