--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\protocol\http\gui\AuthPanel.java	Thu Jul 26 00:34:48 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\protocol\http\gui\AuthPanel.java	Wed Sep 19 05:45:16 2001
@@ -399,7 +399,7 @@
 				{
 					try
 					{
-						File tmp = FileDialoger.promptToOpenFile()
+						File tmp = FileDialoger.promptToSaveFile(null)
 								.getSelectedFile();
 						if (tmp != null)
 						{
