Index: C:/javaopt/eclipse/eclipse_workspaces/jmeter_workspace/JMeter2_2/src/core/org/apache/jmeter/util/Calculator.java
===================================================================
--- C:/javaopt/eclipse/eclipse_workspaces/jmeter_workspace/JMeter2_2/src/core/org/apache/jmeter/util/Calculator.java	(revision 454855)
+++ C:/javaopt/eclipse/eclipse_workspaces/jmeter_workspace/JMeter2_2/src/core/org/apache/jmeter/util/Calculator.java	(working copy)
@@ -105,7 +105,8 @@
         if (startTime == 0){
             startTime=res.getStartTime();
         }
-        elapsedTime=res.getEndTime()-startTime;
+        startTime = Math.min(startTime, res.getStartTime());
+        elapsedTime = Math.max(elapsedTime, res.getEndTime()-startTime);
     }
 
 
