Index: stylesheets/site.vsl
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/xdocs/stylesheets/site.vsl,v
retrieving revision 1.16
diff -u -r1.16 site.vsl
--- stylesheets/site.vsl	13 Jun 2003 14:51:51 -0000	1.16
+++ stylesheets/site.vsl	13 Jun 2003 22:53:13 -0000
@@ -249,7 +249,7 @@
 
 #macro ( component $component)
 	#set ($screenshot = "")
-	#set ($screenshot = "$sshotdir/$component.getAttributeValue('screenshot')" )
+	#set ($screenshot = $component.getAttributeValue('screenshot') )
 	<table border="0" cellspacing="0" cellpadding="2">
 		<tr><td>
 		  <font face="arial,helvetica,sanserif">
@@ -265,7 +265,7 @@
 					#runloop($c_items)
 					#if ($screenshot != "")						
 						<p><b>Control Panel</b></p>
-						<div align="center"><img src="$screenshot"></div>
+						<div align="center"><img src="$sshotdir/$screenshot"></div>
 					#end
 				#else
 					 #unknown($c_items)
Index: stylesheets/site_printable.vsl
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/xdocs/stylesheets/site_printable.vsl,v
retrieving revision 1.7
diff -u -r1.7 site_printable.vsl
--- stylesheets/site_printable.vsl	13 Jun 2003 14:51:51 -0000	1.7
+++ stylesheets/site_printable.vsl	13 Jun 2003 22:53:15 -0000
@@ -243,7 +243,7 @@
 
 #macro ( component $component)
 	#set ($screenshot = "")
-	#set ($screenshot = "$sshotdir/$component.getAttributeValue('screenshot')" )
+	#set ($screenshot = $component.getAttributeValue('screenshot') )
 	<table border="0" cellspacing="0" cellpadding="2">
 		<tr><td>
 		  <font face="arial,helvetica,sanserif">
@@ -254,9 +254,9 @@
 		  #foreach ( $items in $component.getChildren() )
 				#if ($items.getName().equals("description"))
 					#runloop($items)
-					<p><b>Control Panel</b></p>
 					#if ($screenshot != "")
-						<div align="center"><img src="$screenshot"></div>
+					<p><b>Control Panel</b></p>
+						<div align="center"><img src="$sshotdir/$screenshot"></div>
 					#end
 				#else
 					 #unknown($items)