Index: src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java
===================================================================
--- src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java	(revision 910417)
+++ src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java	(working copy)
@@ -337,6 +337,11 @@
             res.setResponseCode("500"); //$NON-NLS-1$
             res.setResponseMessage(ex.toString());
             closeSocket();
+        } catch (Exception ex) {
+            log.error("", ex);
+            isSuccessful=false;
+            res.setResponseCode("500");
+            res.setResponseMessage(ex.toString());
         } finally {
             // Calculate response time
             res.sampleEnd();
