Index: src/core/org/apache/jmeter/control/IfController.java
===================================================================
--- src/core/org/apache/jmeter/control/IfController.java	(revision 1177675)
+++ src/core/org/apache/jmeter/control/IfController.java	(working copy)
@@ -123,6 +123,7 @@
             logger.debug("    >> evaluate Condition -  [ " + cond + "] results is  [" + result + "]");
         } catch (Exception e) {
             logger.error(getName()+": error while processing "+ "[" + cond + "]\n", e);
+            throw new IllegalStateException(getName()+": error while processing "+ "[" + cond + "]\n", e);
         } finally {
             Context.exit();
         }
