Index: src/jorphan/org/apache/jorphan/math/StatCalculator.java
===================================================================
--- src/jorphan/org/apache/jorphan/math/StatCalculator.java	(revision 764194)
+++ src/jorphan/org/apache/jorphan/math/StatCalculator.java	(working copy)
@@ -18,7 +18,7 @@
 
 package org.apache.jorphan.math;
 
-import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Iterator;
@@ -29,7 +29,7 @@
  * not threadsafe.
  */
 public class StatCalculator {
-    List values = new ArrayList();
+    List values = new LinkedList();
 
     double sum = 0;
 
