Index: JMeter.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/JMeter.java,v
retrieving revision 1.9
diff -u -r1.9 JMeter.java
--- JMeter.java	5 Feb 2003 05:12:04 -0000	1.9
+++ JMeter.java	19 Mar 2003 19:11:01 -0000
@@ -153,7 +153,7 @@
 				LOGFILE_OPT,
 				"the file to log samples to"),
 			new CLOptionDescriptor(
-				"logfile",
+				"nogui",
 				CLOptionDescriptor.ARGUMENT_DISALLOWED,
 				NONGUI_OPT,
 				"run JMeter in nongui mode"),
