Index: core/org/apache/jmeter/config/gui/LoginConfigGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/config/gui/LoginConfigGui.java,v
retrieving revision 1.6
diff -u -r1.6 LoginConfigGui.java
--- core/org/apache/jmeter/config/gui/LoginConfigGui.java	11 May 2003 13:02:30 -0000	1.6
+++ core/org/apache/jmeter/config/gui/LoginConfigGui.java	12 Jun 2003 21:18:59 -0000
@@ -114,7 +114,7 @@
     /* Implements JMeterGUIComponent.getStaticLabel() */
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("Login Config Element");
+        return JMeterUtils.getResString("login_config_element");
     }
 
     /**
Index: core/org/apache/jmeter/config/gui/SimpleConfigGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/config/gui/SimpleConfigGui.java,v
retrieving revision 1.5
diff -u -r1.5 SimpleConfigGui.java
--- core/org/apache/jmeter/config/gui/SimpleConfigGui.java	11 May 2003 13:02:30 -0000	1.5
+++ core/org/apache/jmeter/config/gui/SimpleConfigGui.java	12 Jun 2003 21:19:00 -0000
@@ -147,7 +147,7 @@
     /* Implements JMeterGUIComponent.getStaticLabel() */
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("Simple Config Element");
+        return JMeterUtils.getResString("simple_config_element");
     }
 
     /**
Index: core/org/apache/jmeter/control/gui/TestPlanGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/control/gui/TestPlanGui.java,v
retrieving revision 1.13
diff -u -r1.13 TestPlanGui.java
--- core/org/apache/jmeter/control/gui/TestPlanGui.java	13 May 2003 17:42:56 -0000	1.13
+++ core/org/apache/jmeter/control/gui/TestPlanGui.java	12 Jun 2003 21:19:01 -0000
@@ -104,7 +104,7 @@
     /* Implements JMeterGUIComponent.getStaticLabel() */
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("Test Plan");
+        return JMeterUtils.getResString("test_plan");
     }
 
     /**
Index: core/org/apache/jmeter/threads/gui/ThreadGroupGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/threads/gui/ThreadGroupGui.java,v
retrieving revision 1.17
diff -u -r1.17 ThreadGroupGui.java
--- core/org/apache/jmeter/threads/gui/ThreadGroupGui.java	11 Jun 2003 15:38:33 -0000	1.17
+++ core/org/apache/jmeter/threads/gui/ThreadGroupGui.java	12 Jun 2003 21:19:04 -0000
@@ -101,9 +101,9 @@
     private JTextField threadInput;
     private JTextField rampInput;
 
-    private final static String SCHEDULER = "scheduler";
-    private final static String START_TIME= "start_time";
-    private final static String END_TIME= "end_time";
+    //private final static String SCHEDULER = "scheduler";
+    //private final static String START_TIME= "start_time";
+    //private final static String END_TIME= "end_time";
 
     private JDateField start;
     private JDateField end;
@@ -263,7 +263,7 @@
      ***************************************/
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("ThreadGroup");
+        return JMeterUtils.getResString("threadgroup");
     }
 
     private void init()
Index: protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java,v
retrieving revision 1.8
diff -u -r1.8 ParamModifierGui.java
--- protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java	8 May 2003 18:29:15 -0000	1.8
+++ protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java	12 Jun 2003 21:19:06 -0000
@@ -57,7 +57,7 @@
      ***************************************/
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("HTML Parameter Mask");
+        return JMeterUtils.getResString("html_parameter_mask");
     }
 
     /****************************************
@@ -225,7 +225,7 @@
     private long checkTextField(FocusEvent evt, String defaultValue)
     {
         JTextField temp = (JTextField) evt.getSource();
-        boolean pass = true;
+        //boolean pass = true;
         long longVal = 0;
 
         try
Index: protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java,v
retrieving revision 1.7
diff -u -r1.7 UserParameterModifierGui.java
--- protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java	8 May 2003 18:29:15 -0000	1.7
+++ protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java	12 Jun 2003 21:19:12 -0000
@@ -111,7 +111,7 @@
 	 ***********************************************************/
 	public void updateGui() {}
 	public String getStaticLabel() {
-		return JMeterUtils.getResString("HTTP User Parameter Modifier");
+		return JMeterUtils.getResString("http_user_parameter_modifier");
 	}
 	
 	public void configure(TestElement el)
Index: protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java,v
retrieving revision 1.13
diff -u -r1.13 JavaConfigGui.java
--- protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java	8 May 2003 21:03:38 -0000	1.13
+++ protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java	12 Jun 2003 21:19:15 -0000
@@ -136,7 +136,7 @@
     /* Implements JMeterGUIComponent.getStaticLabel() */
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("Java Request Defaults");
+        return JMeterUtils.getResString("java_request_defaults");
     }
 
     /**
Index: protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java,v
retrieving revision 1.10
diff -u -r1.10 JavaTestSamplerGui.java
--- protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java	8 Jun 2003 11:03:26 -0000	1.10
+++ protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java	12 Jun 2003 21:19:15 -0000
@@ -89,7 +89,7 @@
     /* Implements JMeterGuiComponent.getStaticLabel() */
     public String getStaticLabel()
     {
-        return JMeterUtils.getResString("Java Request");
+        return JMeterUtils.getResString("java_request");
     }
 
     /**
