diff -Naur src/protocol/http/org/apache/jmeter/protocol/http/config/gui/HttpDefaultsGui.java new/protocol/http/org/apache/jmeter/protocol/http/config/gui/HttpDefaultsGui.java
--- src/protocol/http/org/apache/jmeter/protocol/http/config/gui/HttpDefaultsGui.java	2002-12-02 01:44:41.000000000 +0100
+++ new/protocol/http/org/apache/jmeter/protocol/http/config/gui/HttpDefaultsGui.java	2002-12-02 01:14:24.000000000 +0100
@@ -23,6 +23,7 @@
  */
 public class HttpDefaultsGui extends AbstractConfigGui
 {
+	JLabeledTextField protocol;
 	JLabeledTextField domain;
 	JLabeledTextField path;
 	JLabeledTextField port;
@@ -51,6 +52,7 @@
 	{
 		ConfigTestElement config = new ConfigTestElement();
 		super.configureTestElement(config);
+		config.setProperty(HTTPSampler.PROTOCOL,protocol.getText());
 		config.setProperty(HTTPSampler.DOMAIN,domain.getText());
 		config.setProperty(HTTPSampler.PATH,path.getText());
 		config.setProperty(HTTPSampler.ARGUMENTS,argPanel.createTestElement());
@@ -61,6 +63,7 @@
 	public void configure(TestElement el)
 	{
 		super.configure(el);
+		protocol.setText(el.getPropertyAsString(HTTPSampler.PROTOCOL));
 		domain.setText(el.getPropertyAsString(HTTPSampler.DOMAIN));
 		path.setText(el.getPropertyAsString(HTTPSampler.PATH));
 		port.setText(el.getPropertyAsString(HTTPSampler.PORT));
@@ -74,6 +77,7 @@
 		this.setLayout(new BorderLayout());
 		argPanel = new HTTPArgumentsPanel();
 		this.add(argPanel,BorderLayout.CENTER);
+		protocol = new JLabeledTextField(JMeterUtils.getResString("url_config_protocol"));
 		domain = new JLabeledTextField(JMeterUtils.getResString("web_server_domain"));
 		path = new JLabeledTextField(JMeterUtils.getResString("path"));
 		port = new JLabeledTextField(JMeterUtils.getResString("web_server_port"));
@@ -86,6 +90,7 @@
 		
 		topPanel.add(title);
 		topPanel.add(getNamePanel());
+		topPanel.add(protocol);
 		topPanel.add(domain);
 		topPanel.add(path);
 		topPanel.add(port);
diff -Naur src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java new/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java
--- src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java	2002-12-02 01:44:41.000000000 +0100
+++ new/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java	2002-12-02 01:23:53.000000000 +0100
@@ -93,24 +93,22 @@
 	protected HTTPArgumentsPanel argsPanel;
 	private static String DOMAIN = "domain";
 	private static String PORT = "port";
+	private static String PROTOCOL = "protocol";
 	private static String PATH = "path";
 	private static String FOLLOW_REDIRECTS = "follow_redirects";
 	private static String USE_KEEPALIVE = "use_keepalive";
 	private static String POST = "post";
 	private static String GET = "get";
-	private static String HTTP = "http";
-	private static String HTTPS = "https";
 	private static String SEND_PARAM = "sendparam";
 
 	private JTextField domain;
 	private JTextField port;
+	private JTextField protocol;
 	private JTextField path;
 	private JCheckBox followRedirects;
 	private JCheckBox useKeepAlive;
 	private JRadioButton post;
 	private JRadioButton get;
-	private JRadioButton http;
-	private JRadioButton https;
 
 
 	/****************************************
@@ -143,11 +141,11 @@
 		element.setProperty(HTTPSampler.ARGUMENTS, args);
 		element.setProperty(HTTPSampler.DOMAIN, domain.getText());
 		element.setProperty(HTTPSampler.PORT, port.getText());
+		element.setProperty(HTTPSampler.PROTOCOL, protocol.getText());
 		element.setProperty(HTTPSampler.METHOD, (post.isSelected() ? "POST" : "GET"));
 		element.setProperty(HTTPSampler.PATH, path.getText());
 		element.setProperty(HTTPSampler.FOLLOW_REDIRECTS, new Boolean(followRedirects.isSelected()));
 		element.setProperty(HTTPSampler.USE_KEEPALIVE, new Boolean(useKeepAlive.isSelected()));
-		element.setProperty(HTTPSampler.PROTOCOL, (http.isSelected() ? "http" : "https"));
 		return element;
 	}
 
@@ -155,6 +153,7 @@
 	{
 		sampler.setArguments((Arguments)argsPanel.createTestElement());
 		sampler.setDomain(domain.getText());
+		sampler.setProtocol(protocol.getText());
 		sampler.setPath(path.getText());
 		sampler.setFollowRedirects(followRedirects.isSelected());
 		sampler.setUseKeepAlive(useKeepAlive.isSelected());
@@ -163,7 +162,6 @@
 			sampler.setPort(Integer.parseInt(port.getText()));
 		}
 		sampler.setMethod((post.isSelected() ? "POST" : "GET"));
-		sampler.setProtocol((http.isSelected() ? "http" : "https"));
 	}
 
 	/****************************************
@@ -177,6 +175,7 @@
 		argsPanel.configure((TestElement)el.getProperty(HTTPSampler.ARGUMENTS));
 		domain.setText((String)el.getProperty(HTTPSampler.DOMAIN));
 		port.setText((String)el.getPropertyAsString(HTTPSampler.PORT));
+		protocol.setText((String)el.getProperty(HTTPSampler.PROTOCOL));
 		if("POST".equals(el.getProperty(HTTPSampler.METHOD)))
 		{
 			post.setSelected(true);
@@ -190,16 +189,6 @@
 		path.setText((String)el.getProperty(HTTPSampler.PATH));
 		followRedirects.setSelected(((AbstractTestElement)el).getPropertyAsBoolean(HTTPSampler.FOLLOW_REDIRECTS));
 		useKeepAlive.setSelected(((AbstractTestElement)el).getPropertyAsBoolean(HTTPSampler.USE_KEEPALIVE));
-		if("http".equals(el.getProperty(HTTPSampler.PROTOCOL)))
-		{
-			http.setSelected(true);
-			https.setSelected(false);
-		}
-		else
-		{
-			https.setSelected(true);
-			http.setSelected(false);
-		}
 	}
 
 	/****************************************
@@ -305,26 +294,14 @@
 	 ***************************************/
 	protected JPanel getProtocolAndMethodPanel()
 	{
-		JPanel panel = new JPanel();
-		panel.setLayout(new BoxLayout(panel, BoxLayout.X_AXIS));
-		panel.setBorder(BorderFactory.createEmptyBorder(0, 10, 0, 10));
-
-		panel.add(new JLabel(JMeterUtils.getResString("protocol")));
-		panel.add(Box.createRigidArea(new Dimension(5, 0)));
+		JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
+		panel.setBorder(BorderFactory.createEmptyBorder(0, 5, 0, 10));
 
 		// PROTOCOL
-		http = new JRadioButton(JMeterUtils.getResString("url_config_http"));
-		https = new JRadioButton(JMeterUtils.getResString("url_config_https"));
-		ButtonGroup protocolButtonGroup = new ButtonGroup();
-		protocolButtonGroup.add(http);
-		protocolButtonGroup.add(https);
-
-		http.setSelected(true);
-
-		panel.add(http);
-		panel.add(https);
-
-		panel.add(Box.createRigidArea(new Dimension(20, 0)));
+		panel.add(new JLabel(JMeterUtils.getResString("protocol")));
+		protocol = new JTextField(4);
+		protocol.setName(PROTOCOL);
+		panel.add(protocol);
 
 		// METHOD
 		post = new JRadioButton(JMeterUtils.getResString("url_config_post"));
@@ -333,8 +310,8 @@
 		methodButtonGroup.add(post);
 		methodButtonGroup.add(get);
 
-		panel.add(new JLabel(JMeterUtils.getResString("method")));
 		panel.add(Box.createRigidArea(new Dimension(5, 0)));
+		panel.add(new JLabel(JMeterUtils.getResString("method")));
 
 		post.setSelected(true);
 
