Index: src/core/org/apache/jmeter/save/CSVSaveService.java
===================================================================
--- src/core/org/apache/jmeter/save/CSVSaveService.java	(revision 1155546)
+++ src/core/org/apache/jmeter/save/CSVSaveService.java	(working copy)
@@ -19,7 +19,7 @@
 package org.apache.jmeter.save;
 
 import java.io.BufferedReader;
-import java.io.ByteArrayOutputStream;
+import java.io.CharArrayWriter;
 import java.io.FileReader;
 import java.io.FileWriter;
 import java.io.IOException;
@@ -937,7 +937,7 @@
         int ch;
         int state = INITIAL;
         List<String> list = new ArrayList<String>();
-        ByteArrayOutputStream baos = new ByteArrayOutputStream(200);
+        CharArrayWriter baos = new CharArrayWriter(200);        
         boolean push = false;
         while(-1 != (ch=infile.read())){
             push = false;
