Index: src/core/org/apache/jmeter/samplers/BatchSampleSender.java
===================================================================
--- src/core/org/apache/jmeter/samplers/BatchSampleSender.java	(revision 1001614)
+++ src/core/org/apache/jmeter/samplers/BatchSampleSender.java	(working copy)
@@ -66,7 +66,7 @@
      * @param listener
      *            that the List of sample events will be sent to.
      */
-    BatchSampleSender(RemoteSampleListener listener) {
+    protected BatchSampleSender(RemoteSampleListener listener) {
         this.listener = listener;
         log.info("Using batching for this run."
                 + " Thresholds: num=" + NUM_SAMPLES_THRESHOLD
@@ -156,4 +156,16 @@
             }
         } // synchronized(sampleStore)
     }
+	/**
+	 * @return the listener
+	 */
+	protected RemoteSampleListener getListener() {
+		return listener;
+	}
+	/**
+	 * @return the sampleStore
+	 */
+	protected List<SampleEvent> getSampleStore() {
+		return sampleStore;
+	}
 }
