diff --git a/src/core/src/main/java/org/apache/jmeter/threads/ThreadGroup.java b/src/core/src/main/java/org/apache/jmeter/threads/ThreadGroup.java
index ad0e001218..a9e6773987 100644
--- a/src/core/src/main/java/org/apache/jmeter/threads/ThreadGroup.java
+++ b/src/core/src/main/java/org/apache/jmeter/threads/ThreadGroup.java
@@ -228,7 +228,7 @@ public class ThreadGroup extends AbstractThreadGroup {
             final JMeterContext context = JMeterContextService.getContext();
             long lastThreadStartInMillis = 0;
             int delayForNextThreadInMillis = 0;
-            final int perThreadDelayInMillis = Math.round((float) rampUpPeriodInSeconds * 1000 / numThreads);
+            final int perThreadDelayInMillis = Math.round((float) rampUpPeriodInSeconds * 1000 / Math.max(numThreads - 1, 1));
             for (int threadNum = 0; running && threadNum < numThreads; threadNum++) {
                 long nowInMillis = System.currentTimeMillis();
                 if(threadNum > 0) {
