Index: src/core/org/apache/jmeter/util/keystore/JmeterKeyStore.java
===================================================================
--- src/core/org/apache/jmeter/util/keystore/JmeterKeyStore.java	(revision 1244197)
+++ src/core/org/apache/jmeter/util/keystore/JmeterKeyStore.java	(working copy)
@@ -202,11 +202,10 @@
 
     private int getNextIndex(int length) {
         synchronized(this) {
-            last_user ++;
             if (last_user >= length) {
                 last_user = 0;
             }
-            return last_user;
+            return last_user ++;
         }
     }
 

