Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java	(revision 1166194)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java	(working copy)
@@ -35,7 +35,7 @@
 /**
  * A parser for html, to find the form tags, and their accept-charset value
  */
-class FormCharSetFinder {
+public class FormCharSetFinder {
     private static final Logger log = LoggingManager.getLoggerForClass();
 
     static {
