Index: src/jorphan/org/apache/jorphan/reflect/Functor.java
===================================================================
--- src/jorphan/org/apache/jorphan/reflect/Functor.java	(revision 385445)
+++ src/jorphan/org/apache/jorphan/reflect/Functor.java	(working copy)
@@ -193,12 +193,10 @@
 
 	protected Class[] getNewArray(int i, Class replacement, Class[] orig) {
 		Class[] newArray = new Class[orig.length];
-		for (int j = 0; j < newArray.length; j++) {
-			newArray[j] = orig[j];
-			if (j == i) {
-				newArray[j] = replacement;
-			}
-		}
+		System.arraycopy(orig, 0, newArray, 0, orig.length);
+		if ((i >= 0) && (i < orig.length))
+			newArray[i] = replacement;
+
 		return newArray;
 	}
 
