Index: util/FileDialoger.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/gui/util/FileDialoger.java,v
retrieving revision 1.9
diff -u -r1.9 FileDialoger.java
--- util/FileDialoger.java	17 Jun 2004 12:41:52 -0000	1.9
+++ util/FileDialoger.java	25 May 2005 16:04:15 -0000
@@ -38,7 +38,6 @@
      */
     private static String lastJFCDirectory = null;
     private static JFileChooser jfc = new JFileChooser();
-    private static String[] extensions = null;
 
     /**
      * Prevent instantiation of utility class.
@@ -107,10 +106,9 @@
 	 * @param filters
 	 * @return
 	 */
-	public static JFileChooser promptToSaveFile(String filename, String[] filters){
-		extensions = filters;
-		JFileChooser chooser = promptToSaveFile(filename);
-		return chooser;
+	public static JFileChooser promptToSaveFile(String filename)
+	{
+		return promptToSaveFile(filename, null);
 	}
 	
     /**
@@ -125,7 +123,7 @@
      *           are finished using it (accept or otherwise).
      * @see             #promptToOpenFile
      */
-    public static JFileChooser promptToSaveFile(String filename)
+    public static JFileChooser promptToSaveFile(String filename, String[] extensions)
     {
         if (lastJFCDirectory == null)
         {
