Index: src/core/org/apache/jmeter/testbeans/TestBeanHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/core/org/apache/jmeter/testbeans/TestBeanHelper.java	(date 1388842110000)
+++ src/core/org/apache/jmeter/testbeans/TestBeanHelper.java	(revision )
@@ -94,7 +94,7 @@
                 {
                     Method writeMethod = desc.getWriteMethod();
                     if (writeMethod!=null) {
-                        invokeOrBailOut(el, writeMethod, new Object[] {value});
+                        invokeOrBailOut(el, writeMethod, new Object[] {value instanceof String?((String) value).trim():value});
                     }
                 }
             }
Index: src/core/org/apache/jmeter/testbeans/gui/GenericTestBeanCustomizer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/core/org/apache/jmeter/testbeans/gui/GenericTestBeanCustomizer.java	(date 1388842110000)
+++ src/core/org/apache/jmeter/testbeans/gui/GenericTestBeanCustomizer.java	(revision )
@@ -747,7 +747,7 @@
                         log.debug("Unset " + name);
                     }
                 } else {
-                    propertyMap.put(name, value);
+                    propertyMap.put(name, value instanceof String?((String) value).trim():value);
                     if (log.isDebugEnabled()) {
                         log.debug("Set " + name + "= " + value);
                     }
