Index: src/core/org/apache/jmeter/threads/JMeterThread.java
===================================================================
--- src/core/org/apache/jmeter/threads/JMeterThread.java	(revision 1416913)
+++ src/core/org/apache/jmeter/threads/JMeterThread.java	(working copy)
@@ -336,8 +336,8 @@
         for (Iterator<Controller> iterator = controllersToReinit
                 .iterator(); iterator.hasNext();) {
             Controller parentController =  iterator.next();
-            if(parentController instanceof ThreadGroup) {
-                ThreadGroup tg = (ThreadGroup) parentController;
+            if(parentController instanceof AbstractThreadGroup) {
+                AbstractThreadGroup tg = (AbstractThreadGroup) parentController;
                 tg.startNextLoop();
             } else {
                 parentController.triggerEndOfLoop();
