Index: SizeAssertionGui.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/components/org/apache/jmeter/assertions/gui/SizeAssertionGui.java,v
retrieving revision 1.11
diff -u -r1.11 SizeAssertionGui.java
--- SizeAssertionGui.java	19 Oct 2003 22:13:11 -0000	1.11
+++ SizeAssertionGui.java	26 Jan 2004 17:32:18 -0000
@@ -236,6 +236,7 @@
             createComparatorButton("<=", SizeAssertion.LESSTHANEQUAL, group);
 
         equalButton.setSelected(true);
+        execState = Integer.parseInt(equalButton.getActionCommand());
 
         //Put the check boxes in a column in a panel
         Box checkPanel = Box.createVerticalBox();
