Index: src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java
===================================================================
--- src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java	(revision 698443)
+++ src/protocol/tcp/org/apache/jmeter/protocol/tcp/sampler/TCPSampler.java	(working copy)
@@ -208,7 +208,7 @@
     }
 
     public void setTimeout(String newTimeout) {
-        this.setProperty(FILENAME, newTimeout);
+        this.setProperty(TIMEOUT, newTimeout);
     }
 
     public int getTimeout() {
