Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java	(revision 789045)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/Proxy.java	(working copy)
@@ -333,7 +333,7 @@
                 final String HTTPS_PREFIX = "https://";
                 if (forcedHTTPS && HTTPConstants.HEADER_LOCATION.equalsIgnoreCase(parts[0])
                         && parts[1].substring(0, HTTPS_PREFIX.length()).equalsIgnoreCase(HTTPS_PREFIX)){
-                    headerLines[i]=headerLines[i].replaceFirst(HTTPS_PREFIX, "http://");
+                	headerLines[i]=headerLines[i].replaceFirst(parts[1].substring(0,HTTPS_PREFIX.length()), "http://");
                     continue;
                 }
             }
