Index: stylesheets/site.vsl
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/xdocs/stylesheets/site.vsl,v
retrieving revision 1.13
diff -u -r1.13 site.vsl
--- stylesheets/site.vsl	15 Oct 2002 20:28:42 -0000	1.13
+++ stylesheets/site.vsl	8 Jun 2003 20:11:00 -0000
@@ -404,7 +404,6 @@
 						  </td>
 					 </tr>
 
-					 <!-- FOOTER -->
 					 <tr><td colspan="2">
 						  <hr noshade="" size="1"/>
 					 </td></tr>
Index: stylesheets/site_printable.vsl
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/xdocs/stylesheets/site_printable.vsl,v
retrieving revision 1.3
diff -u -r1.3 site_printable.vsl
--- stylesheets/site_printable.vsl	24 Apr 2003 20:00:43 -0000	1.3
+++ stylesheets/site_printable.vsl	8 Jun 2003 20:11:02 -0000
@@ -397,7 +397,6 @@
 						  </td>
 					 </tr>
 
-					 <!-- FOOTER -->
 					 <tr><td>
 						  <hr noshade="" size="1"/>
 					 </td></tr>
