diff -ruN /home/schizophrenia/workspace/jmeter/src/core/org/apache/jmeter/control/TransactionSampler.java /home/schizophrenia/workspace/jmeter-TransactionController/src/core/org/apache/jmeter/control/TransactionSampler.java
--- /home/schizophrenia/workspace/jmeter/src/core/org/apache/jmeter/control/TransactionSampler.java	2013-05-15 15:44:42.494060819 +0400
+++ /home/schizophrenia/workspace/jmeter-TransactionController/src/core/org/apache/jmeter/control/TransactionSampler.java	2013-05-16 17:59:38.421521306 +0400
@@ -105,6 +105,12 @@
     public void addSubSamplerResult(SampleResult res) {
         // Another subsample for the transaction
         calls++;
+        
+        // Set Response code of transaction
+        if (noFailingSamples == 0) {
+        	transactionSampleResult.setResponseCode(res.getResponseCode());
+        }
+        
         // The transaction fails if any sub sample fails
         if (!res.isSuccessful()) {
             transactionSampleResult.setSuccessful(false);
