diff --git a/src/core/src/main/java/org/apache/jmeter/functions/gui/FunctionHelper.java b/src/core/src/main/java/org/apache/jmeter/functions/gui/FunctionHelper.java
index 3437fd5c5a..d03df4000b 100644
--- a/src/core/src/main/java/org/apache/jmeter/functions/gui/FunctionHelper.java
+++ b/src/core/src/main/java/org/apache/jmeter/functions/gui/FunctionHelper.java
@@ -239,6 +239,7 @@ public class FunctionHelper extends JDialog implements ActionListener, ChangeLis
             Arguments args = (Arguments) parameterPanel.createTestElement();
             String functionCall = buildFunctionCallString(functionName, args);
             cutPasteFunction.setText(functionCall);
+            cutPasteFunction.setEnabled(true);
             GuiUtils.copyTextToClipboard(cutPasteFunction.getText());
             CompoundVariable function = new CompoundVariable(functionCall);
             JMeterContext threadContext = JMeterContextService.getContext();
