Index: src/core/org/apache/jmeter/JMeter.java
===================================================================
--- src/core/org/apache/jmeter/JMeter.java	(revision 780105)
+++ src/core/org/apache/jmeter/JMeter.java	(working copy)
@@ -736,18 +736,29 @@
             // For GUI runs this is done in Start.java
             convertSubTree(tree);
 
+            Summariser summer = null;
+            String summariserName = JMeterUtils.getPropDefault("summariser.name", "");//$NON-NLS-1$
+            if (summariserName.length() > 0) {
+                log.info("Creating summariser <" + summariserName + ">");
+                println("Creating summariser <" + summariserName + ">");
+                summer = new Summariser(summariserName);
+            }
+            
             if (logFile != null) {
                 ResultCollector logger = new ResultCollector();
                 logger.setFilename(logFile);
+                if (summer != null) {
+                    logger.setSummariser(summer);
+                }
                 tree.add(tree.getArray()[0], logger);
             }
-            String summariserName = JMeterUtils.getPropDefault("summariser.name", "");//$NON-NLS-1$
-            if (summariserName.length() > 0) {
-                log.info("Creating summariser <" + summariserName + ">");
-                println("Creating summariser <" + summariserName + ">");
-                Summariser summer = new Summariser(summariserName);
-                tree.add(tree.getArray()[0], summer);
+            else {
+            	// only add Summariser if it can not be shared with the ResultCollector
+	            if (summer != null) {
+	                tree.add(tree.getArray()[0], summer);
+	            }
             }
+            
             List engines = new LinkedList();
             tree.add(tree.getArray()[0], new ListenToTest(parent, (remoteStart && remoteStop) ? engines : null));
             println("Created the tree successfully using "+testFile);
