diff -r 2ce792adafc4 src/core/org/apache/jmeter/testbeans/gui/TestBeanGUI.java
--- a/src/core/org/apache/jmeter/testbeans/gui/TestBeanGUI.java	Thu Feb 16 11:21:43 2012 +0900
+++ b/src/core/org/apache/jmeter/testbeans/gui/TestBeanGUI.java	Thu Feb 23 13:23:42 2012 +0900
@@ -211,8 +211,8 @@
    public TestElement createTestElement() {
         try {
             TestElement element = (TestElement) testBeanClass.newInstance();
-            // configure(element);
-            // super.clear(); // set name, enabled.
+            element.setName(getName());
+            configure(element);
             modifyTestElement(element); // put the default values back into the
             // new element
             return element;
