Index: src/protocol/http/org/apache/jmeter/protocol/http/gui/HTTPFileArgsPanel.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/gui/HTTPFileArgsPanel.java	(revision 1763872)
+++ src/protocol/http/org/apache/jmeter/protocol/http/gui/HTTPFileArgsPanel.java	(working copy)
@@ -142,17 +142,6 @@
     }
 
     /**
-     * Resize the table columns to appropriate widths.
-     *
-     * @param table
-     *  the table to resize columns for
-     */
-    private void sizeColumns(JTable table) {
-        GuiUtils.fixSize(table.getColumn(PARAMNAME), table);
-        GuiUtils.fixSize(table.getColumn(MIMETYPE), table);
-    }
-
-    /**
      * Save the GUI data in the HTTPSamplerBase element.
      *
      * @param testElement {@link TestElement} to modify
@@ -403,7 +392,6 @@
         p.add(makeButtonPanel(), BorderLayout.SOUTH);
 
         table.revalidate();
-        sizeColumns(table);
     }
 
     private JScrollPane makeScrollPane(Component comp) {
