Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java	(revision 1176336)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/FormCharSetFinder.java	(working copy)
@@ -43,7 +43,7 @@
         log.info("Using htmlparser version: "+Parser.getVersion());
     }
 
-    protected FormCharSetFinder() {
+    public FormCharSetFinder() {
         super();
     }
 
