Index: src/core/org/apache/jmeter/testbeans/gui/ComboStringEditor.java
===================================================================
--- src/core/org/apache/jmeter/testbeans/gui/ComboStringEditor.java	(revision 385445)
+++ src/core/org/apache/jmeter/testbeans/gui/ComboStringEditor.java	(working copy)
@@ -20,6 +20,7 @@
 import java.awt.event.ItemEvent;
 import java.awt.event.ItemListener;
 import java.beans.PropertyEditorSupport;
+import java.util.Arrays;
 
 import javax.swing.DefaultComboBoxModel;
 import javax.swing.JComboBox;
@@ -274,7 +275,7 @@
 	 * @param strings
 	 */
 	public void setTags(String[] strings) {
-		if (tags.equals(strings))
+		if (Arrays.equals(tags, strings))
 			return;
 
 		for (int i = 0; i < tags.length; i++)
