Index: C:/eclipse/eclipse322workspaces/eclipsercp/jmeter/src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java
===================================================================
--- C:/eclipse/eclipse322workspaces/eclipsercp/jmeter/src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java	(revision 532089)
+++ C:/eclipse/eclipse322workspaces/eclipsercp/jmeter/src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java	(working copy)
@@ -152,8 +152,12 @@
 		poolController.setAttribute("blocking", "true");
 		poolController.setAttribute("timeout", getTimeout());
 		poolController.setAttribute("trim-interval", getTrimInterval());
-		poolController.setAttribute("auto-commit", String.valueOf(isAutocommit()));
 		config.addChild(poolController);
+		DefaultConfiguration autoCommit = new DefaultConfiguration("auto-commit");
+		autoCommit.setValue(String.valueOf(isAutocommit()));
+		config.addChild(autoCommit);
+		
+//		config.setAttribute("auto-commit", String.valueOf(isAutocommit()));
 
 		if (log.isDebugEnabled()) {
 			StringBuffer sb = new StringBuffer(40);
