diff -r 707d572b9b88 -r acef377173b5 src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java	Tue Feb 28 10:11:08 2012 +0900
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPHC4Impl.java	Tue Feb 28 14:34:10 2012 +0900
@@ -804,7 +804,8 @@
 
             // Write the request to our own stream
             MultipartEntity multiPart = new MultipartEntity(
-                    getDoBrowserCompatibleMultipart() ? HttpMultipartMode.BROWSER_COMPATIBLE : HttpMultipartMode.STRICT);
+                    getDoBrowserCompatibleMultipart() ? HttpMultipartMode.BROWSER_COMPATIBLE : HttpMultipartMode.STRICT,
+                            null, Charset.forName(contentEncoding));
             // Create the parts
             // Add any parameters
             PropertyIterator args = getArguments().iterator();
