--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\threads\ThreadGroup.java	Thu Aug 30 17:19:04 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\threads\ThreadGroup.java	Tue Sep 18 13:23:18 2001
@@ -83,11 +83,13 @@
 	LoopController control;
 	List listeners = new LinkedList();
 	List remoteListeners = new LinkedList();
-	int numThreads = 1;
+	private final int DEFAULT_NUM_THREADS = 1;
+	private final int DEFAULT_RAMP_UP = 0;
+	int numThreads = DEFAULT_NUM_THREADS;
 	List timers = new LinkedList();
 	private String name;
 	private SampleQueue queue = null;
-	int rampUp = 0;
+	int rampUp = DEFAULT_RAMP_UP;
 	private int threadsStarted = 0;
 
 	private static List addableList;
@@ -199,6 +201,26 @@
 	public int getNumThreads()
 	{
 		return this.numThreads;
+	}
+
+	/************************************************************
+	 *  !ToDoo (Method description)
+	 *
+	 *@return    !ToDo (Return description)
+	 ***********************************************************/
+	public int getDefaultNumThreads()
+	{
+		return this.DEFAULT_NUM_THREADS;
+	}
+
+	/************************************************************
+	 *  !ToDoo (Method description)
+	 *
+	 *@return    !ToDo (Return description)
+	 ***********************************************************/
+	public int getDefaultRampUp()
+	{
+		return this.DEFAULT_RAMP_UP;
 	}
 
 	public synchronized void resetThreadCount()
