*** /home/jmheight/jakarta-jmeter/src/org/apache/jmeter/protocol/http/config/gui/MultipartUrlConfigGui.java	Wed Oct 24 19:34:31 2001
--- MultipartUrlConfigGui.java	Tue Oct 30 11:02:24 2001
***************
*** 127,130 ****
--- 127,131 ----
  		webRequestPanel.add(getParameterPanel());
  		webRequestPanel.add(getFilePanel());
+                 webRequestPanel.add(getBodyPanel());
  
  		// If displayName is TRUE, then this GUI is not embedded in another GUI.
