--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\protocol\http\config\UrlConfig.java	Thu Aug 30 17:19:02 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\protocol\http\config\UrlConfig.java	Sun Sep 16 18:09:14 2001
@@ -428,7 +428,15 @@
 			{
 				String name = args[i].substring(0,index);
 				String value = args[i].substring(index + 1);
-				addArgument(name, URLDecoder.decode(value));
+				try
+				{
+					addArgument(name, URLDecoder.decode(value));
+				}
+				catch (Exception e)
+				{
+					System.out.println("UrlConfig:parseArguments(): Unable to parse argument=[" + value + "]");
+					System.out.println("UrlConfig:parseArguments(): queryString=[" + queryString + "]");
+				}
 			}
 		}
 	}
