Index: src/core/org/apache/jmeter/gui/action/RemoteStart.java
===================================================================
--- src/core/org/apache/jmeter/gui/action/RemoteStart.java	(revision 1531796)
+++ src/core/org/apache/jmeter/gui/action/RemoteStart.java	(working copy)
@@ -127,7 +127,10 @@
     private void doRemoteStop(String name, boolean now) {
         GuiPackage.getInstance().getMainFrame().showStoppingMessage(name);
         JMeterEngine engine = remoteEngines.get(name);
-        engine.stopTest(now);
+        // Engine may be null if it has not correctly started
+        if(engine != null) {
+            engine.stopTest(now);
+        }
     }
 
     /**
