Index: src/core/org/apache/jmeter/control/gui/TestFragmentControllerGui.java
===================================================================
--- src/core/org/apache/jmeter/control/gui/TestFragmentControllerGui.java	(revision 1656228)
+++ src/core/org/apache/jmeter/control/gui/TestFragmentControllerGui.java	(working copy)
@@ -46,6 +46,7 @@
     @Override
     public TestElement createTestElement() {
         TestFragmentController controller = new TestFragmentController();
+        setEnabled(false);
         modifyTestElement(controller);
         return controller;
     }
