From 9c1e902b6f40a44a26bdb27e58c22058002e983b Mon Sep 17 00:00:00 2001
From: benoit <b.wiart@ubik-ingenierie.com>
Date: Sun, 14 Apr 2013 18:43:17 +0200
Subject: [PATCH] ajout icone


diff --git bin/jmeter.sh bin/jmeter.sh
index b0dfa63..a004ebf 100644
--- bin/jmeter.sh
+++ bin/jmeter.sh
@@ -31,7 +31,7 @@ JMETER_OPTS=""
 case $(uname) in
    Darwin*)
    # Add Mac-specific property - should be ignored elsewhere (Bug 47064)
-   JMETER_OPTS="-Xdock:name=JMeter -Dapple.laf.useScreenMenuBar=true -Dapple.eawt.quitStrategy=CLOSE_ALL_WINDOWS"
+   JMETER_OPTS="-Xdock:name=JMeter -Xdock:icon="`dirname $0`/../docs/images/logo.jpg" -Dapple.laf.useScreenMenuBar=true -Dapple.eawt.quitStrategy=CLOSE_ALL_WINDOWS"
    ;;
 esac
 
-- 
1.7.12

