diff -ur old/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngine.java new/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngine.java
--- old/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngine.java	Mon Mar  4 07:51:54 2002
+++ new/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngine.java	Mon May 13 02:27:30 2002
@@ -55,7 +55,6 @@
 package org.apache.jmeter.engine;
 
 import java.rmi.*;
-import org.apache.jmeter.samplers.RemoteSampleListener;
 import org.apache.jmeter.threads.ThreadGroup;
 
 /************************************************************
@@ -66,7 +65,7 @@
  ***********************************************************/
 public interface RemoteJMeterEngine extends Remote
 {
-	void addThreadGroup(ThreadGroup tGroup) throws RemoteException;
+	void addThreadGroup(org.apache.jmeter.threads.ThreadGroup tGroup) throws RemoteException;
 
 	void runTest() throws RemoteException;
 
diff -ur old/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java new/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java
--- old/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java	Mon Mar  4 07:51:54 2002
+++ new/jakarta-jmeter/src/org/apache/jmeter/engine/RemoteJMeterEngineImpl.java	Mon May 13 02:27:30 2002
@@ -57,7 +57,6 @@
 import java.rmi.*;
 
 import java.util.*;
-import org.apache.jmeter.threads.JMeterThread;
 import org.apache.jmeter.threads.ThreadGroup;
 
 /************************************************************
@@ -98,7 +97,7 @@
 	 *      attribute
 	 *@exception  RemoteException  Description of Exception
 	 ***********************************************************/
-	public void addThreadGroup(ThreadGroup tGroup) throws RemoteException
+	public void addThreadGroup(org.apache.jmeter.threads.ThreadGroup tGroup) throws RemoteException
 	{
 		backingEngine.addThreadGroup(tGroup);
 	}
