Index: JMeterThread.java
===================================================================
--- JMeterThread.java	(revision 1628107)
+++ JMeterThread.java	(working copy)
@@ -439,6 +439,10 @@
                     result.setThreadName(threadName);
                     threadContext.setPreviousResult(result);
                     runPostProcessors(pack.getPostProcessors());
+                    
+                    // Update the label of the sampler after post-processor
+                    result.setSampleLabel(pack.getSampler().getName());
+                    
                     checkAssertions(pack.getAssertions(), result, threadContext);
                     // Do not send subsamples to listeners which receive the transaction sample
                     List<SampleListener> sampleListeners = getSampleListeners(pack, transactionPack, transactionSampler);
