--- SSLManager.orig.java	Tue Aug  7 14:49:50 2007
+++ SSLManager.java	Tue Aug  7 14:51:51 2007
@@ -246,7 +246,7 @@
 	 */
 	public static final SSLManager getInstance() {
 		if (null == SSLManager.manager) {
-			return new JsseSSLManager(SSLManager.sslProvider);
+			SSLManager.manager = new JsseSSLManager(SSLManager.sslProvider);
 //			if (SSLManager.isSSLSupported) {
 //				String classname = null;
 //				classname = "org.apache.jmeter.util.JsseSSLManager"; // $NON-NLS-1$
