Index: src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java
===================================================================
--- src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java	(Revision 1740364)
+++ src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/config/DataSourceElement.java	(Arbeitskopie)
@@ -25,6 +25,7 @@
 import java.util.Set;
 
 import org.apache.commons.dbcp2.BasicDataSource;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.jmeter.config.ConfigElement;
 import org.apache.jmeter.testbeans.TestBean;
 import org.apache.jmeter.testbeans.TestBeanHelper;
@@ -209,7 +210,10 @@
 
         if(isKeepAlive()) {
             dataSource.setTestWhileIdle(true);
-            dataSource.setValidationQuery(getCheckQuery());
+            String validationQuery = getCheckQuery();
+            if (!StringUtils.isBlank(validationQuery)) {
+                dataSource.setValidationQuery(validationQuery);
+            }
             dataSource.setSoftMinEvictableIdleTimeMillis(Long.parseLong(getConnectionAge()));
             dataSource.setTimeBetweenEvictionRunsMillis(Integer.parseInt(getTrimInterval()));
         }
