Index: /src/protocol/http/org/apache/jmeter/protocol/http/proxy/HttpRequestHdr.java
===================================================================
--- /src/protocol/http/org/apache/jmeter/protocol/http/proxy/HttpRequestHdr.java	(revision 801478)
+++ /src/protocol/http/org/apache/jmeter/protocol/http/proxy/HttpRequestHdr.java	(working copy)
@@ -24,7 +24,6 @@
 import java.io.InputStream;
 import java.io.UnsupportedEncodingException;
 import java.net.MalformedURLException;
-import java.net.ProtocolException;
 import java.net.URL;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -238,7 +237,8 @@
         Iterator keys = headers.keySet().iterator();
         while (keys.hasNext()) {
             String key = (String) keys.next();
-            if (!key.equals(PROXY_CONNECTION) && !key.equals(CONTENT_LENGTH)) {
+            if (!key.equals(PROXY_CONNECTION) && !key.equals(CONTENT_LENGTH) 
+                    && !key.equals(HTTPConstants.HEADER_CONNECTION.toLowerCase())) {
                 manager.add((Header) headers.get(key));
             }
         }

