--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\timers\RandomTimer.java	Fri Aug 31 00:46:46 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\timers\RandomTimer.java	Tue Sep 18 14:16:18 2001
@@ -76,8 +76,10 @@
 		JMeterComponentModel, Saveable, Serializable
 {
 	protected Random random;
-	protected long delay = 300;
-	protected double range = 100.0;
+	protected final long DEFAULT_DELAY = 300;
+	protected final double DEFAULT_RANGE = 100.0;
+	protected long delay = DEFAULT_DELAY;
+	protected double range = DEFAULT_RANGE;
 	private String name;
 	private static List addableList = new LinkedList();
 
@@ -142,6 +144,16 @@
 	 *
 	 *@return    !ToDo (Return description)
 	 ***********************************************************/
+	public double getDefaultRange()
+	{
+		return DEFAULT_RANGE;
+	}
+
+	/************************************************************
+	 *  !ToDoo (Method description)
+	 *
+	 *@return    !ToDo (Return description)
+	 ***********************************************************/
 	public Class getTagHandlerClass()
 	{
 		return org.apache.jmeter.save.handlers.TimerHandler.class;
@@ -159,6 +171,15 @@
 		return delay;
 	}
 
+	/************************************************************
+	 *  !ToDoo (Method description)
+	 *
+	 *@return    !ToDo (Return description)
+	 ***********************************************************/
+	public long getDefaultDelay()
+	{
+		return DEFAULT_DELAY;
+	}
 
 	/************************************************************
 	 *  !ToDoo (Method description)
