Index: src/core/org/apache/jmeter/control/GenericController.java
===================================================================
--- src/core/org/apache/jmeter/control/GenericController.java	(revision 1166194)
+++ src/core/org/apache/jmeter/control/GenericController.java	(working copy)
@@ -226,8 +226,11 @@
                 if (currentElement instanceof Sampler) {
                     nextIsASampler((Sampler) currentElement);
                 } else { // must be a controller
-                    nextIsAController((Controller) currentElement);
-                    reInitializeSubController();
+                    Sampler s = nextIsAController((Controller) currentElement);
+                    if(s != null)
+                    {
+                        reInitializeSubController();                        
+                    }
                 }
             }
         } catch (NextIsNullException e) {
