Index: src/core/org/apache/jmeter/threads/TestCompiler.java
===================================================================
--- src/core/org/apache/jmeter/threads/TestCompiler.java	(revision 1242852)
+++ src/core/org/apache/jmeter/threads/TestCompiler.java	(working copy)
@@ -27,6 +27,7 @@
 import java.util.Set;
 
 import org.apache.jmeter.assertions.Assertion;
+import org.apache.jmeter.config.CSVDataSet;
 import org.apache.jmeter.config.ConfigTestElement;
 import org.apache.jmeter.control.Controller;
 import org.apache.jmeter.control.TransactionController;
@@ -285,7 +286,10 @@
     private void configureWithConfigElements(Sampler sam, List<ConfigTestElement> configs) {
         sam.clearTestElementChildren();
         for (ConfigTestElement config  : configs) {
-            sam.addTestElement(config);
+            if (!(config instanceof CSVDataSet)) 
+            {
+                sam.addTestElement(config);
+            }
         }
     }
 }
