--- /home/schizophrenia/test/jmeter/src/protocol/mail/org/apache/jmeter/protocol/smtp/sampler/protocol/SendMailCommand.java	2013-04-04 16:59:52.962771294 +0400
+++ src/protocol/mail/org/apache/jmeter/protocol/smtp/sampler/protocol/SendMailCommand.java	2013-04-04 17:03:50.498764277 +0400
@@ -219,7 +219,9 @@
             message.setReplyTo(replyTo.toArray(to));
         }
 
-        message.setSubject(subject);
+        if (null != subject) {
+            message.setSubject(subject);
+        }
 
         if (receiverTo != null) {
             InternetAddress[] to = new InternetAddress[receiverTo.size()];
@@ -772,4 +774,4 @@
     public void setReplyTo(List<InternetAddress> replyTo) {
         this.replyTo = replyTo;
     }
-}
\ No newline at end of file
+}
