Index: src/core/org/apache/jmeter/JMeter.java
===================================================================
--- src/core/org/apache/jmeter/JMeter.java	(revision 1645532)
+++ src/core/org/apache/jmeter/JMeter.java	(working copy)
@@ -97,9 +97,17 @@
  * Main JMeter class; processes options and starts the GUI, non-GUI or server as appropriate.
  */
 public class JMeter implements JMeterPlugin {
-
     private static final Logger log = LoggingManager.getLoggerForClass();
-
+/*    
+    static {
+        String jMeterLaf = LookAndFeelCommand.getJMeterLaf();
+        try {
+            UIManager.setLookAndFeel(jMeterLaf);
+        } catch (Exception e) {
+            log.warn("Could not set LAF to:"+jMeterLaf);
+        }
+    }
+*/
     public static final int UDP_PORT_DEFAULT = 4445; // needed for ShutdownClient
 
     public static final String HTTP_PROXY_PASS = "http.proxyPass"; // $NON-NLS-1$
