Index: src/core/org/apache/jmeter/gui/tree/JMeterCellRenderer.java
===================================================================
--- src/core/org/apache/jmeter/gui/tree/JMeterCellRenderer.java	(revision 1772362)
+++ src/core/org/apache/jmeter/gui/tree/JMeterCellRenderer.java	(working copy)
@@ -27,12 +27,17 @@
 import javax.swing.border.Border;
 import javax.swing.tree.DefaultTreeCellRenderer;
 
+import org.apache.commons.lang3.StringUtils;
+import org.apache.jorphan.util.JOrphanUtils;
+
 /**
  * Class to render the test tree - sets the enabled/disabled versions of the icons
  */
 public class JMeterCellRenderer extends DefaultTreeCellRenderer {
     private static final long serialVersionUID = 241L;
 
+    private static final int DEFAULT_LENGTH = 15;
+
     private static final Border RED_BORDER = BorderFactory.createLineBorder(Color.red);
 
     public JMeterCellRenderer() {
@@ -42,8 +47,9 @@
     public Component getTreeCellRendererComponent(JTree tree, Object value, boolean sel, boolean expanded,
             boolean leaf, int row, boolean p_hasFocus) {
         JMeterTreeNode node = (JMeterTreeNode) value;
-        super.getTreeCellRendererComponent(tree, node.getName(), sel, expanded, leaf, row,
-                p_hasFocus);
+        super.getTreeCellRendererComponent(tree,
+                JOrphanUtils.isBlank(node.getName()) ? StringUtils.repeat(' ', DEFAULT_LENGTH) : node.getName(), 
+                        sel, expanded, leaf, row, p_hasFocus);
         boolean enabled = node.isEnabled();
         ImageIcon ic = node.getIcon(enabled);
         if (ic != null) {
