diff -ruN jmeter-2.4.orig/src/protocol/http/org/apache/jmeter/protocol/http/modifier/URLRewritingModifier.java jmeter-2.4/src/protocol/http/org/apache/jmeter/protocol/http/modifier/URLRewritingModifier.java
--- jmeter-2.4.orig/src/protocol/http/org/apache/jmeter/protocol/http/modifier/URLRewritingModifier.java	2010-07-08 22:41:26.000000000 +0200
+++ jmeter-2.4/src/protocol/http/org/apache/jmeter/protocol/http/modifier/URLRewritingModifier.java	2010-11-16 18:53:21.643909702 +0100
@@ -150,19 +150,19 @@
     private void initRegex(String argName) {
         String quotedArg = Perl5Compiler.quotemeta(argName);// Don't get tripped up by RE chars in the arg name
         pathExtensionEqualsQuestionmarkRegexp = JMeterUtils.getPatternCache().getPattern(
-                SEMI_COLON + quotedArg + "=([^\"'>&\\s;]*)[&\\s\"'>;]?$?", // $NON-NLS-1$
+                SEMI_COLON + quotedArg + "=([^\"'<>&\\s;]*)[&\\s\"'>;]?$?", // $NON-NLS-1$
                 Perl5Compiler.MULTILINE_MASK | Perl5Compiler.READ_ONLY_MASK);
 
         pathExtensionEqualsNoQuestionmarkRegexp = JMeterUtils.getPatternCache().getPattern(
-                SEMI_COLON + quotedArg + "=([^\"'>&\\s;?]*)[&\\s\"'>;?]?$?", // $NON-NLS-1$
+                SEMI_COLON + quotedArg + "=([^\"'<>&\\s;?]*)[&\\s\"'>;?]?$?", // $NON-NLS-1$
                 Perl5Compiler.MULTILINE_MASK | Perl5Compiler.READ_ONLY_MASK);
 
         pathExtensionNoEqualsQuestionmarkRegexp = JMeterUtils.getPatternCache().getPattern(
-                SEMI_COLON + quotedArg + "([^\"'>&\\s;]*)[&\\s\"'>;]?$?", // $NON-NLS-1$
+                SEMI_COLON + quotedArg + "([^\"'<>&\\s;]*)[&\\s\"'>;]?$?", // $NON-NLS-1$
                 Perl5Compiler.MULTILINE_MASK | Perl5Compiler.READ_ONLY_MASK);
 
         pathExtensionNoEqualsNoQuestionmarkRegexp = JMeterUtils.getPatternCache().getPattern(
-                SEMI_COLON + quotedArg + "([^\"'>&\\s;?]*)[&\\s\"'>;?]?$?", // $NON-NLS-1$
+                SEMI_COLON + quotedArg + "([^\"'<>&\\s;?]*)[&\\s\"'>;?]?$?", // $NON-NLS-1$
                 Perl5Compiler.MULTILINE_MASK | Perl5Compiler.READ_ONLY_MASK);
 
         parameterRegexp = JMeterUtils.getPatternCache().getPattern(
