Index: src/core/org/apache/jmeter/threads/JMeterThread.java
===================================================================
--- src/core/org/apache/jmeter/threads/JMeterThread.java	(revision 1176336)
+++ src/core/org/apache/jmeter/threads/JMeterThread.java	(working copy)
@@ -321,13 +321,17 @@
             log.error("Test failed!", e);
         } finally {
             currentSampler = null; // prevent any further interrupts
-            interruptLock.lock();  // make sure current interrupt is finished, prevent another starting yet
-            threadContext.clear();
-            log.info("Thread finished: " + threadName);
-            threadFinished(iterationListener);
-            monitor.threadFinished(this); // Tell the engine we are done
-            JMeterContextService.removeContext(); // Remove the ThreadLocal entry
-            interruptLock.unlock(); // Allow any pending interrupt to complete (OK because currentSampler == null)
+            try {
+                interruptLock.lock();  // make sure current interrupt is finished, prevent another starting yet
+                threadContext.clear();
+                log.info("Thread finished: " + threadName);
+                threadFinished(iterationListener);
+                monitor.threadFinished(this); // Tell the engine we are done
+                JMeterContextService.removeContext(); // Remove the ThreadLocal entry
+            }
+            finally {
+                interruptLock.unlock(); // Allow any pending interrupt to complete (OK because currentSampler == null)
+            }
         }
     }
 
