Index: src/core/org/apache/jmeter/testelement/property/FunctionProperty.java
===================================================================
--- src/core/org/apache/jmeter/testelement/property/FunctionProperty.java	(revision 1166194)
+++ src/core/org/apache/jmeter/testelement/property/FunctionProperty.java	(working copy)
@@ -64,10 +64,11 @@
 
     @Override
     public int hashCode(){
-        int hash = super.hashCode();
+
         if (function != null) {
-            hash = hash*37 + function.hashCode();
+            return function.hashCode();
         }
+        int hash = super.hashCode();
         return hash;
     }
 
