Index: StringFromFile.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/functions/org/apache/jmeter/functions/StringFromFile.java,v
retrieving revision 1.6
diff -u -r1.6 StringFromFile.java
--- StringFromFile.java	5 Jun 2003 00:24:22 -0000	1.6
+++ StringFromFile.java	8 Jun 2003 18:57:59 -0000
@@ -67,7 +67,7 @@
 import org.apache.jmeter.samplers.Sampler;
 import org.apache.jmeter.threads.JMeterVariables;
 import org.apache.jmeter.util.JMeterUtils;
-import org.apache.log.Hierarchy;
+import org.apache.jorphan.logging.LoggingManager;
 import org.apache.log.Logger;
 
 
@@ -101,8 +101,7 @@
  */
 public class StringFromFile extends AbstractFunction implements Serializable
 {
-	transient private static Logger log = Hierarchy.getDefaultHierarchy().getLoggerFor(
-			"jmeter.elements");
+	transient private static Logger log = LoggingManager.getLoggerForClass();
 
 	private static final List desc = new LinkedList();
 	private static final String KEY = "_StringFromFile"; // Function name (only 1 _)
@@ -196,7 +195,7 @@
 		values = parameters.toArray();
 		
 		if (( values.length > 2 ) || (values.length < 1)) {
-			throw new InvalidVariableException();
+			throw new InvalidVariableException("Wrong number of parameters");
 		}
 		
 		openFile();
