--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\gui\MainFrame.java	Thu Jul 26 00:34:44 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\gui\MainFrame.java	Sun Sep 16 19:10:32 2001
@@ -160,7 +160,8 @@
 		JSplitPane treeAndMain = new JSplitPane(JSplitPane.HORIZONTAL_SPLIT);
 		treeAndMain.setLeftComponent(treePanel);
 		treeAndMain.setRightComponent(new JScrollPane(mainPanel));
-		treeAndMain.setResizeWeight(.2);
+		// Commented out by khammond 2001.09.16 - This method is new in JDK1.3.
+                //treeAndMain.setResizeWeight(.2);
 		treeAndMain.setContinuousLayout(true);
 		{
 			gbc.gridy++;
