Index: /workspace/rel-2-1/src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java
===================================================================
--- /workspace/rel-2-1/src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java	(revision 408847)
+++ /workspace/rel-2-1/src/components/org/apache/jmeter/visualizers/StatGraphVisualizer.java	(working copy)
@@ -120,6 +120,10 @@
     
     JLabeledTextField graphTitle = 
         new JLabeledTextField(JMeterUtils.getResString("aggregate_graph_user_title"));
+    
+    JLabeledTextField maxLengthXAxisLabel = 
+        new JLabeledTextField(JMeterUtils.getResString("aggregate_graph_max_length_xaxis_label"));
+    
     JLabeledTextField graphWidth = 
         new JLabeledTextField(JMeterUtils.getResString("aggregate_graph_width"));
     JLabeledTextField graphHeight = 
@@ -222,6 +226,7 @@
         
         graph.add(graphLabel);
         graph.add(graphTitle);
+        graph.add(maxLengthXAxisLabel);
         graph.add(graphWidth);
         graph.add(graphHeight);
         graph.add(buttonpanel);
@@ -246,19 +251,25 @@
     public void makeGraph() {
         String wstr = graphWidth.getText();
         String hstr = graphHeight.getText();
+        String lstr = maxLengthXAxisLabel.getText();
         if (wstr.length() == 0) {
             wstr = "450";
         }
         if (hstr.length() == 0) {
             hstr = "250";
         }
+        if (lstr.length() == 0) {
+            lstr = "20";
+        }
         int width = Integer.parseInt(wstr);
         int height = Integer.parseInt(hstr);
+        int maxLength = Integer.parseInt(lstr);
 
         graphPanel.setData(this.getData());
         graphPanel.setHeight(height);
         graphPanel.setWidth(width);
         graphPanel.setTitle(graphTitle.getText());
+        graphPanel.setMaxLenght(maxLength);
         graphPanel.setXAxisLabels(getAxisLabels());
         graphPanel.setXAxisTitle(columns.getText());
         graphPanel.setYAxisLabels(this.yAxisLabel);
