Index: org/apache/jmeter/threads/TestCompiler.java
===================================================================
--- org/apache/jmeter/threads/TestCompiler.java	(revision 1509349)
+++ org/apache/jmeter/threads/TestCompiler.java	(working copy)
@@ -137,7 +137,12 @@
     /** {@inheritDoc} */
     @Override
     public void addNode(Object node, HashTree subTree) {
+	//Added to resolve the bug Bug 55334 - https://issues.apache.org/bugzilla/show_bug.cgi?id=55334
+    	if(node instanceof JMeterTreeNode){
+    		stack.addLast(((JMeterTreeNode)node).getTestElement());
+    	}else{
         stack.addLast((TestElement) node);
+    	}
     }
 
     /** {@inheritDoc} */
