Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/tree/JMeterTreeModel.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/tree/JMeterTreeModel.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/tree/JMeterTreeModel.java	(working copy)
@@ -104,12 +104,8 @@
 		}
 		component.setProperty(TestElement.GUI_CLASS, NameUpdater.getCurrentName(component
 				.getPropertyAsString(TestElement.GUI_CLASS)));
+        // Save current GUI state to test element for current node
 		GuiPackage.getInstance().updateCurrentNode();
-		JMeterGUIComponent guicomp = GuiPackage.getInstance().getGui(component);
-		guicomp.configure(component);
-		guicomp.modifyTestElement(component);
-		GuiPackage.getInstance().getCurrentGui(); // put the gui object back
-													// to the way it was.
 		JMeterTreeNode newNode = new JMeterTreeNode(component, this);
 
 		// This check the state of the TestElement and if returns false it
