Index: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java	(revision 1606604)
+++ src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java	(working copy)
@@ -200,6 +200,15 @@
         }
     }
 
+    @Override
+    public String getDataEncodingWithDefault(String defaultEncoding) {
+        String dataEncodingNoDefault = getDataEncodingNoDefault();
+        if(dataEncodingNoDefault != null && dataEncodingNoDefault.length()> 0) {
+            return dataEncodingNoDefault;
+        }
+        return defaultEncoding;
+    }
+    
     /**
      * Overrides the method from SampleResult - so the encoding can be extracted from
      * the Meta content-type if necessary.
