Index: /src/core/org/apache/jmeter/engine/StandardJMeterEngine.java
===================================================================
--- /src/core/org/apache/jmeter/engine/StandardJMeterEngine.java	(revision 798325)
+++ /src/core/org/apache/jmeter/engine/StandardJMeterEngine.java	(working copy)
@@ -366,6 +366,9 @@
                 } // else will be done by threadFinished()
             } else {
                 stopAllThreads();
+                if (serialized) {
+                    notifyTestListenersOfEnd(testListenersSave);
+                }
             }
         }
     }

