Index: src/core/org/apache/jmeter/JMeter.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/core/org/apache/jmeter/JMeter.java,v
retrieving revision 1.8
diff -u -w -r1.8 JMeter.java
--- src/core/org/apache/jmeter/JMeter.java	11 Jan 2003 11:21:07 -0000	1.8
+++ src/core/org/apache/jmeter/JMeter.java	20 Jan 2003 10:00:16 -0000
@@ -363,7 +363,7 @@
 			println("Created the tree successfully");
 			StandardJMeterEngine engine = new StandardJMeterEngine();
 			engine.configure(tree);
-			println("Starting the test");
+			println("Starting the test: "+tree);
 			engine.runTest();
 
 		} catch (Exception e) {
