Index: src/protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java
===================================================================
--- src/protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java	(revision 1166194)
+++ src/protocol/java/org/apache/jmeter/protocol/java/control/gui/JavaTestSamplerGui.java	(working copy)
@@ -84,4 +84,13 @@
         super.configure(el);
         javaPanel.configure(el);
     }
+
+    /* (non-Javadoc)
+     * @see org.apache.jmeter.gui.AbstractJMeterGuiComponent#clearGui()
+     */
+    @Override
+    public void clearGui() {
+        super.clearGui();
+        javaPanel.clearGui();
+    }
 }
Index: src/protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java
===================================================================
--- src/protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java	(revision 1166194)
+++ src/protocol/java/org/apache/jmeter/protocol/java/config/gui/JavaConfigGui.java	(working copy)
@@ -239,4 +239,15 @@
         ((JavaConfig) config).setArguments((Arguments) argsPanel.createTestElement());
         ((JavaConfig) config).setClassname(String.valueOf(classnameCombo.getSelectedItem()));
     }
+
+    /* (non-Javadoc)
+     * @see org.apache.jmeter.gui.AbstractJMeterGuiComponent#clearGui()
+     */
+    @Override
+    public void clearGui() {
+        super.clearGui();
+        this.displayName = true;
+        argsPanel.clearGui();
+        classnameCombo.setSelectedIndex(0);
+    }
 }
