From 74e0b2232cccfac80e6236d2a81a46dd536bcf03 Mon Sep 17 00:00:00 2001
From: Felix Schumacher <felix.schumacher@internetallee.de>
Date: Sun, 21 Feb 2021 18:56:33 +0100
Subject: [PATCH] Correct usage of clipboard in systemsamplergui

---
 .../protocol/system/gui/SystemSamplerGui.java | 26 ++++++++++++-------
 1 file changed, 17 insertions(+), 9 deletions(-)

diff --git a/src/protocol/native/src/main/java/org/apache/jmeter/protocol/system/gui/SystemSamplerGui.java b/src/protocol/native/src/main/java/org/apache/jmeter/protocol/system/gui/SystemSamplerGui.java
index 6c884a531b..0fea2d7072 100644
--- a/src/protocol/native/src/main/java/org/apache/jmeter/protocol/system/gui/SystemSamplerGui.java
+++ b/src/protocol/native/src/main/java/org/apache/jmeter/protocol/system/gui/SystemSamplerGui.java
@@ -213,19 +213,27 @@ public class SystemSamplerGui extends AbstractSamplerGui implements ItemListener
         return panel;
     }
 
-
     /**
      * @return JPanel Arguments Panel
      */
     private JPanel makeArgumentsPanel() {
-        argsPanel = new ArgumentsPanel(JMeterUtils.getResString("arguments_panel_title"), null, true, false ,  // $NON-NLS-1$
-                new ObjectTableModel(new String[] { ArgumentsPanel.COLUMN_RESOURCE_NAMES_1 },
-                        Argument.class,
-                        new Functor[] {
-                        new Functor("getValue") },  // $NON-NLS-1$
-                        new Functor[] {
-                        new Functor("setValue") }, // $NON-NLS-1$
-                        new Class[] {String.class }));
+        final ObjectTableModel objectTableModel = new ObjectTableModel(
+                new String[] { ArgumentsPanel.COLUMN_RESOURCE_NAMES_1 }, Argument.class,
+                new Functor[] { new Functor("getValue") }, // $NON-NLS-1$
+                new Functor[] { new Functor("setValue") }, // $NON-NLS-1$
+                new Class[] { String.class });
+        argsPanel = new ArgumentsPanel(JMeterUtils.getResString("arguments_panel_title"), null, true, false, // $NON-NLS-1$
+                objectTableModel) {
+            private static final long serialVersionUID = 1L;
+
+            @Override
+            protected Argument createArgumentFromClipboard(String[] clipboardCols) {
+                Argument argument = makeNewArgument();
+                argument.setName(clipboardCols[0]);
+                argument.setValue(clipboardCols[0]);
+                return argument;
+            }
+        };
         return argsPanel;
     }
 
-- 
2.25.1

