Index: MachineName.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/functions/org/apache/jmeter/functions/MachineName.java,v
retrieving revision 1.7
diff -u -r1.7 MachineName.java
--- MachineName.java	10 Mar 2003 17:28:11 -0000	1.7
+++ MachineName.java	8 Jun 2003 19:03:39 -0000
@@ -75,6 +75,8 @@
 	private static final List desc = new LinkedList();
 	private static final String KEY = "__machineName";
 	
+	// Number of parameters expected - used to reject invalid calls
+	private static final int PARAMETER_COUNT = 1;
 	static {
 //		desc.add("Use fully qualified host name: TRUE/FALSE (Default FALSE)");
 		desc.add(JMeterUtils.getResString("function_name_param"));
@@ -126,8 +128,8 @@
 
 		values = parameters.toArray();
 		
-		if ( values.length < 2 ) {
-			throw new InvalidVariableException();
+		if ( values.length != PARAMETER_COUNT ) {
+			throw new InvalidVariableException("Parameter Count != "+PARAMETER_COUNT);
 		}
 
 	}
