Index: src/components/org/apache/jmeter/config/CSVDataSet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/components/org/apache/jmeter/config/CSVDataSet.java	(date 1388842110000)
+++ src/components/org/apache/jmeter/config/CSVDataSet.java	(revision )
@@ -156,7 +156,7 @@
             delim=",";
         }
         if (vars == null) {
-            String _fileName = getFilename();
+            String _fileName = getFilename().trim();
             String mode = getShareMode();
             int modeInt = CSVDataSetBeanInfo.getShareModeAsInt(mode);
             switch(modeInt){
Index: src/protocol/http/org/apache/jmeter/protocol/http/util/accesslog/TCLogParser.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/util/accesslog/TCLogParser.java	(date 1388842110000)
+++ src/protocol/http/org/apache/jmeter/protocol/http/util/accesslog/TCLogParser.java	(revision )
@@ -183,7 +183,7 @@
      */
     @Override
     public void setSourceFile(String source) {
-        this.FILENAME = source;
+        this.FILENAME = source.trim();
     }
 
     /**
\ No newline at end of file
