Index: ArgumentsPanel.java
===================================================================
--- ArgumentsPanel.java	(revision 1188659)
+++ ArgumentsPanel.java	(working copy)
@@ -358,32 +358,31 @@
     protected void deleteArgument() {
         cancelEditing();
 
-        int rowSelected = table.getSelectedRow();
-        if (rowSelected >= 0) {
-            tableModel.removeRow(rowSelected);
-            tableModel.fireTableDataChanged();
+        int[] rowsSelected = table.getSelectedRows();
+        int anchorSelection = table.getSelectionModel().getAnchorSelectionIndex();
+        table.clearSelection();
+        if (rowsSelected.length > 0) {
+            for (int i = rowsSelected.length - 1; i >= 0; i--) {
+                tableModel.removeRow(rowsSelected[i]);
+            }
 
             // Disable DELETE if there are no rows in the table to delete.
             if (tableModel.getRowCount() == 0) {
                 delete.setEnabled(false);
             }
+            // Table still contains one or more rows, so highlight (select)
+            // the appropriate one.
+            else if (tableModel.getRowCount() > 0) {
+                if (anchorSelection >= tableModel.getRowCount()) {
+                    anchorSelection = tableModel.getRowCount() - 1;
+                }
+                table.setRowSelectionInterval(anchorSelection, anchorSelection);
+            }
             
             if(enableUpDown && tableModel.getRowCount()>1) {
                 up.setEnabled(true);
                 down.setEnabled(true);
             }
-
-            // Table still contains one or more rows, so highlight (select)
-            // the appropriate one.
-            else {
-                int rowToSelect = rowSelected;
-
-                if (rowSelected >= tableModel.getRowCount()) {
-                    rowToSelect = rowSelected - 1;
-                }
-
-                table.setRowSelectionInterval(rowToSelect, rowToSelect);
-            }
         }
     }
 
@@ -467,7 +466,7 @@
         initializeTableModel();
         table = new JTable(tableModel);
         table.getTableHeader().setDefaultRenderer(new HeaderAsPropertyRenderer());
-        table.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
+        table.setSelectionMode(ListSelectionModel.MULTIPLE_INTERVAL_SELECTION);
         if (this.background != null) {
             table.setBackground(this.background);
         }
