From 282e945b5312f02ec26f9fe0298a1be20844afc9 Mon Sep 17 00:00:00 2001
From: benoit <b.wiart@ubik-ingenierie.com>
Date: Sat, 13 Apr 2013 23:43:22 +0200
Subject: [PATCH] handle mac cmd-Q


diff --git bin/jmeter.sh bin/jmeter.sh
index 125b2e9..b0dfa63 100644
--- bin/jmeter.sh
+++ bin/jmeter.sh
@@ -31,7 +31,7 @@ JMETER_OPTS=""
 case $(uname) in
    Darwin*)
    # Add Mac-specific property - should be ignored elsewhere (Bug 47064)
-   JMETER_OPTS="-Xdock:name=JMeter -Dapple.laf.useScreenMenuBar=true"
+   JMETER_OPTS="-Xdock:name=JMeter -Dapple.laf.useScreenMenuBar=true -Dapple.eawt.quitStrategy=CLOSE_ALL_WINDOWS"
    ;;
 esac
 
-- 
1.7.12

