Index: src/components/org/apache/jmeter/visualizers/RenderAsJSON.java
===================================================================
--- src/components/org/apache/jmeter/visualizers/RenderAsJSON.java	(Revision 1835976)
+++ src/components/org/apache/jmeter/visualizers/RenderAsJSON.java	(Arbeitskopie)
@@ -31,7 +31,7 @@
 import net.minidev.json.parser.ParseException;
 
 public class RenderAsJSON extends SamplerResultTab implements ResultRenderer {
-    private static final String TAB_SEPARATOR = ":   "; //$NON-NLS-1$
+    private static final String TAB_SEPARATOR = "    "; //$NON-NLS-1$
 
     /** {@inheritDoc} */
     @Override
Index: test/src/org/apache/jmeter/visualizers/TestRenderAsJson.java
===================================================================
--- test/src/org/apache/jmeter/visualizers/TestRenderAsJson.java	(Revision 1835976)
+++ test/src/org/apache/jmeter/visualizers/TestRenderAsJson.java	(Arbeitskopie)
@@ -28,7 +28,7 @@
 public class TestRenderAsJson {
 
     private Method prettyJSON;
-    private final String TAB = ":   ";
+    private final String TAB = "    ";
 
     private String prettyJSON(String prettify) throws Exception {
         return (String) prettyJSON.invoke(null, prettify);
