--- /src/components/org/apache/jmeter/extractor/XPathExtractor.java
+++ /otherpath/src/org/apache/jmeter/extractor/XPathExtractor.java
@@ -17,6 +17,8 @@
  */
 package org.apache.jmeter.extractor;
 
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.Serializable;
@@ -197,7 +199,24 @@
     }
 
     public String getXPathQuery(){
-        return getPropertyAsString(XPATH_QUERY);
+	String defaultValue=getPropertyAsString(XPATH_QUERY);
+
+	JMeterContext context = getThreadContext();
+	JMeterVariables vars = context.getVariables();
+
+	if ( defaultValue != null && vars != null ){
+		Pattern p = Pattern.compile("\\$\\{([a-zA-Z0-9]*)\\}");
+		Matcher m = p.matcher(defaultValue);
+		if ( m.find() ) {
+			for ( int i = 0; i < m.groupCount(); i++){
+				String vName = m.group(i);
+				if ( vars.get(vName) != null ){
+					defaultValue = defaultValue.replaceAll("\\$\\{"+vName+"\\}", vars.get(vName));
+				}
+			}
+		}
+	}
+        return defaultValue;
     }
 
     public void setRefName(String refName) {
