Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	(revision 512172)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSamplerBase.java	(working copy)
@@ -765,7 +765,9 @@
 				try {
 					URL url = (URL) binURL;
                     String urlstr = url.toString();
+                    // Encode any spaces, and change any &amp; to &
                     String urlStrEnc=encodeSpaces(urlstr);
+                    urlStrEnc = JOrphanUtils.substitute(urlStrEnc, "&amp;", "&"); // $NON-NLS-1$//$NON-NLS-2$
                     if (!urlstr.equals(urlStrEnc)){// There were some spaces in the URL
                         try {
                             url = new URL(urlStrEnc);