Index: SoapSampler.java
===================================================================
--- SoapSampler.java	(revision 587078)
+++ SoapSampler.java	(working copy)
@@ -23,6 +23,7 @@
 import org.apache.commons.httpclient.methods.RequestEntity;
 import org.apache.jmeter.protocol.http.control.Header;
 import org.apache.jmeter.protocol.http.control.HeaderManager;
+import org.apache.jmeter.testelement.property.BooleanProperty;
 import org.apache.jorphan.logging.LoggingManager;
 import org.apache.jorphan.util.JOrphanUtils;
 import org.apache.log.Logger;
@@ -53,11 +54,13 @@
 	public static final String SEND_SOAP_ACTION = "SoapSampler.SEND_SOAP_ACTION"; //$NON-NLS-1$
 
 	public static final String XML_DATA_FILE = "SoapSampler.xml_data_file"; //$NON-NLS-1$
+	
+	public final static String USE_KEEPALIVE = "HTTPSampler.use_keepalive"; // $NON-NLS-1$
 
 	private static final String DOUBLE_QUOTE = "\""; //$NON-NLS-1$
 
 	private static final String SOAPACTION = "SOAPAction"; //$NON-NLS-1$
-
+	
 	public void setXmlData(String data) {
 		setProperty(XML_DATA, data);
 	}
@@ -75,7 +78,7 @@
     public void setXmlFile(String filename) {
         setProperty(XML_DATA_FILE, filename);
     }
-
+    
     /**
      * Get the file location of the xml file.
      *
@@ -84,7 +87,16 @@
     public String getXmlFile() {
         return getPropertyAsString(XML_DATA_FILE);
     }
+    
+	public void setUseKeepAlive(boolean value) {
+		setProperty(new BooleanProperty(USE_KEEPALIVE, value));
+	}
 
+	public boolean getUseKeepAlive() {
+		return getPropertyAsBoolean(USE_KEEPALIVE);
+	}
+
+
 	public String getURLData() {
 		return getPropertyAsString(URL_DATA);
 	}
