? jmeter-noroot.diff
Index: src/components/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java
===================================================================
RCS file: /home/cvspublic/jakarta-jmeter/src/components/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java,v
retrieving revision 1.48
diff -c -r1.48 ViewResultsFullVisualizer.java
*** src/components/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java	26 Apr 2005 00:14:45 -0000	1.48
--- src/components/org/apache/jmeter/visualizers/ViewResultsFullVisualizer.java	27 Apr 2005 15:10:50 -0000
***************
*** 61,66 ****
--- 61,67 ----
  import javax.swing.tree.DefaultMutableTreeNode;
  import javax.swing.tree.DefaultTreeCellRenderer;
  import javax.swing.tree.DefaultTreeModel;
+ import javax.swing.tree.TreePath;
  import javax.swing.tree.TreeSelectionModel;
  import javax.xml.parsers.DocumentBuilder;
  import javax.xml.parsers.DocumentBuilderFactory;
***************
*** 167,172 ****
--- 168,177 ----
          DefaultMutableTreeNode currNode = new DefaultMutableTreeNode(res);
          treeModel.insertNodeInto(currNode, root, root.getChildCount());
          addSubResults(currNode, res);
+ 
+         if(root.getChildCount() == 1)
+         	jTree.expandPath(new TreePath(root));
+ 
          log.debug("End : updateGui1");
      }
  
***************
*** 665,671 ****
          jTree.getSelectionModel().setSelectionMode(
              TreeSelectionModel.SINGLE_TREE_SELECTION);
          jTree.addTreeSelectionListener(this);
!         jTree.setShowsRootHandles(true);
  
          JScrollPane treePane = new JScrollPane(jTree);
          treePane.setPreferredSize(new Dimension(200, 300));
--- 670,676 ----
          jTree.getSelectionModel().setSelectionMode(
              TreeSelectionModel.SINGLE_TREE_SELECTION);
          jTree.addTreeSelectionListener(this);
!         jTree.setRootVisible(false);
  
          JScrollPane treePane = new JScrollPane(jTree);
          treePane.setPreferredSize(new Dimension(200, 300));
