diff --git a/src/core/org/apache/jmeter/report/processor/graph/impl/TimeVSThreadGraphConsumer.java b/src/core/org/apache/jmeter/report/processor/graph/impl/TimeVSThreadGraphConsumer.java
index 48ced57..6285aba 100644
--- a/src/core/org/apache/jmeter/report/processor/graph/impl/TimeVSThreadGraphConsumer.java
+++ b/src/core/org/apache/jmeter/report/processor/graph/impl/TimeVSThreadGraphConsumer.java
@@ -65,7 +65,7 @@ public class TimeVSThreadGraphConsumer extends AbstractGraphConsumer {
 
         groupInfos.put(AbstractGraphConsumer.DEFAULT_GROUP, new GroupInfo(
                 new MeanAggregatorFactory(), new NameSeriesSelector(),
-                new ElapsedTimeValueSelector(), false, false));
+                new ElapsedTimeValueSelector(), false, true));
 
         return groupInfos;
     }
