diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java b/src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java
index 603a238..8637565 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/config/gui/UrlConfigGui.java
@@ -597,7 +597,7 @@ public class UrlConfigGui extends JPanel implements ChangeListener {
 
         if (notConfigOnly){
             method = new JLabeledChoice(JMeterUtils.getResString("method"), // $NON-NLS-1$
-                    HTTPSamplerBase.getValidMethodsAsArray());
+                    HTTPSamplerBase.getValidMethodsAsArray(), true);
         }
 
         JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
diff --git a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HttpWebdav.java b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HttpWebdav.java
index f7e4dbf..d2236f5 100644
--- a/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HttpWebdav.java
+++ b/src/protocol/http/org/apache/jmeter/protocol/http/sampler/HttpWebdav.java
@@ -67,6 +67,6 @@ public final class HttpWebdav extends HttpEntityEnclosingRequestBase {
      * @return <code>true</code> if method is a Webdav one
      */
     public static boolean isWebdavMethod(String method) {
-        return WEBDAV_METHODS.contains(method);
+        return true;
     }
 }
