Index: src/core/org/apache/jmeter/gui/MainFrame.java
===================================================================
--- src/core/org/apache/jmeter/gui/MainFrame.java    (revision 1239287)
+++ src/core/org/apache/jmeter/gui/MainFrame.java    (working copy)
@@ -420,6 +420,9 @@
         topAndDown.setResizeWeight(.8);
         topAndDown.setContinuousLayout(true);
         topAndDown.setBorder(null); // see bug jdk 4131528
+        if(!DISPLAY_LOGGER_PANEL) {
+            topAndDown.setDividerSize(0);
+        }
         
         mainPanel = createMainPanel();
         logPanel = createLoggerPanel();
