Index: LdapExtConfigGui.java
===================================================================
--- LdapExtConfigGui.java	(revision 901273)
+++ LdapExtConfigGui.java	(working copy)
@@ -228,7 +228,7 @@
             cl.show(cards, CARDS_BIND);
         } else if (testType.equals(LDAPExtSampler.SBIND)) {
             sbind.setSelected(true);
-            cl.show(cards, CARDS_DEFAULT);
+            cl.show(cards, CARDS_BIND);
         } else if (testType.equals(LDAPExtSampler.COMPARE)) {
             compare.setSelected(true);
             cl.show(cards, CARDS_COMPARE);
