--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\timers\gui\GaussianRandomTimerGui.java	Thu Jul 26 00:34:52 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\timers\gui\GaussianRandomTimerGui.java	Tue Sep 18 14:13:18 2001
@@ -109,11 +109,45 @@
 		}
 		else if (temp.equals("Timer Delay"))
 		{
-			model.setDelay(Long.parseLong(delayField.getText()));
+			try
+			{
+				model.setDelay(Long.parseLong(delayField.getText()));
+			}
+			catch (NumberFormatException nfe)
+			{
+				if (delayField.getText().length() > 0)
+				{
+					JOptionPane.showMessageDialog(this, "You must enter a valid number",
+							"Invalid data", JOptionPane.WARNING_MESSAGE);
+					model.setDelay(model.getDefaultDelay());
+					delayField.setText("");
+				}
+				else
+				{
+					model.setDelay(model.getDefaultDelay());
+				}
+			}
 		}
 		else if (temp.equals("Timer Delay Range"))
 		{
-			model.setRange(Double.parseDouble(rangeField.getText()));
+			try
+			{
+				model.setRange(Double.parseDouble(rangeField.getText()));
+			}
+			catch (NumberFormatException nfe)
+			{
+				if (rangeField.getText().length() > 0)
+				{
+					JOptionPane.showMessageDialog(this, "You must enter a valid number",
+							"Invalid data", JOptionPane.WARNING_MESSAGE);
+					model.setRange(model.getDefaultRange());
+					rangeField.setText("");
+				}
+				else
+				{
+					model.setRange(model.getDefaultRange());
+				}
+			}
 		}
 	}
 
