--- \jakarta-jmeter-cvs\jakarta-jmeter\src\org\apache\jmeter\timers\ConstantTimer.java	Fri Aug 31 00:46:46 2001
+++ \jakarta-jmeter-nightly\2001.09.15\jakarta-jmeter\src\org\apache\jmeter\timers\ConstantTimer.java	Tue Sep 18 14:04:42 2001
@@ -77,7 +77,8 @@
 public class ConstantTimer implements Timer, JMeterComponentModel, Saveable,
 		Serializable
 {
-	private long delay = 300;
+	private final long DEFAULT_DELAY = 300;
+	private long delay = DEFAULT_DELAY;
 	private String name;
 	private static List addableList = new LinkedList();
 
@@ -135,6 +136,16 @@
 	public long getDelay()
 	{
 		return delay;
+	}
+
+	/************************************************************
+	 *  !ToDoo (Method description)
+	 *
+	 *@return    !ToDo (Return description)
+	 ***********************************************************/
+	public long getDefaultDelay()
+	{
+		return DEFAULT_DELAY;
 	}
 
 	/************************************************************
