Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/action/Save.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/action/Save.java	(revision 536067)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/gui/action/Save.java	(working copy)
@@ -117,6 +117,7 @@
 		} finally {
             JOrphanUtils.closeQuietly(ostream);
 		}
+        GuiPackage.getInstance().updateCurrentGui();
 	}
 
 	// package protected to all for separate test code
