Index: src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java	(revision 816394)
+++ src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPSampleResult.java	(working copy)
@@ -39,7 +39,7 @@
 
     public HTTPSampleResult() {
         super();
-        setDataEncoding(DEFAULT_HTTP_ENCODING); // default if encoding not provided be the page
+//        setDataEncoding(DEFAULT_HTTP_ENCODING); // default if encoding not provided be the page
     }
 
     public HTTPSampleResult(long elapsed) {
