Index: src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java	(revision 6d30c96148394d484dddb09e07b7bf4856e406b0)
+++ src/protocol/http/org/apache/jmeter/protocol/http/proxy/ProxyControl.java	(revision )
@@ -798,7 +798,10 @@
             try {
                 log.debug("Creating HTTP Authentication manager for authorization:"+authorization);
                 AuthManager authManager = newAuthorizationManager(authorization);
-                jmeterTreeModel.addComponent(authManager, target);
+                JMeterTreeNode authManagerNode = new JMeterTreeNode(authManager,jmeterTreeModel);
+                int index=(target.getChildCount()>0&&
+                        ((JMeterTreeNode)target.getFirstChild()).getTestElement() instanceof TestPlan)?2:0;
+                jmeterTreeModel.insertNodeInto(authManagerNode, target,index);
             } catch (IllegalUserActionException e) {
                 log.error("Failed to add Authorization Manager to target node:" + target.getName(), e);
             }
