Index: src/protocol/http/org/apache/jmeter/protocol/http/control/DNSCacheManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/protocol/http/org/apache/jmeter/protocol/http/control/DNSCacheManager.java	(revision c6287a820eb61f5fae533647c7da68a5e335a303)
+++ src/protocol/http/org/apache/jmeter/protocol/http/control/DNSCacheManager.java	(revision )
@@ -62,11 +62,11 @@
 
     private static final Logger log = LoggingManager.getLoggerForClass();
 
-    private SystemDefaultDnsResolver systemDefaultDnsResolver = null;
+    private transient SystemDefaultDnsResolver systemDefaultDnsResolver = null;
 
     private Map<String, InetAddress[]> cache = null;
 
-    private Resolver resolver = null;
+    private transient Resolver resolver = null;
 
     //++ JMX tag values
     public static final String CLEAR_CACHE_EACH_ITER = "DNSCacheManager.clearEachIteration"; // $NON-NLS-1$
\ No newline at end of file
