Index: src/core/org/apache/jmeter/reporters/ResultCollector.java
===================================================================
--- src/core/org/apache/jmeter/reporters/ResultCollector.java	(revision 1615530)
+++ src/core/org/apache/jmeter/reporters/ResultCollector.java	(working copy)
@@ -592,6 +592,15 @@
     }
 
     /**
+     * Flush PrintWriter to synchronize file contents
+     */
+    protected void flushFile() {
+	if (out != null) {
+            out.flush();
+	}
+    }
+
+    /**
      * Flush PrintWriter, called by Shutdown Hook to ensure no data is lost
      */
     private static void flushFileOutput() {
