Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/control/gui/HttpTestSampleGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/control/gui/HttpTestSampleGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/control/gui/HttpTestSampleGui.java	(working copy)
@@ -140,6 +140,8 @@
 	public void clear() {
 		super.clear();
 		getImages.setSelected(false);
+		isMon.setSelected(false);
 		urlConfigGui.clear();
+		embeddedRE.setText(""); // $NON-NLS-1$
 	}
 }
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/URLRewritingModifierGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/URLRewritingModifierGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/URLRewritingModifierGui.java	(working copy)
@@ -104,6 +104,19 @@
         rewritingModifier.setShouldCache((shouldCache.isSelected()));
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        argumentName.setText(""); //$NON-NLS-1$
+        pathExt.setSelected(false);
+        pathExtNoEquals.setSelected(false);
+        pathExtNoQuestionmark.setSelected(false);
+        shouldCache.setSelected(false);
+    }    
+
 	public void configure(TestElement el) {
 		URLRewritingModifier rewritingModifier = ((URLRewritingModifier) el);
         argumentName.setText(rewritingModifier.getArgumentName());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/UserParameterModifierGui.java	(working copy)
@@ -68,6 +68,14 @@
 		this.configureTestElement(mod);
 		((UserParameterModifier) mod).setXmlUri(fileNameField.getText());
 	}
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        fileNameField.setText("users.xml"); //$NON-NLS-1$
+    }    
 
 	public void updateGui() {
 	}
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/http/org/apache/jmeter/protocol/http/modifier/gui/ParamModifierGui.java	(working copy)
@@ -106,6 +106,20 @@
 			mask.setSuffix(_suffix.getText());
 		}
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        _fieldName.setText(""); //$NON-NLS-1$
+        _prefix.setText(""); //$NON-NLS-1$
+        _lowerBound.setText("0"); //$NON-NLS-1$
+        _upperBound.setText("10"); //$NON-NLS-1$
+        _increment.setText("1"); //$NON-NLS-1$
+        _suffix.setText(""); //$NON-NLS-1$        
+    }    
 
 	public void focusGained(FocusEvent evt) {
 	}
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BSFSamplerGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BSFSamplerGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/protocol/java/org/apache/jmeter/protocol/java/control/gui/BSFSamplerGui.java	(working copy)
@@ -78,6 +79,18 @@
 		te.setProperty(BSFSampler.PARAMETERS, parameters.getText());
 		te.setProperty(BSFSampler.SCRIPT, scriptField.getText());
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        scriptField.setText(""); //$NON-NLS-1$
+        langField.setText(""); //$NON-NLS-1$
+        filename.setText(""); //$NON-NLS-1$
+        parameters.setText(""); //$NON-NLS-1$
+    }    
 
 	public String getLabelResource() {
 		return "bsf_sampler_title";
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/UserParametersGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/UserParametersGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/UserParametersGui.java	(working copy)
@@ -129,6 +129,17 @@
 		super.configureTestElement(params);
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        initTableModel();
+        paramTable.setModel(tableModel);
+        perIterationCheck.setSelected(false);
+    }    
+
 	private void init() {
 		setBorder(makeBorder());
 		setLayout(new BorderLayout());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/CounterConfigGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/CounterConfigGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/modifiers/gui/CounterConfigGui.java	(working copy)
@@ -73,6 +74,20 @@
 		}
 		super.configureTestElement(c);
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        startField.setText(""); //$NON-NLS-1$
+        incrField.setText(""); //$NON-NLS-1$
+        endField.setText(""); //$NON-NLS-1$
+        varNameField.setText(""); //$NON-NLS-1$
+        formatField.setText(""); //$NON-NLS-1$
+        perUserField.setSelected(false);
+    }
 
 	public void configure(TestElement element) {
 		super.configure(element);
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/sampler/gui/TestActionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/sampler/gui/TestActionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/sampler/gui/TestActionGui.java	(working copy)
@@ -128,6 +129,22 @@
 		ta.setDuration(duration);
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        targetBox.setSelectedIndex(0);
+        durationField.setText(""); //$NON-NLS-1$
+        pauseButton.setSelected(true);
+        stopButton.setSelected(false);
+        action = TestAction.PAUSE;
+        target = TestAction.THREAD;
+        duration = 0;
+        
+    }    
+
 	private void init() {
 		setLayout(new VerticalLayout(5, VerticalLayout.LEFT, VerticalLayout.TOP));
 		setBorder(makeBorder());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/HTMLAssertionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/HTMLAssertionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/HTMLAssertionGui.java	(working copy)
@@ -138,6 +138,22 @@
 		((HTMLAssertion) inElement).setFilename(filePanel.getFilename());
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+
+        docTypeBox.setSelectedIndex(0);
+        htmlRadioButton.setSelected(true);
+        xhtmlRadioButton.setSelected(false);
+        xmlRadioButton.setSelected(false);
+        errorThresholdField.setText("0"); //$NON-NLS-1$
+        warningThresholdField.setText("0"); //$NON-NLS-1$
+        filePanel.setFilename(""); //$NON-NLS-1$
+        errorsOnly.setSelected(false);
+    }    
+
 	/**
 	 * Configures the associated test element.
 	 * 
@@ -161,6 +177,10 @@
 			warningThresholdField.setEnabled(false);
 			warningThresholdField.setEditable(false);
 		}
+		else {
+			warningThresholdField.setEnabled(true);
+			warningThresholdField.setEditable(true);
+		}
 		filePanel.setFilename(lAssertion.getFilename());
 	}
 
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathAssertionGui.java	(working copy)
@@ -116,4 +116,16 @@
 			assertion.setXPathString(xpath.getXPath());
 		}
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        xpath.setXPath("/"); //$NON-NLS-1$
+        xpath.setNegated(false);
+        
+        xml.setDefaultValues();
+    }    
 }
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/DurationAssertionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/DurationAssertionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/DurationAssertionGui.java	(working copy)
@@ -68,6 +68,15 @@
 		el.setProperty(DurationAssertion.DURATION_KEY,duration.getText());
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        duration.setText(""); //$NON-NLS-1$
+    }    
+
 	public void configure(TestElement el) {
 		super.configure(el);
 		duration.setText(el.getPropertyAsString(DurationAssertion.DURATION_KEY));
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathPanel.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathPanel.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XPathPanel.java	(working copy)
@@ -68,6 +68,7 @@
 	 */
 	public XPathPanel(boolean isDoubleBuffered) {
 		super(isDoubleBuffered);
+        init();
 	}
 
 	/**
@@ -75,6 +76,7 @@
 	 */
 	public XPathPanel(LayoutManager layout) {
 		super(layout);
+        init();
 	}
 
 	/**
@@ -83,7 +85,7 @@
 	 */
 	public XPathPanel(LayoutManager layout, boolean isDoubleBuffered) {
 		super(layout, isDoubleBuffered);
-
+		init();
 	}
 
 	private void init() {
@@ -99,8 +101,15 @@
 		vbox.add(getNegatedCheckBox());
 
 		add(vbox);
+        
+        setDefaultValues();
 	}
 
+    public void setDefaultValues() {
+        setXPath("/"); //$NON-NLS-1$
+        setNegated(false);
+    }
+
 	/**
 	 * Get the XPath String
 	 * 
@@ -170,7 +179,6 @@
 	public JTextField getXPathTextField() {
 		if (xpath == null) {
 			xpath = new JTextField(50);
-			xpath.setText("/"); //$NON-NLS-1$
 		}
 		return xpath;
 	}
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLConfPanel.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLConfPanel.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLConfPanel.java	(working copy)
@@ -51,8 +51,16 @@
 		add(getNamespace());
 		add(getValidate());
 		add(getWhitespace());
+		setDefaultValues();
 	}
 
+    public void setDefaultValues() {
+        setWhitespace(false);
+        setValidate(false);
+        setTolerant(false);
+        setNamespace(false);
+    }
+
 	/**
 	 * @return Returns the namespace.
 	 */
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/MD5HexAssertionGUI.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/MD5HexAssertionGUI.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/MD5HexAssertionGUI.java	(working copy)
@@ -102,4 +102,13 @@
 		}
 		((MD5HexAssertion) element).setAllowedMD5Hex(md5HexString);
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        md5HexInput.setText(""); //$NON-NLS-1$
+    }    
 }
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/SizeAssertionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/SizeAssertionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/SizeAssertionGui.java	(working copy)
@@ -84,6 +84,22 @@
 		((SizeAssertion) el).setAllowedSize(assertionSize);
 		((SizeAssertion) el).setCompOper(getState());
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        size.setText(""); //$NON-NLS-1$
+        equalButton.setSelected(true);
+        notequalButton.setSelected(false);
+        greaterthanButton.setSelected(false);
+        lessthanButton.setSelected(false);
+        greaterthanequalButton.setSelected(false);
+        lessthanequalButton.setSelected(false);
+        execState = SizeAssertion.EQUAL;
+    }    
 
 	public void configure(TestElement el) {
 		super.configure(el);
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/AssertionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/AssertionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/AssertionGui.java	(working copy)
@@ -160,7 +160,27 @@
 			}
 		}
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        tableModel.clearData();
 
+        responseStringButton.setSelected(true);
+        urlButton.setSelected(false);
+        responseCodeButton.setSelected(false);
+        responseMessageButton.setSelected(false);
+        assumeSuccess.setSelected(false);
+        
+        containsBox.setSelected(true);
+        matchesBox.setSelected(false);
+        equalsBox.setSelected(false);
+        notBox.setSelected(false);
+    }    
+
 	/**
 	 * A newly created component can be initialized with the contents of a Test
 	 * Element object by calling this method. The component is responsible for
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLSchemaAssertionGUI.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLSchemaAssertionGUI.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/assertions/gui/XMLSchemaAssertionGUI.java	(working copy)
@@ -81,6 +81,15 @@
 		((XMLSchemaAssertion) inElement).setXsdFileName(xmlSchema.getText());
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        xmlSchema.setText(""); //$NON-NLS-1$
+    }    
+
 	/**
 	 * Configures the GUI from the associated test element.
 	 * 
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/XPathExtractorGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/XPathExtractorGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/XPathExtractorGui.java	(working copy)
@@ -88,6 +88,18 @@
         }
     }
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        xpathQueryField.setText(""); // $NON-NLS-1$
+        defaultField.setText(""); // $NON-NLS-1$
+        refNameField.setText(""); // $NON-NLS-1$
+        tolerant.setSelected(false);
+    }
+
 	private void init() {
 		setLayout(new BorderLayout());
 		setBorder(makeBorder());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/RegexExtractorGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/RegexExtractorGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/components/org/apache/jmeter/extractor/gui/RegexExtractorGui.java	(working copy)
@@ -108,6 +108,23 @@
 			regex.setMatchNumber(matchNumberField.getText());
 		}
 	}
+    
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        useBody.setSelected(true);
+        useHeaders.setSelected(false);
+        useURL.setSelected(false);
+        
+        regexField.setText(""); //$NON-NLS-1$
+        templateField.setText(""); //$NON-NLS-1$
+        defaultField.setText(""); //$NON-NLS-1$
+        refNameField.setText(""); //$NON-NLS-1$
+        matchNumberField.setText(""); //$NON-NLS-1$
+    }    
 
 	private void init() {
 		setLayout(new BorderLayout());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultActionGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultActionGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultActionGui.java	(working copy)
@@ -76,6 +76,15 @@
 		((OnErrorTestElement) te).setErrorAction(errorPanel.getOnErrorSetting());
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        errorPanel.configure(OnErrorTestElement.ON_ERROR_CONTINUE);
+    }
+    
 	private void init() {
 		setLayout(new BorderLayout());
 		setBorder(makeBorder());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultSaverGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultSaverGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/reporters/gui/ResultSaverGui.java	(working copy)
@@ -85,6 +85,16 @@
 		te.setProperty(ResultSaver.ERRORS_ONLY, errorsOnly.isSelected());
 	}
 
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        filename.setText(""); //$NON-NLS-1$
+        errorsOnly.setSelected(false);
+    }
+
 	private void init() {
 		setLayout(new BorderLayout());
 		setBorder(makeBorder());
Index: C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/config/gui/LoginConfigGui.java
===================================================================
--- C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/config/gui/LoginConfigGui.java	(revision 534327)
+++ C:/Documents and Settings/alf.hogemark/workspace/Jmeter 2.2 official/src/core/org/apache/jmeter/config/gui/LoginConfigGui.java	(working copy)
@@ -107,6 +108,15 @@
 		String passwordString = new String(password.getPassword());
 		element.setProperty(new StringProperty(ConfigTestElement.PASSWORD, passwordString));
 	}
+    /**
+     * Implements JMeterGUIComponent.clear
+     */
+    public void clear() {
+        super.clear();
+        
+        username.setText(""); //$NON-NLS-1$
+        password.setText(""); //$NON-NLS-1$
+    }    
 
 	/**
 	 * Initialize the components and layout of this component.
